# Package index

## All functions

- [`add_new_metadata()`](add_new_metadata.md) : Replace the meta data

- [`add_nhoods_info()`](add_nhoods_info.md) : Add neighbourhood info on
  majority cell type

- [`add_snf_data_modality()`](add_snf_data_modality.md) : Add a data
  modality for SNF generation

- [`all_limma_contrasts()`](all_limma_contrasts.md) : Create all limma
  contrasts (based on combination of everything)

- [`anndata_parser`](anndata_parser.md) : Class for Anndata

- [`assertCellsExist()`](assertCellsExist.md) : Assert neighbour
  generation parameters

- [`assertCistargetParams()`](assertCistargetParams.md) : Assert
  CisTarget parameters

- [`assertCoReMoParams()`](assertCoReMoParams.md) : Assert CoReMo
  parameter

- [`assertCommunityParams()`](assertCommunityParams.md) : Assert
  community detection parameter

- [`assertCorGraphParams()`](assertCorGraphParams.md) : Assert
  correlation graph parameters

- [`assertDGRDLparams()`](assertDGRDLparams.md) : Assert DGRDL parameter

- [`assertFileExists()`](assertFileExists.md) : Assert that files exist

- [`assertGSEAParams()`](assertGSEAParams.md) : Assert GSEA parameter

- [`assertGSVAParams()`](assertGSVAParams.md) : Assert GSVA parameter

- [`assertGraphResParams()`](assertGraphResParams.md) : Assert
  resolution graph parameters

- [`assertIcaIterParams()`](assertIcaIterParams.md) : Assert ICA
  randomisation parameters

- [`assertIcaNcomps()`](assertIcaNcomps.md) : Assert ICA no of component
  parameters

- [`assertIcaParams()`](assertIcaParams.md) : Assert ICA parameters

- [`assertSNFParams()`](assertSNFParams.md) : Assert SNF parameter

- [`assertScBbknn()`](assertScBbknn.md) : Assert BBKNN parameters

- [`assertScBoost()`](assertScBoost.md) : Assert Boost parameters

- [`assertScFastmnn()`](assertScFastmnn.md) : Assert fastMNN parameters

- [`assertScHotspot()`](assertScHotspot.md) : Assert HotSpot parameters

- [`assertScHvg()`](assertScHvg.md) : Assert HVG selection parameters

- [`assertScMetacells()`](assertScMetacells.md) : Assert metacell
  generation parameters

- [`assertScMiloR()`](assertScMiloR.md) : Assert MiloR parameters

- [`assertScMinQC()`](assertScMinQC.md) : Assert SC minimum QC
  parameters

- [`assertScMtxIO()`](assertScMtxIO.md) : Assert SC MTX load parameters

- [`assertScNeighbours()`](assertScNeighbours.md) : Assert neighbour
  generation parameters

- [`assertScScrublet()`](assertScScrublet.md) : Assert Scrublet
  parameters

- [`assertScSeacells()`](assertScSeacells.md) : Assert SEACells
  parameters

- [`assertScSupercell()`](assertScSupercell.md) : Assert SuperCell
  parameters

- [`assertScSyntheticData()`](assertScSyntheticData.md) : Assert
  synthetic data parameters

- [`assertScVision()`](assertScVision.md) : Assert VISION parameters

- [`assertSingleSampleGSEAparams()`](assertSingleSampleGSEAparams.md) :
  Assert ssGSEA parameter

- [`aucell_sc()`](aucell_sc.md) : Calculate AUC scores (akin to AUCell)

- [`batch_correction_bulk_dge()`](batch_correction_bulk_dge.md) : Run a
  linear batch correction

- [`bbknn_sc()`](bbknn_sc.md) : Run BBKNN

- [`bixverse_base_class()`](bixverse_base_class.md) : bixverse base
  class

- [`bulk_coexp()`](bulk_coexp.md) : Bulk RNAseq co-expression modules

- [`bulk_dge()`](bulk_dge.md) : Bulk RNAseq differential gene expression
  class

- [`bulk_dge_from_h5ad()`](bulk_dge_from_h5ad.md) : Wrapper function to
  generate bulk_dge object from h5ad

- [`c_pca_plot_alphas()`](c_pca_plot_alphas.md) : Plot various alphas
  for the contrastive PCA

- [`calc_fgsea()`](calc_fgsea.md) : Bixverse implementation of the fgsea
  algorithm

- [`calc_fgsea_simple()`](calc_fgsea_simple.md) : Bixverse
  implementation of the simple fgsea algorithm

- [`calc_gsea_traditional()`](calc_gsea_traditional.md) : Bixverse
  implementation of the traditional GSEA algorithm

- [`calc_gsva()`](calc_gsva.md) : Bixverse implementation of GSVA

- [`calc_meta_cell_purity()`](calc_meta_cell_purity.md) : Calculate meta
  cell purity

- [`calc_mitch()`](calc_mitch.md) : Calculate a mitch gene set
  enrichments on contrast

- [`calc_ssgsea()`](calc_ssgsea.md) : Bixverse implementation of ssGSEA

- [`calculate_all_dges()`](calculate_all_dges.md) : Calculate all
  possible DGE variants (DEPRECATED!)

- [`calculate_critical_value()`](calculate_critical_value.md) :
  Calculates the critical value

- [`calculate_dge_hedges()`](calculate_dge_hedges.md) : Calculates the
  Hedge's G effect size

- [`calculate_dge_limma()`](calculate_dge_limma.md) : Calculates the
  Limma Voom DGE

- [`calculate_diffusion_auc()`](calculate_diffusion_auc.md) : Calculate
  the AUROC for a diffusion score

- [`calculate_effect_size()`](calculate_effect_size.md) : Calculate the
  Hedge G effect between two matrices

- [`calculate_information_content()`](calculate_information_content.md)
  : Calculate the information content for each ontology term

- [`calculate_kbet_sc()`](calculate_kbet_sc.md) : Calculate kBET scores

- [`calculate_pca_bulk_dge()`](calculate_pca_bulk_dge.md) : Calculate
  PCA on the expression.

- [`calculate_pca_sc()`](calculate_pca_sc.md) : Run PCA for single cell

- [`calculate_rpkm()`](calculate_rpkm.md) : RPKM calculation

- [`calculate_semantic_sim()`](calculate_semantic_sim.md) : Calculate
  the Resnik or Lin semantic similarity

- [`calculate_semantic_sim_mat()`](calculate_semantic_sim_mat.md) :
  Calculate the Resnik or Lin semantic similarity matrix

- [`calculate_semantic_sim_onto()`](calculate_semantic_sim_onto.md) :
  Calculate the Resnik or Lin semantic similarity for an ontology.

- [`calculate_sparsity_stats()`](calculate_sparsity_stats.md) : Helper
  function to calculate the induced sparsity

- [`calculate_tom()`](calculate_tom.md) : Calculate the TOM from a
  correlation matrix

- [`calculate_tom_from_exp()`](calculate_tom_from_exp.md) : Calculate
  the TOM from an expression matrix

- [`calculate_tpm()`](calculate_tpm.md) : TPM calculation

- [`calculate_wang_sim()`](calculate_wang_sim.md) : Calculate the Wang
  similarities between terms

- [`calculate_wang_sim_mat()`](calculate_wang_sim_mat.md) : Calculate
  the Wang similarity matrix

- [`calculate_wang_sim_onto()`](calculate_wang_sim_onto.md) : Calculate
  the Wang similarity for an ontology.

- [`call_doublets_manual()`](call_doublets_manual.md) : Helper function
  to manually readjust Scrublet thresholds

- [`cell_cycle_genes`](cell_cycle_genes.md) : Cell cycle genes

- [`change_gene_identifier()`](change_gene_identifier.md) : Change the
  primary gene identifier of bulk_dge

- [`checkCellsExist()`](checkCellsExist.md) : Check that the cell name
  exists in the object

- [`checkCistargetParams()`](checkCistargetParams.md) : Check CisTarget
  parameters

- [`checkCoReMoParams()`](checkCoReMoParams.md) : Check CoReMo
  parameters

- [`checkCommunityParams()`](checkCommunityParams.md) : Check community
  detection parameters

- [`checkCorGraphParams()`](checkCorGraphParams.md) : Check correlation
  graph parameters

- [`checkDGRDLparams()`](checkDGRDLparams.md) : Check DGRDL parameters

- [`checkFilesExist()`](checkFilesExist.md) : Check that files exist

- [`checkGSEAParams()`](checkGSEAParams.md) : Check GSEA parameters

- [`checkGSVAParams()`](checkGSVAParams.md) : Check GSVA parameters

- [`checkGraphResParams()`](checkGraphResParams.md) : Check resolution
  graph parameters

- [`checkIcaIterParams()`](checkIcaIterParams.md) : Check ICA
  randomisation parameters

- [`checkIcaNcomps()`](checkIcaNcomps.md) : Check ICA no of component
  parameters

- [`checkIcaParams()`](checkIcaParams.md) : Check ICA parameters

- [`checkSNFParams()`](checkSNFParams.md)
  [`testSNFParams()`](checkSNFParams.md) : Check SNF parameters

- [`checkScBbknn()`](checkScBbknn.md) : Check BBKNN parameters

- [`checkScBoost()`](checkScBoost.md) : Check Boost parameters

- [`checkScFastmnn()`](checkScFastmnn.md) : Check fastMNN parameters

- [`checkScHotspot()`](checkScHotspot.md) : Check HotSpot parameters

- [`checkScHvg()`](checkScHvg.md) : Check HVG selection parameters

- [`checkScMetacells()`](checkScMetacells.md) : Check metacell
  generation parameters

- [`checkScMiloR()`](checkScMiloR.md) : Check MiloR parameters

- [`checkScMinQC()`](checkScMinQC.md) : Check SC minimum QC parameters

- [`checkScMtxIO()`](checkScMtxIO.md) : Check SC MTX load parameters

- [`checkScNeighbours()`](checkScNeighbours.md) : Check neighbour
  generation parameters

- [`checkScScrublet()`](checkScScrublet.md) : Check Scrublet parameters

- [`checkScSeacells()`](checkScSeacells.md) : Check SEACells parameters

- [`checkScSupercell()`](checkScSupercell.md) : Check SuperCell
  parameters

- [`checkScSyntheticData()`](checkScSyntheticData.md) : Check synthetic
  data parameters

- [`checkScVision()`](checkScVision.md) : Check VISION parameters

- [`checkSingleSampleGSEAparams()`](checkSingleSampleGSEAparams.md) :
  Check ssGSEA parameters

- [`check_dim()`](check_dim.md) : Helper to assert same number of
  samples

- [`check_pca_grp_differences()`](check_pca_grp_differences.md) : Helper
  function to check if PC1 and/or PC2 distinguish groups

- [`community_detection()`](community_detection.md) : Identify
  privileged communities based on a given diffusion vector

- [`community_stability()`](community_stability.md) : Assess stability
  of ICA components

- [`component_mutual_information()`](component_mutual_information.md) :
  Assess the NMI of the ICA components

- [`constrained_page_rank()`](constrained_page_rank.md) : Constrained
  personalised page rank

- [`constrained_page_rank_ls()`](constrained_page_rank_ls.md) :
  Constrained personalised page rank over a list

- [`contrastive_pca()`](contrastive_pca.md) : Apply contrastive PCA.

- [`contrastive_pca_processing()`](contrastive_pca_processing.md) :
  Prepare class for contrastive PCA

- [`cor_module_check_epsilon()`](cor_module_check_epsilon.md) : Iterate
  through different epsilon parameters

- [`cor_module_coremo_clustering()`](cor_module_coremo_clustering.md) :
  Generates CoReMo-based gene modules

- [`cor_module_coremo_cor_sign()`](cor_module_coremo_cor_sign.md) :
  Split CoReMo modules by correlation sign

- [`cor_module_coremo_eigengene()`](cor_module_coremo_eigengene.md) :
  Calculate Eigengenes for CoReMo modules

- [`cor_module_coremo_stability()`](cor_module_coremo_stability.md) :
  Assesses CoReMo-based gene module stability

- [`cor_module_graph_check_res()`](cor_module_graph_check_res.md) :
  Iterate through Leiden resolutions for graph-based community
  detection.

- [`cor_module_graph_final_modules()`](cor_module_graph_final_modules.md)
  : Identify correlation-based gene modules via graphs

- [`cor_module_processing()`](cor_module_processing.md) : Prepare
  correlation-based module detection

- [`cor_module_tom()`](cor_module_tom.md) : Update the correlation
  matrix to a TOM

- [`coremo_cluster_quality()`](coremo_cluster_quality.md) : Coremo:
  measures the quality of the clusters

- [`coremo_tree_cut()`](coremo_tree_cut.md) : Coremo: cuts a
  hierarchical cluster based on k

- [`create_dist_obj()`](create_dist_obj.md) : Create distance object
  from a vector

- [`create_sparse_matrix()`](create_sparse_matrix.md) : Helper function
  transform Rust counts into sparse matrices

- [`dgrdl_grid_search()`](dgrdl_grid_search.md) : Grid search over DGRDL
  parameters

- [`dgrdl_result()`](dgrdl_result.md) : Run DGRDL with the specified
  parameters

- [`diffcor_module_processing()`](diffcor_module_processing.md) :
  Prepare differential correlation-based module detection

- [`diffuse_seed_nodes()`](diffuse_seed_nodes.md) : Diffuse seed genes
  over a network

- [`doublet_detection_boost_sc()`](doublet_detection_boost_sc.md) :
  Doublet detection with boosted doublet classification

- [`f1_score_confusion_mat()`](f1_score_confusion_mat.md) : F1 scores on
  top of a confusion matrix

- [`fast_ica_rust()`](fast_ica_rust.md) : Fast ICA via Rust

- [`fast_ica_rust_helper()`](fast_ica_rust_helper.md) : Fast ICA via
  Rust from processed data

- [`fast_mnn_sc()`](fast_mnn_sc.md) : Run fastMNN

- [`fgsea_go_elim()`](fgsea_go_elim.md) : Run GO enrichment with
  elimination method over a continuous vectors

- [`fgsea_simple_go_elim()`](fgsea_simple_go_elim.md) : Run GO
  enrichment with elimination with fgsea simple

- [`filter_similarities()`](filter_similarities.md) : Filter the
  calculated similarities

- [`finalise_matrix()`](finalise_matrix.md) : Finalise the count matrix

- [`find_all_markers_sc()`](find_all_markers_sc.md) : Find all markers

- [`find_clusters_sc()`](find_clusters_sc.md) : Graph-based clustering
  of cells on the sNN graph

- [`find_hvg_batch_aware_sc()`](find_hvg_batch_aware_sc.md) : Identify
  HVGs (batch aware)

- [`find_hvg_sc()`](find_hvg_sc.md) : Identify HVGs

- [`find_markers_sc()`](find_markers_sc.md) : Calculate DGE between two
  cell groups

- [`find_neighbours_sc()`](find_neighbours_sc.md) : Find the neighbours
  for single cell.

- [`find_rbh_communities()`](find_rbh_communities.md) : Find RBH
  communities

- [`fix_contrast_names()`](fix_contrast_names.md) : Fixes contrast names
  for DGEs

- [`fix_meta_data_column()`](fix_meta_data_column.md) : Helper to fix
  meta-data columns to be R conform

- [`flip_ica_loading_signs()`](flip_ica_loading_signs.md) : Flips the
  ICA source sign

- [`gene_ontology_data()`](gene_ontology_data.md) : Gene Ontology data

- [`gene_set_proportions_sc()`](gene_set_proportions_sc.md) : Calculate
  the proportions of reads for specific gene sets

- [`generate_gene_module_data()`](generate_gene_module_data.md) :
  Generates synthetic gene module data.

- [`generate_knn_sc()`](generate_knn_sc.md) : Generate the KNN data with
  distances

- [`generate_null_perm_gs()`](generate_null_perm_gs.md) : Generate
  random gene sets for VISION p-value calculations

- [`generate_perm_diffusion_vecs()`](generate_perm_diffusion_vecs.md) :
  Generates random permutation vectors

- [`generate_personalisation_vec()`](generate_personalisation_vec.md) :
  Helper function to create personalisation vectors

- [`generate_rbh_graph()`](generate_rbh_graph.md) : Generate an RBH
  graph.

- [`generate_single_cell_test_data()`](generate_single_cell_test_data.md)
  : Single cell test data

- [`get_available_embeddings()`](get_available_embeddings.md) : Get the
  available embeddings from the cache

- [`get_c_pca_factors()`](get_c_pca_factors.md) : Get the contrastive
  PCA factors

- [`get_c_pca_loadings()`](get_c_pca_loadings.md) : Get the contrastive
  PCA loadings

- [`get_cell_indices()`](get_cell_indices.md) : Get the index position
  for a gene

- [`get_cell_names()`](get_cell_names.md) : Get the cell names

- [`get_cell_ranger_params()`](get_cell_ranger_params.md) : Helper to
  generate cell ranger input parameters

- [`get_cells_to_keep()`](get_cells_to_keep.md) : Get the cells to keep

- [`get_cor_graph()`](get_cor_graph.md) : Get correlation-based graph

- [`get_cores()`](get_cores.md) : Helper function for core detection.

- [`get_counts_from_rust()`](get_counts_from_rust.md) : Helper function
  to get counts from Rust

- [`get_dge_effect_sizes()`](get_dge_effect_sizes.md) : Return the
  effect size results

- [`get_dge_limma_voom()`](get_dge_limma_voom.md) : Return the Limma
  Voom results

- [`get_dge_list()`](get_dge_list.md) : Return the DGEList

- [`get_dge_qc_plot()`](get_dge_qc_plot.md) : Return QC plots

- [`get_diffcor_graph()`](get_diffcor_graph.md) : Get differential
  correlation-based graph

- [`get_differential_abundance_res()`](get_differential_abundance_res.md)
  : Get the differential abundance results

- [`get_diffusion_perms()`](get_diffusion_perms.md) : Get the diffusion
  permutations

- [`get_diffusion_vector()`](get_diffusion_vector.md) : Get the
  diffusion vector

- [`get_embedding()`](get_embedding.md) : Get the embedding from the
  cache

- [`get_epsilon_res()`](get_epsilon_res.md) : Return the epsilon data

- [`get_fpkm_counts()`](get_fpkm_counts.md) : Return the FPKM-normalised
  counts

- [`get_gene_indices()`](get_gene_indices.md) : Get the index position
  for a gene

- [`get_gene_lengths()`](get_gene_lengths.md) : Get the gene lengths

- [`get_gene_lengths.bulk_coexp`](get_gene_lengths.bulk_coexp.md) : Get
  gene set lengths for a bulk_coexp class

- [`get_gene_lengths(`*`<matrix>`*`)`](get_gene_lengths.matrix.md) : Get
  gene set lengths for a matrix

- [`get_gene_names()`](get_gene_names.md) : Get the gene names

- [`get_gene_names_from_idx()`](get_gene_names_from_idx.md) : Get the
  gene names based on the gene idx

- [`get_go_data_human()`](get_go_data_human.md) : Wrapper function to
  load and process the gene ontology data.

- [`get_go_levels()`](get_go_levels.md) : Helper to get the Gene
  Ontology levels

- [`get_grid_search_res()`](get_grid_search_res.md) : Get the grid
  search results

- [`get_h5ad_dimensions()`](get_h5ad_dimensions.md) : Helper function to
  get the dimensions and compressed sparse format

- [`get_hotspot_membership()`](get_hotspot_membership.md) : Get the
  hotspot gene membership table

- [`get_hvg()`](get_hvg.md) : Get the HVG

- [`get_ica_stability_res()`](get_ica_stability_res.md) : Get the ICA
  component data (stability, convergence, nMI)

- [`get_index_cells()`](get_index_cells.md) : Get the index cells

- [`get_inflection_point()`](get_inflection_point.md) : Identify the
  inflection point for elbow-like data

- [`get_knn_dist()`](get_knn_dist.md) : Get the KNN distance measures

- [`get_knn_mat()`](get_knn_mat.md) : Get the KNN matrix

- [`get_meta_cell_matrices()`](get_meta_cell_matrices.md) : Create
  sparse dgRMatrix matrices for raw and norm counts

- [`get_meta_cells_sc()`](get_meta_cells_sc.md) :

  Generate meta cells based on hdWGCNA and return a `meta_cells` object

- [`get_metadata()`](get_metadata.md) : Return the metadata

- [`get_miloR_abundances_sc()`](get_miloR_abundances_sc.md) : Generate
  an miloR abundance object for differential abundance testing

- [`get_model_fit()`](get_model_fit.md) : Get the fitted model

- [`get_obs_data()`](get_obs_data.md) : Get the ready obs data from
  various sub method

- [`get_ontology_ancestry()`](get_ontology_ancestry.md) : Return
  ancestry terms from an ontology

- [`get_outputs()`](get_outputs.md) : Return the outputs

- [`get_params()`](get_params.md)
  [`get_params.sc_hotspot()`](get_params.md)
  [`get_params.sc_miloR()`](get_params.md) : Get the parameters that
  were used.

- [`get_pca_factors()`](get_pca_factors.md) : Get the PCA factors

- [`get_pca_loadings()`](get_pca_loadings.md) : Get the PCA loadings

- [`get_pca_singular_val()`](get_pca_singular_val.md) : Get the PCA
  singular values

- [`get_rbh_res()`](get_rbh_res.md) : Get the RBH results

- [`get_resolution_res()`](get_resolution_res.md) : Return the
  resolution results

- [`get_results()`](get_results.md) : Get the final results from the
  class

- [`get_rust_count_cell_f_path()`](get_rust_count_cell_f_path.md) : Get
  the file path to the counts in cell format

- [`get_rust_count_gene_f_path()`](get_rust_count_gene_f_path.md) : Get
  the file path to the counts in gene format

- [`get_sc_cache()`](get_sc_cache.md) : Getter the memory-stored data
  from the class

- [`get_sc_counts()`](get_sc_counts.md) : Getter the counts

- [`get_sc_duckdb()`](get_sc_duckdb.md) : Getter for the single cell
  DuckDB connection

- [`get_sc_map()`](get_sc_map.md) : Getter for the different maps in the
  object

- [`get_sc_obs()`](get_sc_obs.md) : Getter the obs table

- [`get_sc_rust_ptr()`](get_sc_rust_ptr.md) : Getter for the single cell
  Rust pointer

- [`get_sc_var()`](get_sc_var.md) : Getter the var table

- [`get_seacells_sc()`](get_seacells_sc.md) :

  Generate meta cells based on SEACells and return a `meta_cells` object

- [`get_seurat_counts_to_list()`](get_seurat_counts_to_list.md) :
  Transform Seurat raw counts into a List

- [`get_sim_matrix()`](get_sim_matrix.md) : Get the similarity matrix

- [`get_snf_adjcacency_mat()`](get_snf_adjcacency_mat.md) : Get an
  individual affinity matrix

- [`get_snf_final_mat()`](get_snf_final_mat.md) : Get the final SNF
  matrix

- [`get_snf_params()`](get_snf_params.md) : Get the SNF params

- [`get_snn_graph()`](get_snn_graph.md) : Get the sNN graph

- [`get_supercells_sc()`](get_supercells_sc.md) :

  Generate SuperCells and return a `meta_cells` object

- [`get_tpm_counts()`](get_tpm_counts.md) : Return the TPM-normalised
  counts

- [`gse_go_elim_method()`](gse_go_elim_method.md) : Run gene ontology
  enrichment with elimination method.

- [`gse_go_elim_method_list()`](gse_go_elim_method_list.md) : Run gene
  ontology enrichment with elimination method over a list.

- [`gse_hypergeometric()`](gse_hypergeometric.md) : Gene set enrichment
  (GSE) based on a hypergeometric test.

- [`gse_hypergeometric_list()`](gse_hypergeometric_list.md) : Gene set
  enrichment (GSE) based on a hypergeometric test over a list.

- [`hedges_g_dge()`](hedges_g_dge.md) : Calculate the effect size

- [`hotspot_autocor_sc()`](hotspot_autocor_sc.md) : Calculate the local
  auto-correlation of a gene

- [`hotspot_gene_cor_sc()`](hotspot_gene_cor_sc.md) : Calculate the
  local pairwise gene-gene correlation

- [`ica_evaluate_comp()`](ica_evaluate_comp.md) : Iterate over different
  ncomp parameters for ICA

- [`ica_optimal_ncomp()`](ica_optimal_ncomp.md) : Identify stability
  inflection point

- [`ica_processing()`](ica_processing.md) : Prepare class for ICA

- [`ica_stabilised_results()`](ica_stabilised_results.md) : Run
  stabilised ICA with a given number of components

- [`knn_graph_label_propagation()`](knn_graph_label_propagation.md) :
  kNN-based graph label propagation

- [`load_existing()`](load_existing.md) : Load an existing
  single_cell_exp from disk

- [`load_go_human_data()`](load_go_human_data.md) : Get the Gene
  Ontology data human

- [`load_h5ad()`](load_h5ad.md) :

  Load in h5ad to `single_cell_exp`

- [`load_mtx()`](load_mtx.md) :

  Load in mtx/plain text files to `single_cell_exp` (nightly!)

- [`load_r_data()`](load_r_data.md) : Load in data directly from R
  objects.

- [`load_seurat()`](load_seurat.md) :

  Load in Seurat to `single_cell_exp`

- [`meta_cells()`](meta_cells.md) : bixverse meta cell class

- [`module_scores_sc()`](module_scores_sc.md) : Calculate module
  activity scores

- [`multilevel_error()`](multilevel_error.md) : Helper function to
  calculate the multi-level error

- [`network_diffusions()`](network_diffusions.md) : Network diffusion
  class

- [`new_sc_cache()`](new_sc_cache.md) : Helper function to hold relevant
  cached data

- [`new_sc_hotspot_res()`](new_sc_hotspot_res.md) : Helper function to
  generate HotSpot data

- [`new_sc_knn()`](new_sc_knn.md) : Helper function to generate kNN data
  with distances

- [`new_sc_mapper()`](new_sc_mapper.md) : Helper function to construct
  relevant maps

- [`new_sc_miloR_res()`](new_sc_miloR_res.md) : Generate a new miloR
  result

- [`nightly_feature()`](nightly_feature.md) : Helper functions to mark
  something unstable/nightly

- [`normalise_bulk_dge()`](normalise_bulk_dge.md) : Normalise the count
  data for DGE.

- [`one_hot_encode()`](one_hot_encode.md) : Helper to generate one-hot
  encodings

- [`ontology()`](ontology.md) : Ontology class

- [`ot_harmonic_score()`](ot_harmonic_score.md) : Calculates a harmonic
  sum normalised between 0 to 1.

- [`params_boost()`](params_boost.md) : Wrapper function for Boost
  parameters

- [`params_cistarget()`](params_cistarget.md) : Wrapper function to
  CisTarget parameters

- [`params_community_detection()`](params_community_detection.md) :
  Wrapper function to generate community detection parameters

- [`params_cor_graph()`](params_cor_graph.md) : Wrapper function for
  graph generation

- [`params_coremo()`](params_coremo.md) : Wrapper function to generate
  CoReMo parameters

- [`params_dgrdl()`](params_dgrdl.md) : Wrapper function to generate
  DGRDL parameters

- [`params_graph_resolution()`](params_graph_resolution.md) : Wrapper
  function to generate resolution parameters for Leiden or Louvain
  clustering.

- [`params_gsea()`](params_gsea.md) : Wrapper function to generate GSEA
  parameters

- [`params_gsva()`](params_gsva.md) : Wrapper function to generate GSVA
  parameters

- [`params_hvg_defaults()`](params_hvg_defaults.md) : Helper function to
  generate HVG defaults

- [`params_ica_general()`](params_ica_general.md) : Wrapper function for
  standard ICA parameters

- [`params_ica_ncomp()`](params_ica_ncomp.md) : Wrapper function for ICA
  ncomp iterations

- [`params_ica_randomisation()`](params_ica_randomisation.md) : Wrapper
  function for ICA randomisation

- [`params_knn_defaults()`](params_knn_defaults.md) : Helper function to
  generate kNN defaults

- [`params_norm_doublet_detection_defaults()`](params_norm_doublet_detection_defaults.md)
  : Helper function to generate normalisation defaults for doublet
  detection.

- [`params_pca_defaults()`](params_pca_defaults.md) : Helper function to
  generate default parameters for PCA

- [`params_sc_bbknn()`](params_sc_bbknn.md) : Wrapper function for the
  BBKNN parameters

- [`params_sc_fastmnn()`](params_sc_fastmnn.md) : Wrapper function for
  the fastMNN parameters

- [`params_sc_hotspot()`](params_sc_hotspot.md) : Wrapper function for
  parameters for HotSpot

- [`params_sc_hvg()`](params_sc_hvg.md) : Wrapper function for HVG
  detection parameters.

- [`params_sc_metacells()`](params_sc_metacells.md) : Wrapper function
  for parameters for meta cell generation

- [`params_sc_miloR()`](params_sc_miloR.md) : Wrapper function for
  parameters for MiloR

- [`params_sc_min_quality()`](params_sc_min_quality.md) : Wrapper
  function to generate QC metric params for single cell

- [`params_sc_mtx_io()`](params_sc_mtx_io.md) : Wrapper function to
  provide data for mtx-based loading

- [`params_sc_neighbours()`](params_sc_neighbours.md) : Wrapper function
  for parameters for neighbour identification in single cell

- [`params_sc_seacells()`](params_sc_seacells.md) : Wrapper function for
  the SEACells parameters

- [`params_sc_supercell()`](params_sc_supercell.md) : Wrapper function
  for parameters for SuperCell generation

- [`params_sc_synthetic_data()`](params_sc_synthetic_data.md) : Default
  parameters for generation of synthetic data

- [`params_sc_vision()`](params_sc_vision.md) : Wrapper function for
  parameters for VISION with auto-correlation

- [`params_scrublet()`](params_scrublet.md) : Wrapper function for
  Scrublet doublet detection parameters

- [`params_snf()`](params_snf.md) : Wrapper function to generate SNF
  parameters

- [`params_ssgsea()`](params_ssgsea.md) : Wrapper function to generate
  ssGSEA parameters

- [`permute_seed_nodes()`](permute_seed_nodes.md) : Generate permuation
  scores for the diffusion

- [`plot(`*`<cpca_synthetic_data>`*`)`](plot.cpca_synthetic_data.md) :
  Plot the contrastive PCA example data

- [`plot(`*`<synthetic_matrix_simple>`*`)`](plot.synthetic_matrix_simple.md)
  : Plot the (simple) synthetic gene expression

- [`plot_boxplot_normalization()`](plot_boxplot_normalization.md) :
  Helper plot function for boxplot of normalized data

- [`plot_epsilon_res()`](plot_epsilon_res.md) : Plot the epsilon vs.
  power law goodness of fit result

- [`plot_hvgs()`](plot_hvgs.md) : Plot the highly variable genes

- [`plot_ica_ncomp_params()`](plot_ica_ncomp_params.md) : Plot various
  parameters with no comp

- [`plot_ica_stability_individual()`](plot_ica_stability_individual.md)
  : Plot the stability of the ICA components

- [`plot_optimal_cuts()`](plot_optimal_cuts.md) : Plot the k cuts vs
  median R2

- [`plot_pca()`](plot_pca.md) : Helper plot function for pca with
  contrasts

- [`plot_pca_res()`](plot_pca_res.md) : Plot the PCA data

- [`plot_preprocessing_genes()`](plot_preprocessing_genes.md) : Helper
  plot function of distribution of genes by samples

- [`plot_preprocessing_outliers()`](plot_preprocessing_outliers.md) :
  Helper plot function for identification of outliers

- [`plot_rbf_impact()`](plot_rbf_impact.md) : Helper function to plot
  distance to affinity relationship

- [`plot_resolution_res()`](plot_resolution_res.md) : Plot the
  resolution results.

- [`plot_voom_normalization()`](plot_voom_normalization.md) : Helper
  plot function for Voom normalisation

- [`pre_process_sim_onto()`](pre_process_sim_onto.md) : Pre-process data
  for subsequent ontology similarity

- [`prep_data_gower_hamming_dist()`](prep_data_gower_hamming_dist.md) :
  Transform data.tables into matrices for distance calculations

- [`prep_limma_contrasts()`](prep_limma_contrasts.md) : Create all limma
  contrasts from a provided string

- [`prep_stats_pathways()`](prep_stats_pathways.md) : Helper function to
  prepare data for GSEA

- [`preprocess_bulk_coexp()`](preprocess_bulk_coexp.md) : Process the
  raw data

- [`preprocess_bulk_dge()`](preprocess_bulk_dge.md) : QC on the bulk dge
  data (DEPRECATED!)

- [`print.bulk_coexp`](print.bulk_coexp.md) : print Method for
  bulk_coexp object

- [`print.gene_ontology_data`](print.gene_ontology_data.md) : print
  Method for gene_ontology_data object

- [`print.ontology`](print.ontology.md) : print Method for ontology
  object

- [`process_cistarget_res()`](process_cistarget_res.md) : Helper to
  process CisTarget results

- [`process_go_data()`](process_go_data.md) : Process Gene Ontology data
  into the right format

- [`qc_bulk_dge()`](qc_bulk_dge.md) : QC on the bulk dge data

- [`rbh_graph()`](rbh_graph.md) : Reciprocal best hit graph

- [`read_motif_annotation_file()`](read_motif_annotation_file.md) : Read
  in the motif annotation file

- [`read_motif_ranking()`](read_motif_ranking.md) : Read in the motif
  rankings and transform them into a matrix

- [`remove_knn()`](remove_knn.md) : Remove the KNN data

- [`remove_samples()`](remove_samples.md) : Remove samples from object

- [`remove_snn_graph()`](remove_snn_graph.md) : Remove the sNN graph

- [`robust_scale()`](robust_scale.md) : Robust scaler.

- [`rs_2d_loess()`](rs_2d_loess.md) : Rust implementation of a Loess
  function

- [`rs_aucell()`](rs_aucell.md) : Calculate AUCell in Rust

- [`rs_bbknn()`](rs_bbknn.md) : BBKNN implementation in Rust

- [`rs_bbknn_filtering()`](rs_bbknn_filtering.md) : Reduce BBKNN matrix
  to Top X neighbours

- [`rs_calc_es()`](rs_calc_es.md) : Calculates the traditional GSEA
  enrichment score

- [`rs_calc_gsea_stat_cumulative_batch()`](rs_calc_gsea_stat_cumulative_batch.md)
  : Helper function to generate fgsea simple-based permutations

- [`rs_calc_gsea_stat_traditional_batch()`](rs_calc_gsea_stat_traditional_batch.md)
  : Helper function to generate traditional GSEA-based permutations

- [`rs_calc_gsea_stats()`](rs_calc_gsea_stats.md) : Rust implementation
  of the fgsea::calcGseaStat() function

- [`rs_calc_multi_level()`](rs_calc_multi_level.md) : Calculates
  p-values for pre-processed data

- [`rs_calculate_dge_mann_whitney()`](rs_calculate_dge_mann_whitney.md)
  : Calculate DGEs between cells based on Mann Whitney stats

- [`rs_cistarget()`](rs_cistarget.md) : Run CisTarget motif enrichment
  analysis

- [`rs_cluster_stability()`](rs_cluster_stability.md) : Helper function
  to assess cluster stability

- [`rs_constrained_page_rank()`](rs_constrained_page_rank.md) :
  Calculate a constrained page-rank score

- [`rs_constrained_page_rank_list()`](rs_constrained_page_rank_list.md)
  : Calculate a constrained page-rank score over a list.

- [`rs_contrastive_pca()`](rs_contrastive_pca.md) : Calculate the
  contrastive PCA

- [`rs_cor()`](rs_cor.md) : Calculate the column wise correlations.

- [`rs_cor2()`](rs_cor2.md) : Calculate the column wise correlations.

- [`rs_cor_upper_triangle()`](rs_cor_upper_triangle.md) : Calculate the
  column wise correlations.

- [`rs_coremo_quality()`](rs_coremo_quality.md) : Helper function to
  assess CoReMo cluster quality

- [`rs_coremo_stability()`](rs_coremo_stability.md) : Helper function to
  assess CoReMo cluster stability

- [`rs_cos()`](rs_cos.md) : Calculate the column wise cosine
  similarities

- [`rs_count_zeroes()`](rs_count_zeroes.md) : Helper to get zero stats
  from a given matrix

- [`rs_cov2cor()`](rs_cov2cor.md) : Calculates the correlation matrix
  from the co-variance matrix

- [`rs_covariance()`](rs_covariance.md) : Calculate the column-wise
  co-variance.

- [`rs_create_random_aucs()`](rs_create_random_aucs.md) : Create random
  AUCs

- [`rs_critval()`](rs_critval.md) : Calculate the critical value

- [`rs_critval_mat()`](rs_critval_mat.md) : Calculate the critical value

- [`rs_dense_to_upper_triangle()`](rs_dense_to_upper_triangle.md) :
  Generate a vector-based representation of the upper triangle of a
  matrix

- [`rs_differential_cor()`](rs_differential_cor.md) : Calculate the
  column wise differential correlation between two sets of data.

- [`rs_dist()`](rs_dist.md) : Calculate the pairwise column distance in
  a matrix

- [`rs_fast_auc()`](rs_fast_auc.md) : Fast AUC calculation

- [`rs_fast_ica()`](rs_fast_ica.md) : Run the Rust implementation of
  fast ICA.

- [`rs_fdr_adjustment()`](rs_fdr_adjustment.md) : Calculate a BH-based
  FDR

- [`rs_filter_onto_sim()`](rs_filter_onto_sim.md) : Filter the term
  similarities for a specific critical value

- [`rs_generate_bulk_rnaseq()`](rs_generate_bulk_rnaseq.md) : Generation
  of bulkRNAseq-like data with optional correlation structure

- [`rs_geom_elim_fgsea_simple()`](rs_geom_elim_fgsea_simple.md) : Run
  fgsea simple method for gene ontology with elimination method

- [`rs_get_gs_indices()`](rs_get_gs_indices.md) : Helper function to
  rapidly retrieve the indices of the gene set members

- [`rs_get_metacells()`](rs_get_metacells.md) : Generate meta cells
  (hdWGCNA method)

- [`rs_get_seacells()`](rs_get_seacells.md) : Generate SEACells

- [`rs_gower_dist()`](rs_gower_dist.md) : Calculates the Gower distance
  for a given matrix

- [`rs_gse_geom_elim()`](rs_gse_geom_elim.md) : Run hypergeometric
  enrichment over the gene ontology

- [`rs_gse_geom_elim_list()`](rs_gse_geom_elim_list.md) : Run
  hypergeometric enrichment a list of target genes over the gene
  ontology

- [`rs_gsva()`](rs_gsva.md) : Rust version of the GSVA algorithm

- [`rs_h5ad_data()`](rs_h5ad_data.md) : Load in h5ad data via Rust

- [`rs_hamming_dist()`](rs_hamming_dist.md) : Calculates the Hamming
  distance between categorical columns

- [`rs_hedges_g()`](rs_hedges_g.md) : Calculate the Hedge's G effect

- [`rs_hotspot_autocor()`](rs_hotspot_autocor.md) : Calculate gene
  spatial auto-correlations

- [`rs_hotspot_cluster_genes()`](rs_hotspot_cluster_genes.md) : Cluster
  the genes by Z-score together

- [`rs_hotspot_gene_cor()`](rs_hotspot_gene_cor.md) : Calculate
  gene\<\>gene spatial correlations

- [`rs_hypergeom_test()`](rs_hypergeom_test.md) : Run a single
  hypergeometric test.

- [`rs_hypergeom_test_list()`](rs_hypergeom_test_list.md) : Run a
  hypergeometric test over a list of target genes

- [`rs_ica_iters()`](rs_ica_iters.md) : Run ICA over a given no_comp
  with random initilisations of w_init

- [`rs_ica_iters_cv()`](rs_ica_iters_cv.md) : Run ICA with
  cross-validation and random initialsiation

- [`rs_kbet()`](rs_kbet.md) : Calculate kBET type scores

- [`rs_knn_label_propagation()`](rs_knn_label_propagation.md) : kNN
  label propagation

- [`rs_knn_mat_to_edge_list()`](rs_knn_mat_to_edge_list.md) : Flatten
  kNN matrix to edge list

- [`rs_mad_outlier()`](rs_mad_outlier.md) : Calculate MAD outlier
  detection in Rust.

- [`rs_make_milor_nhoods()`](rs_make_milor_nhoods.md) : Generate the
  neighbourhoods akin to the miloR approach

- [`rs_mitch_calc()`](rs_mitch_calc.md) : Calculate mitch enrichment
  leveraging Rust under the hood

- [`rs_mnn()`](rs_mnn.md) : FastMNN batch correction in Rust

- [`rs_module_scoring()`](rs_module_scoring.md) : Calculate module
  activity scores in Rust

- [`rs_mutual_info()`](rs_mutual_info.md) : Calculates the mutual
  information matrix

- [`rs_onto_semantic_sim()`](rs_onto_semantic_sim.md) : Calculate the
  semantic similarity in an ontology

- [`rs_onto_semantic_sim_mat()`](rs_onto_semantic_sim_mat.md) :
  Calculate the semantic similarity in an ontology

- [`rs_onto_sim_wang()`](rs_onto_sim_wang.md) : Calculate the Wang
  similarity for specific terms

- [`rs_onto_sim_wang_mat()`](rs_onto_sim_wang_mat.md) : Calculate the
  Wang similarity matrix for an ontology

- [`rs_ot_harmonic_sum()`](rs_ot_harmonic_sum.md) : Calculate the OT
  harmonic sum

- [`rs_page_rank()`](rs_page_rank.md) : Rust version of calcaluting the
  personalised page rank

- [`rs_page_rank_parallel()`](rs_page_rank_parallel.md) : Calculate
  massively parallelised personalised page rank scores

- [`rs_phyper()`](rs_phyper.md) : Calculate the hypergeometric rest in
  Rust

- [`rs_pointwise_mutual_info()`](rs_pointwise_mutual_info.md) :
  Calculates the point wise mutual information

- [`rs_prcomp()`](rs_prcomp.md) : Rust implementation of prcomp

- [`rs_prepare_gsva_gs()`](rs_prepare_gsva_gs.md) : Prepare a pathway
  list for GSVA

- [`rs_prepare_whitening()`](rs_prepare_whitening.md) : Prepare the data
  for whitening

- [`rs_random_svd()`](rs_random_svd.md) : Run randomised SVD over a
  matrix

- [`rs_range_norm()`](rs_range_norm.md) : Apply a range normalisation on
  a vector.

- [`rs_rbf_function()`](rs_rbf_function.md) : Apply a Radial Basis
  Function

- [`rs_rbf_function_mat()`](rs_rbf_function_mat.md) : Apply a Radial
  Basis Function (to a matrix)

- [`rs_rbf_iterate_epsilons()`](rs_rbf_iterate_epsilons.md) : Helper to
  identify the right epsilon parameter

- [`rs_rbh_cor()`](rs_rbh_cor.md) : Generate reciprocal best hits based
  on correlations

- [`rs_rbh_sets()`](rs_rbh_sets.md) : Generate reciprocal best hits
  based on set similarities

- [`rs_sample_ids_for_cell_types()`](rs_sample_ids_for_cell_types.md) :
  Helper function to generate sample identifiers based on cells

- [`rs_sc_doublet_detection()`](rs_sc_doublet_detection.md) : Detect
  Doublets via BoostClassifier (in Rust)

- [`rs_sc_get_gene_set_perc()`](rs_sc_get_gene_set_perc.md) : Calculate
  the percentage of gene sets in the cells

- [`rs_sc_get_top_genes_perc()`](rs_sc_get_top_genes_perc.md) :
  Calculates the cumulative proportion of the top X genes

- [`rs_sc_hvg()`](rs_sc_hvg.md) : Calculate the percentage of gene sets
  in the cells

- [`rs_sc_hvg_batch_aware()`](rs_sc_hvg_batch_aware.md) : Calculate HVG
  per batch

- [`rs_sc_knn()`](rs_sc_knn.md) : Generates the kNN graph

- [`rs_sc_knn_w_dist()`](rs_sc_knn_w_dist.md) : Generates the kNN graph
  with additional distances

- [`rs_sc_pca()`](rs_sc_pca.md) : Calculates PCA for single cell

- [`rs_sc_scrublet()`](rs_sc_scrublet.md) : Scrublet Rust interface

- [`rs_sc_snn()`](rs_sc_snn.md) : Generates the sNN graph for igraph

- [`rs_set_similarity()`](rs_set_similarity.md) : Set similarities

- [`rs_set_similarity_list()`](rs_set_similarity_list.md) : Set
  similarities over one list

- [`rs_set_similarity_list2()`](rs_set_similarity_list2.md) : Set
  similarities over two list

- [`rs_simple_and_multi_err()`](rs_simple_and_multi_err.md) : Calculates
  the simple and multi error for fgsea multi level

- [`rs_simulate_dropouts()`](rs_simulate_dropouts.md) : Sparsify
  bulkRNAseq like data

- [`rs_snf()`](rs_snf.md) : Similarity network fusion

- [`rs_snf_affinity_cat()`](rs_snf_affinity_cat.md) : Calculate the SNF
  affinity matrix for categorical values

- [`rs_snf_affinity_continuous()`](rs_snf_affinity_continuous.md) :
  Calculate the SNF affinity matrix for continuous values

- [`rs_snf_affinity_mixed()`](rs_snf_affinity_mixed.md) : Calculate the
  SNF affinity matrix for mixed values

- [`rs_sparse_dict_dgrdl()`](rs_sparse_dict_dgrdl.md) : Generate a
  sparse dictionary with DGRDL

- [`rs_sparse_dict_dgrdl_grid_search()`](rs_sparse_dict_dgrdl_grid_search.md)
  : Generate a sparse dictionary with DGRDL

- [`rs_spectral_clustering()`](rs_spectral_clustering.md) : Rust
  implementation of spectral clustering

- [`rs_spectral_clustering_sim()`](rs_spectral_clustering_sim.md) : Rust
  implementation of spectral clustering

- [`rs_split_cor_signs()`](rs_split_cor_signs.md) : Helper function to
  split correlation matrices by sign

- [`rs_ssgsea()`](rs_ssgsea.md) : Rust version of the ssGSEA algorithm

- [`rs_supercell()`](rs_supercell.md) : Generate SuperCells.

- [`rs_synthetic_sc_data_csc()`](rs_synthetic_sc_data_csc.md) : Generate
  synthetic single cell data (Seurat type)

- [`rs_synthetic_sc_data_csr()`](rs_synthetic_sc_data_csr.md) : Generate
  synthetic single cell data (h5ad type)

- [`rs_synthetic_sc_data_with_cell_types()`](rs_synthetic_sc_data_with_cell_types.md)
  : Generates synthetic data for single cell

- [`rs_tied_diffusion_parallel()`](rs_tied_diffusion_parallel.md) :
  Calculate massively parallelised tied diffusion scores

- [`rs_tom()`](rs_tom.md) : Calculates the TOM over an affinity matrix

- [`rs_upper_triangle_to_dense()`](rs_upper_triangle_to_dense.md) :
  Reconstruct a matrix from a flattened upper triangle vector

- [`rs_upper_triangle_to_sparse()`](rs_upper_triangle_to_sparse.md) :
  Generate sparse data from an upper triangle

- [`rs_vision()`](rs_vision.md) : Calculate VISION pathway scores in
  Rust

- [`rs_vision_with_autocorrelation()`](rs_vision_with_autocorrelation.md)
  : Calculate VISION pathway scores in Rust with auto-correlation

- [`run_cistarget()`](run_cistarget.md) : Main function to run CisTarget

- [`run_limma_voom()`](run_limma_voom.md) : Wrapper for a Limma Voom
  analysis

- [`run_snf()`](run_snf.md) : Run the SNF algorithm

- [`save_sc_exp_to_disk()`](save_sc_exp_to_disk.md) : Save memory-bound
  data to disk

- [`scale_free_fit()`](scale_free_fit.md) : Calculate the goodness of
  fit for a power law distribution.

- [`scrublet_sc()`](scrublet_sc.md) : Doublet detection with Scrublet

- [`select_user_option()`](select_user_option.md) : Helper function for
  user option selection

- [`set_cell_mapping()`](set_cell_mapping.md) : Set cell mapping for
  sc_mapper object

- [`set_cells_to_keep()`](set_cells_to_keep.md) : Set cells to keep for
  sc_mapper object

- [`set_embedding()`](set_embedding.md) : Add additional embeddings to
  the class

- [`set_gene_mapping()`](set_gene_mapping.md) : Set gene mapping for
  sc_mapper object

- [`set_hotspot_membership()`](set_hotspot_membership.md) : Identify
  hotspot gene clusters

- [`set_hvg()`](set_hvg.md) : Set the HVG genes

- [`set_knn()`](set_knn.md) : Set/add KNN

- [`set_pca_factors()`](set_pca_factors.md) : Set/add PCA factors

- [`set_pca_loadings()`](set_pca_loadings.md) : Set/add PCA loadings

- [`set_pca_singular_vals()`](set_pca_singular_vals.md) : Set/add PCA
  singular values

- [`set_sc_new_obs_col()`](set_sc_new_obs_col.md) : Add a new column to
  the obs table

- [`set_sc_new_obs_col_multiple()`](set_sc_new_obs_col_multiple.md) :
  Add multiple new columns to the obs table

- [`set_sc_new_var_cols()`](set_sc_new_var_cols.md) : Add a new column
  to the var table

- [`set_snn_graph()`](set_snn_graph.md) : Set/add KNN

- [`simplify_hypergeom_res()`](simplify_hypergeom_res.md) : Simplify
  gene set results via ontologies

- [`simulate_dropouts()`](simulate_dropouts.md) : Simulate dropouts via
  different functions on synthetic bulk data

- [`single_cell_duckdb_base`](single_cell_duckdb_base.md) : Base class
  for the single cell DuckDB connection

- [`single_cell_duckdb_con`](single_cell_duckdb_con.md) : Class for
  storing single cell experimental data in DuckDB (nightly!)

- [`single_cell_exp()`](single_cell_exp.md) : bixverse single cell class

- [`snf()`](snf.md) : Similarity network fusion

- [`snf_process_aff_cat_mixed()`](snf_process_aff_cat_mixed.md) : Helper
  function to process categorical or mixed data for SNF

- [`snf_process_aff_continuous()`](snf_process_aff_continuous.md) :
  Helper function to process continuous data for SNF

- [`sparse_list_to_mat()`](sparse_list_to_mat.md) : Helper function to
  transform the Rust-exported sparse matrices into R ones

- [`sparse_mat_to_list()`](sparse_mat_to_list.md) : Helper function to
  transform R sparse matrices to list

- [`stream_h5ad()`](stream_h5ad.md) :

  Stream in h5ad to `single_cell_exp`

- [`summarise_scores()`](summarise_scores.md) : Summarise gene scores if
  they are duplicates.

- [`synthetic_bulk_cor_matrix()`](synthetic_bulk_cor_matrix.md) :
  Generates synthetic bulk RNAseq data

- [`synthetic_c_pca_data()`](synthetic_c_pca_data.md) : Generates
  synthetic data for contrastive PCA exploration.

- [`synthetic_signal_matrix()`](synthetic_signal_matrix.md) : Generates
  a simple synthetic, pseudo gene expression matrix

- [`test_nhoods()`](test_nhoods.md) : Test neighbourhoods for
  differential abundance

- [`tied_diffusion()`](tied_diffusion.md) : Diffuse seed genes in a tied
  manner over a network

- [`to_snake_case()`](to_snake_case.md) : Helper function to transform
  strings to snake_case

- [`top_genes_perc_sc()`](top_genes_perc_sc.md) : Calculate the
  proportions of reads for the Top N genes

- [`tree_cut_iter()`](tree_cut_iter.md) : Coremo: Iterate over k for
  gene module detection.

- [`update_metadata_values()`](update_metadata_values.md) : Replace
  values in a metadata column

- [`upper_triangle_diffcor_mat`](upper_triangle_diffcor_mat.md) : Class
  for symmetric differential correlation matrices

- [`upper_triangle_to_sparse()`](upper_triangle_to_sparse.md) :
  Transform an upper triangle-stored matrix to a sparse one

- [`upper_triangular_sym_mat`](upper_triangular_sym_mat.md) : Class for
  symmetric correlation matrices

- [`vision_sc()`](vision_sc.md) : Calculate VISION scores

- [`vision_w_autocor_sc()`](vision_w_autocor_sc.md) : Calculate VISION
  scores (with auto-correlation scores)

- [`write_cellranger_output()`](write_cellranger_output.md) : Helper
  function to write data to a cell ranger like output

- [`write_h5ad_sc()`](write_h5ad_sc.md) : Helper function to write data
  to h5ad format
