<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ bixverse</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">bixverse</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="add_new_metadata.html">add_new_metadata()</a></code>

        </dt>
        <dd>Replace the meta data</dd>

        <dt>

          <code><a href="add_nhoods_info.html">add_nhoods_info()</a></code>

        </dt>
        <dd>Add neighbourhood info on majority cell type</dd>

        <dt>

          <code><a href="add_snf_data_modality.html">add_snf_data_modality()</a></code>

        </dt>
        <dd>Add a data modality for SNF generation</dd>

        <dt>

          <code><a href="all_limma_contrasts.html">all_limma_contrasts()</a></code>

        </dt>
        <dd>Create all limma contrasts (based on combination of everything)</dd>

        <dt>

          <code><a href="anndata_parser.html">anndata_parser</a></code>

        </dt>
        <dd>Class for Anndata</dd>

        <dt>

          <code><a href="assertCellsExist.html">assertCellsExist()</a></code>

        </dt>
        <dd>Assert neighbour generation parameters</dd>

        <dt>

          <code><a href="assertCistargetParams.html">assertCistargetParams()</a></code>

        </dt>
        <dd>Assert CisTarget parameters</dd>

        <dt>

          <code><a href="assertCoReMoParams.html">assertCoReMoParams()</a></code>

        </dt>
        <dd>Assert CoReMo parameter</dd>

        <dt>

          <code><a href="assertCommunityParams.html">assertCommunityParams()</a></code>

        </dt>
        <dd>Assert community detection parameter</dd>

        <dt>

          <code><a href="assertCorGraphParams.html">assertCorGraphParams()</a></code>

        </dt>
        <dd>Assert correlation graph parameters</dd>

        <dt>

          <code><a href="assertDGRDLparams.html">assertDGRDLparams()</a></code>

        </dt>
        <dd>Assert DGRDL parameter</dd>

        <dt>

          <code><a href="assertFileExists.html">assertFileExists()</a></code>

        </dt>
        <dd>Assert that files exist</dd>

        <dt>

          <code><a href="assertGSEAParams.html">assertGSEAParams()</a></code>

        </dt>
        <dd>Assert GSEA parameter</dd>

        <dt>

          <code><a href="assertGSVAParams.html">assertGSVAParams()</a></code>

        </dt>
        <dd>Assert GSVA parameter</dd>

        <dt>

          <code><a href="assertGraphResParams.html">assertGraphResParams()</a></code>

        </dt>
        <dd>Assert resolution graph parameters</dd>

        <dt>

          <code><a href="assertIcaIterParams.html">assertIcaIterParams()</a></code>

        </dt>
        <dd>Assert ICA randomisation parameters</dd>

        <dt>

          <code><a href="assertIcaNcomps.html">assertIcaNcomps()</a></code>

        </dt>
        <dd>Assert ICA no of component parameters</dd>

        <dt>

          <code><a href="assertIcaParams.html">assertIcaParams()</a></code>

        </dt>
        <dd>Assert ICA parameters</dd>

        <dt>

          <code><a href="assertSNFParams.html">assertSNFParams()</a></code>

        </dt>
        <dd>Assert SNF parameter</dd>

        <dt>

          <code><a href="assertScBbknn.html">assertScBbknn()</a></code>

        </dt>
        <dd>Assert BBKNN parameters</dd>

        <dt>

          <code><a href="assertScBoost.html">assertScBoost()</a></code>

        </dt>
        <dd>Assert Boost parameters</dd>

        <dt>

          <code><a href="assertScFastmnn.html">assertScFastmnn()</a></code>

        </dt>
        <dd>Assert fastMNN parameters</dd>

        <dt>

          <code><a href="assertScHotspot.html">assertScHotspot()</a></code>

        </dt>
        <dd>Assert HotSpot parameters</dd>

        <dt>

          <code><a href="assertScHvg.html">assertScHvg()</a></code>

        </dt>
        <dd>Assert HVG selection parameters</dd>

        <dt>

          <code><a href="assertScMetacells.html">assertScMetacells()</a></code>

        </dt>
        <dd>Assert metacell generation parameters</dd>

        <dt>

          <code><a href="assertScMiloR.html">assertScMiloR()</a></code>

        </dt>
        <dd>Assert MiloR parameters</dd>

        <dt>

          <code><a href="assertScMinQC.html">assertScMinQC()</a></code>

        </dt>
        <dd>Assert SC minimum QC parameters</dd>

        <dt>

          <code><a href="assertScMtxIO.html">assertScMtxIO()</a></code>

        </dt>
        <dd>Assert SC MTX load parameters</dd>

        <dt>

          <code><a href="assertScNeighbours.html">assertScNeighbours()</a></code>

        </dt>
        <dd>Assert neighbour generation parameters</dd>

        <dt>

          <code><a href="assertScScrublet.html">assertScScrublet()</a></code>

        </dt>
        <dd>Assert Scrublet parameters</dd>

        <dt>

          <code><a href="assertScSeacells.html">assertScSeacells()</a></code>

        </dt>
        <dd>Assert SEACells parameters</dd>

        <dt>

          <code><a href="assertScSupercell.html">assertScSupercell()</a></code>

        </dt>
        <dd>Assert SuperCell parameters</dd>

        <dt>

          <code><a href="assertScSyntheticData.html">assertScSyntheticData()</a></code>

        </dt>
        <dd>Assert synthetic data parameters</dd>

        <dt>

          <code><a href="assertScVision.html">assertScVision()</a></code>

        </dt>
        <dd>Assert VISION parameters</dd>

        <dt>

          <code><a href="assertSingleSampleGSEAparams.html">assertSingleSampleGSEAparams()</a></code>

        </dt>
        <dd>Assert ssGSEA parameter</dd>

        <dt>

          <code><a href="aucell_sc.html">aucell_sc()</a></code>

        </dt>
        <dd>Calculate AUC scores (akin to AUCell)</dd>

        <dt>

          <code><a href="batch_correction_bulk_dge.html">batch_correction_bulk_dge()</a></code>

        </dt>
        <dd>Run a linear batch correction</dd>

        <dt>

          <code><a href="bbknn_sc.html">bbknn_sc()</a></code>

        </dt>
        <dd>Run BBKNN</dd>

        <dt>

          <code><a href="bixverse_base_class.html">bixverse_base_class()</a></code>

        </dt>
        <dd>bixverse base class</dd>

        <dt>

          <code><a href="bulk_coexp.html">bulk_coexp()</a></code>

        </dt>
        <dd>Bulk RNAseq co-expression modules</dd>

        <dt>

          <code><a href="bulk_dge.html">bulk_dge()</a></code>

        </dt>
        <dd>Bulk RNAseq differential gene expression class</dd>

        <dt>

          <code><a href="bulk_dge_from_h5ad.html">bulk_dge_from_h5ad()</a></code>

        </dt>
        <dd>Wrapper function to generate bulk_dge object from h5ad</dd>

        <dt>

          <code><a href="c_pca_plot_alphas.html">c_pca_plot_alphas()</a></code>

        </dt>
        <dd>Plot various alphas for the contrastive PCA</dd>

        <dt>

          <code><a href="calc_fgsea.html">calc_fgsea()</a></code>

        </dt>
        <dd>Bixverse implementation of the fgsea algorithm</dd>

        <dt>

          <code><a href="calc_fgsea_simple.html">calc_fgsea_simple()</a></code>

        </dt>
        <dd>Bixverse implementation of the simple fgsea algorithm</dd>

        <dt>

          <code><a href="calc_gsea_traditional.html">calc_gsea_traditional()</a></code>

        </dt>
        <dd>Bixverse implementation of the traditional GSEA algorithm</dd>

        <dt>

          <code><a href="calc_gsva.html">calc_gsva()</a></code>

        </dt>
        <dd>Bixverse implementation of GSVA</dd>

        <dt>

          <code><a href="calc_meta_cell_purity.html">calc_meta_cell_purity()</a></code>

        </dt>
        <dd>Calculate meta cell purity</dd>

        <dt>

          <code><a href="calc_mitch.html">calc_mitch()</a></code>

        </dt>
        <dd>Calculate a mitch gene set enrichments on contrast</dd>

        <dt>

          <code><a href="calc_ssgsea.html">calc_ssgsea()</a></code>

        </dt>
        <dd>Bixverse implementation of ssGSEA</dd>

        <dt>

          <code><a href="calculate_all_dges.html">calculate_all_dges()</a></code>

        </dt>
        <dd>Calculate all possible DGE variants (DEPRECATED!)</dd>

        <dt>

          <code><a href="calculate_critical_value.html">calculate_critical_value()</a></code>

        </dt>
        <dd>Calculates the critical value</dd>

        <dt>

          <code><a href="calculate_dge_hedges.html">calculate_dge_hedges()</a></code>

        </dt>
        <dd>Calculates the Hedge's G effect size</dd>

        <dt>

          <code><a href="calculate_dge_limma.html">calculate_dge_limma()</a></code>

        </dt>
        <dd>Calculates the Limma Voom DGE</dd>

        <dt>

          <code><a href="calculate_diffusion_auc.html">calculate_diffusion_auc()</a></code>

        </dt>
        <dd>Calculate the AUROC for a diffusion score</dd>

        <dt>

          <code><a href="calculate_effect_size.html">calculate_effect_size()</a></code>

        </dt>
        <dd>Calculate the Hedge G effect between two matrices</dd>

        <dt>

          <code><a href="calculate_information_content.html">calculate_information_content()</a></code>

        </dt>
        <dd>Calculate the information content for each ontology term</dd>

        <dt>

          <code><a href="calculate_kbet_sc.html">calculate_kbet_sc()</a></code>

        </dt>
        <dd>Calculate kBET scores</dd>

        <dt>

          <code><a href="calculate_pca_bulk_dge.html">calculate_pca_bulk_dge()</a></code>

        </dt>
        <dd>Calculate PCA on the expression.</dd>

        <dt>

          <code><a href="calculate_pca_sc.html">calculate_pca_sc()</a></code>

        </dt>
        <dd>Run PCA for single cell</dd>

        <dt>

          <code><a href="calculate_rpkm.html">calculate_rpkm()</a></code>

        </dt>
        <dd>RPKM calculation</dd>

        <dt>

          <code><a href="calculate_semantic_sim.html">calculate_semantic_sim()</a></code>

        </dt>
        <dd>Calculate the Resnik or Lin semantic similarity</dd>

        <dt>

          <code><a href="calculate_semantic_sim_mat.html">calculate_semantic_sim_mat()</a></code>

        </dt>
        <dd>Calculate the Resnik or Lin semantic similarity matrix</dd>

        <dt>

          <code><a href="calculate_semantic_sim_onto.html">calculate_semantic_sim_onto()</a></code>

        </dt>
        <dd>Calculate the Resnik or Lin semantic similarity for an ontology.</dd>

        <dt>

          <code><a href="calculate_sparsity_stats.html">calculate_sparsity_stats()</a></code>

        </dt>
        <dd>Helper function to calculate the induced sparsity</dd>

        <dt>

          <code><a href="calculate_tom.html">calculate_tom()</a></code>

        </dt>
        <dd>Calculate the TOM from a correlation matrix</dd>

        <dt>

          <code><a href="calculate_tom_from_exp.html">calculate_tom_from_exp()</a></code>

        </dt>
        <dd>Calculate the TOM from an expression matrix</dd>

        <dt>

          <code><a href="calculate_tpm.html">calculate_tpm()</a></code>

        </dt>
        <dd>TPM calculation</dd>

        <dt>

          <code><a href="calculate_wang_sim.html">calculate_wang_sim()</a></code>

        </dt>
        <dd>Calculate the Wang similarities between terms</dd>

        <dt>

          <code><a href="calculate_wang_sim_mat.html">calculate_wang_sim_mat()</a></code>

        </dt>
        <dd>Calculate the Wang similarity matrix</dd>

        <dt>

          <code><a href="calculate_wang_sim_onto.html">calculate_wang_sim_onto()</a></code>

        </dt>
        <dd>Calculate the Wang similarity for an ontology.</dd>

        <dt>

          <code><a href="call_doublets_manual.html">call_doublets_manual()</a></code>

        </dt>
        <dd>Helper function to manually readjust Scrublet thresholds</dd>

        <dt>

          <code><a href="cell_cycle_genes.html">cell_cycle_genes</a></code>

        </dt>
        <dd>Cell cycle genes</dd>

        <dt>

          <code><a href="change_gene_identifier.html">change_gene_identifier()</a></code>

        </dt>
        <dd>Change the primary gene identifier of bulk_dge</dd>

        <dt>

          <code><a href="checkCellsExist.html">checkCellsExist()</a></code>

        </dt>
        <dd>Check that the cell name exists in the object</dd>

        <dt>

          <code><a href="checkCistargetParams.html">checkCistargetParams()</a></code>

        </dt>
        <dd>Check CisTarget parameters</dd>

        <dt>

          <code><a href="checkCoReMoParams.html">checkCoReMoParams()</a></code>

        </dt>
        <dd>Check CoReMo parameters</dd>

        <dt>

          <code><a href="checkCommunityParams.html">checkCommunityParams()</a></code>

        </dt>
        <dd>Check community detection parameters</dd>

        <dt>

          <code><a href="checkCorGraphParams.html">checkCorGraphParams()</a></code>

        </dt>
        <dd>Check correlation graph parameters</dd>

        <dt>

          <code><a href="checkDGRDLparams.html">checkDGRDLparams()</a></code>

        </dt>
        <dd>Check DGRDL parameters</dd>

        <dt>

          <code><a href="checkFilesExist.html">checkFilesExist()</a></code>

        </dt>
        <dd>Check that files exist</dd>

        <dt>

          <code><a href="checkGSEAParams.html">checkGSEAParams()</a></code>

        </dt>
        <dd>Check GSEA parameters</dd>

        <dt>

          <code><a href="checkGSVAParams.html">checkGSVAParams()</a></code>

        </dt>
        <dd>Check GSVA parameters</dd>

        <dt>

          <code><a href="checkGraphResParams.html">checkGraphResParams()</a></code>

        </dt>
        <dd>Check resolution graph parameters</dd>

        <dt>

          <code><a href="checkIcaIterParams.html">checkIcaIterParams()</a></code>

        </dt>
        <dd>Check ICA randomisation parameters</dd>

        <dt>

          <code><a href="checkIcaNcomps.html">checkIcaNcomps()</a></code>

        </dt>
        <dd>Check ICA no of component parameters</dd>

        <dt>

          <code><a href="checkIcaParams.html">checkIcaParams()</a></code>

        </dt>
        <dd>Check ICA parameters</dd>

        <dt>

          <code><a href="checkSNFParams.html">checkSNFParams()</a></code> <code><a href="checkSNFParams.html">testSNFParams()</a></code>

        </dt>
        <dd>Check SNF parameters</dd>

        <dt>

          <code><a href="checkScBbknn.html">checkScBbknn()</a></code>

        </dt>
        <dd>Check BBKNN parameters</dd>

        <dt>

          <code><a href="checkScBoost.html">checkScBoost()</a></code>

        </dt>
        <dd>Check Boost parameters</dd>

        <dt>

          <code><a href="checkScFastmnn.html">checkScFastmnn()</a></code>

        </dt>
        <dd>Check fastMNN parameters</dd>

        <dt>

          <code><a href="checkScHotspot.html">checkScHotspot()</a></code>

        </dt>
        <dd>Check HotSpot parameters</dd>

        <dt>

          <code><a href="checkScHvg.html">checkScHvg()</a></code>

        </dt>
        <dd>Check HVG selection parameters</dd>

        <dt>

          <code><a href="checkScMetacells.html">checkScMetacells()</a></code>

        </dt>
        <dd>Check metacell generation parameters</dd>

        <dt>

          <code><a href="checkScMiloR.html">checkScMiloR()</a></code>

        </dt>
        <dd>Check MiloR parameters</dd>

        <dt>

          <code><a href="checkScMinQC.html">checkScMinQC()</a></code>

        </dt>
        <dd>Check SC minimum QC parameters</dd>

        <dt>

          <code><a href="checkScMtxIO.html">checkScMtxIO()</a></code>

        </dt>
        <dd>Check SC MTX load parameters</dd>

        <dt>

          <code><a href="checkScNeighbours.html">checkScNeighbours()</a></code>

        </dt>
        <dd>Check neighbour generation parameters</dd>

        <dt>

          <code><a href="checkScScrublet.html">checkScScrublet()</a></code>

        </dt>
        <dd>Check Scrublet parameters</dd>

        <dt>

          <code><a href="checkScSeacells.html">checkScSeacells()</a></code>

        </dt>
        <dd>Check SEACells parameters</dd>

        <dt>

          <code><a href="checkScSupercell.html">checkScSupercell()</a></code>

        </dt>
        <dd>Check SuperCell parameters</dd>

        <dt>

          <code><a href="checkScSyntheticData.html">checkScSyntheticData()</a></code>

        </dt>
        <dd>Check synthetic data parameters</dd>

        <dt>

          <code><a href="checkScVision.html">checkScVision()</a></code>

        </dt>
        <dd>Check VISION parameters</dd>

        <dt>

          <code><a href="checkSingleSampleGSEAparams.html">checkSingleSampleGSEAparams()</a></code>

        </dt>
        <dd>Check ssGSEA parameters</dd>

        <dt>

          <code><a href="check_dim.html">check_dim()</a></code>

        </dt>
        <dd>Helper to assert same number of samples</dd>

        <dt>

          <code><a href="check_pca_grp_differences.html">check_pca_grp_differences()</a></code>

        </dt>
        <dd>Helper function to check if PC1 and/or PC2 distinguish groups</dd>

        <dt>

          <code><a href="community_detection.html">community_detection()</a></code>

        </dt>
        <dd>Identify privileged communities based on a given diffusion vector</dd>

        <dt>

          <code><a href="community_stability.html">community_stability()</a></code>

        </dt>
        <dd>Assess stability of ICA components</dd>

        <dt>

          <code><a href="component_mutual_information.html">component_mutual_information()</a></code>

        </dt>
        <dd>Assess the NMI of the ICA components</dd>

        <dt>

          <code><a href="constrained_page_rank.html">constrained_page_rank()</a></code>

        </dt>
        <dd>Constrained personalised page rank</dd>

        <dt>

          <code><a href="constrained_page_rank_ls.html">constrained_page_rank_ls()</a></code>

        </dt>
        <dd>Constrained personalised page rank over a list</dd>

        <dt>

          <code><a href="contrastive_pca.html">contrastive_pca()</a></code>

        </dt>
        <dd>Apply contrastive PCA.</dd>

        <dt>

          <code><a href="contrastive_pca_processing.html">contrastive_pca_processing()</a></code>

        </dt>
        <dd>Prepare class for contrastive PCA</dd>

        <dt>

          <code><a href="cor_module_check_epsilon.html">cor_module_check_epsilon()</a></code>

        </dt>
        <dd>Iterate through different epsilon parameters</dd>

        <dt>

          <code><a href="cor_module_coremo_clustering.html">cor_module_coremo_clustering()</a></code>

        </dt>
        <dd>Generates CoReMo-based gene modules</dd>

        <dt>

          <code><a href="cor_module_coremo_cor_sign.html">cor_module_coremo_cor_sign()</a></code>

        </dt>
        <dd>Split CoReMo modules by correlation sign</dd>

        <dt>

          <code><a href="cor_module_coremo_eigengene.html">cor_module_coremo_eigengene()</a></code>

        </dt>
        <dd>Calculate Eigengenes for CoReMo modules</dd>

        <dt>

          <code><a href="cor_module_coremo_stability.html">cor_module_coremo_stability()</a></code>

        </dt>
        <dd>Assesses CoReMo-based gene module stability</dd>

        <dt>

          <code><a href="cor_module_graph_check_res.html">cor_module_graph_check_res()</a></code>

        </dt>
        <dd>Iterate through Leiden resolutions for graph-based community detection.</dd>

        <dt>

          <code><a href="cor_module_graph_final_modules.html">cor_module_graph_final_modules()</a></code>

        </dt>
        <dd>Identify correlation-based gene modules via graphs</dd>

        <dt>

          <code><a href="cor_module_processing.html">cor_module_processing()</a></code>

        </dt>
        <dd>Prepare correlation-based module detection</dd>

        <dt>

          <code><a href="cor_module_tom.html">cor_module_tom()</a></code>

        </dt>
        <dd>Update the correlation matrix to a TOM</dd>

        <dt>

          <code><a href="coremo_cluster_quality.html">coremo_cluster_quality()</a></code>

        </dt>
        <dd>Coremo: measures the quality of the clusters</dd>

        <dt>

          <code><a href="coremo_tree_cut.html">coremo_tree_cut()</a></code>

        </dt>
        <dd>Coremo: cuts a hierarchical cluster based on k</dd>

        <dt>

          <code><a href="create_dist_obj.html">create_dist_obj()</a></code>

        </dt>
        <dd>Create distance object from a vector</dd>

        <dt>

          <code><a href="create_sparse_matrix.html">create_sparse_matrix()</a></code>

        </dt>
        <dd>Helper function transform Rust counts into sparse matrices</dd>

        <dt>

          <code><a href="dgrdl_grid_search.html">dgrdl_grid_search()</a></code>

        </dt>
        <dd>Grid search over DGRDL parameters</dd>

        <dt>

          <code><a href="dgrdl_result.html">dgrdl_result()</a></code>

        </dt>
        <dd>Run DGRDL with the specified parameters</dd>

        <dt>

          <code><a href="diffcor_module_processing.html">diffcor_module_processing()</a></code>

        </dt>
        <dd>Prepare differential correlation-based module detection</dd>

        <dt>

          <code><a href="diffuse_seed_nodes.html">diffuse_seed_nodes()</a></code>

        </dt>
        <dd>Diffuse seed genes over a network</dd>

        <dt>

          <code><a href="doublet_detection_boost_sc.html">doublet_detection_boost_sc()</a></code>

        </dt>
        <dd>Doublet detection with boosted doublet classification</dd>

        <dt>

          <code><a href="f1_score_confusion_mat.html">f1_score_confusion_mat()</a></code>

        </dt>
        <dd>F1 scores on top of a confusion matrix</dd>

        <dt>

          <code><a href="fast_ica_rust.html">fast_ica_rust()</a></code>

        </dt>
        <dd>Fast ICA via Rust</dd>

        <dt>

          <code><a href="fast_ica_rust_helper.html">fast_ica_rust_helper()</a></code>

        </dt>
        <dd>Fast ICA via Rust from processed data</dd>

        <dt>

          <code><a href="fast_mnn_sc.html">fast_mnn_sc()</a></code>

        </dt>
        <dd>Run fastMNN</dd>

        <dt>

          <code><a href="fgsea_go_elim.html">fgsea_go_elim()</a></code>

        </dt>
        <dd>Run GO enrichment with elimination method over a continuous vectors</dd>

        <dt>

          <code><a href="fgsea_simple_go_elim.html">fgsea_simple_go_elim()</a></code>

        </dt>
        <dd>Run GO enrichment with elimination with fgsea simple</dd>

        <dt>

          <code><a href="filter_similarities.html">filter_similarities()</a></code>

        </dt>
        <dd>Filter the calculated similarities</dd>

        <dt>

          <code><a href="finalise_matrix.html">finalise_matrix()</a></code>

        </dt>
        <dd>Finalise the count matrix</dd>

        <dt>

          <code><a href="find_all_markers_sc.html">find_all_markers_sc()</a></code>

        </dt>
        <dd>Find all markers</dd>

        <dt>

          <code><a href="find_clusters_sc.html">find_clusters_sc()</a></code>

        </dt>
        <dd>Graph-based clustering of cells on the sNN graph</dd>

        <dt>

          <code><a href="find_hvg_batch_aware_sc.html">find_hvg_batch_aware_sc()</a></code>

        </dt>
        <dd>Identify HVGs (batch aware)</dd>

        <dt>

          <code><a href="find_hvg_sc.html">find_hvg_sc()</a></code>

        </dt>
        <dd>Identify HVGs</dd>

        <dt>

          <code><a href="find_markers_sc.html">find_markers_sc()</a></code>

        </dt>
        <dd>Calculate DGE between two cell groups</dd>

        <dt>

          <code><a href="find_neighbours_sc.html">find_neighbours_sc()</a></code>

        </dt>
        <dd>Find the neighbours for single cell.</dd>

        <dt>

          <code><a href="find_rbh_communities.html">find_rbh_communities()</a></code>

        </dt>
        <dd>Find RBH communities</dd>

        <dt>

          <code><a href="fix_contrast_names.html">fix_contrast_names()</a></code>

        </dt>
        <dd>Fixes contrast names for DGEs</dd>

        <dt>

          <code><a href="fix_meta_data_column.html">fix_meta_data_column()</a></code>

        </dt>
        <dd>Helper to fix meta-data columns to be R conform</dd>

        <dt>

          <code><a href="flip_ica_loading_signs.html">flip_ica_loading_signs()</a></code>

        </dt>
        <dd>Flips the ICA source sign</dd>

        <dt>

          <code><a href="gene_ontology_data.html">gene_ontology_data()</a></code>

        </dt>
        <dd>Gene Ontology data</dd>

        <dt>

          <code><a href="gene_set_proportions_sc.html">gene_set_proportions_sc()</a></code>

        </dt>
        <dd>Calculate the proportions of reads for specific gene sets</dd>

        <dt>

          <code><a href="generate_gene_module_data.html">generate_gene_module_data()</a></code>

        </dt>
        <dd>Generates synthetic gene module data.</dd>

        <dt>

          <code><a href="generate_knn_sc.html">generate_knn_sc()</a></code>

        </dt>
        <dd>Generate the KNN data with distances</dd>

        <dt>

          <code><a href="generate_null_perm_gs.html">generate_null_perm_gs()</a></code>

        </dt>
        <dd>Generate random gene sets for VISION p-value calculations</dd>

        <dt>

          <code><a href="generate_perm_diffusion_vecs.html">generate_perm_diffusion_vecs()</a></code>

        </dt>
        <dd>Generates random permutation vectors</dd>

        <dt>

          <code><a href="generate_personalisation_vec.html">generate_personalisation_vec()</a></code>

        </dt>
        <dd>Helper function to create personalisation vectors</dd>

        <dt>

          <code><a href="generate_rbh_graph.html">generate_rbh_graph()</a></code>

        </dt>
        <dd>Generate an RBH graph.</dd>

        <dt>

          <code><a href="generate_single_cell_test_data.html">generate_single_cell_test_data()</a></code>

        </dt>
        <dd>Single cell test data</dd>

        <dt>

          <code><a href="get_available_embeddings.html">get_available_embeddings()</a></code>

        </dt>
        <dd>Get the available embeddings from the cache</dd>

        <dt>

          <code><a href="get_c_pca_factors.html">get_c_pca_factors()</a></code>

        </dt>
        <dd>Get the contrastive PCA factors</dd>

        <dt>

          <code><a href="get_c_pca_loadings.html">get_c_pca_loadings()</a></code>

        </dt>
        <dd>Get the contrastive PCA loadings</dd>

        <dt>

          <code><a href="get_cell_indices.html">get_cell_indices()</a></code>

        </dt>
        <dd>Get the index position for a gene</dd>

        <dt>

          <code><a href="get_cell_names.html">get_cell_names()</a></code>

        </dt>
        <dd>Get the cell names</dd>

        <dt>

          <code><a href="get_cell_ranger_params.html">get_cell_ranger_params()</a></code>

        </dt>
        <dd>Helper to generate cell ranger input parameters</dd>

        <dt>

          <code><a href="get_cells_to_keep.html">get_cells_to_keep()</a></code>

        </dt>
        <dd>Get the cells to keep</dd>

        <dt>

          <code><a href="get_cor_graph.html">get_cor_graph()</a></code>

        </dt>
        <dd>Get correlation-based graph</dd>

        <dt>

          <code><a href="get_cores.html">get_cores()</a></code>

        </dt>
        <dd>Helper function for core detection.</dd>

        <dt>

          <code><a href="get_counts_from_rust.html">get_counts_from_rust()</a></code>

        </dt>
        <dd>Helper function to get counts from Rust</dd>

        <dt>

          <code><a href="get_dge_effect_sizes.html">get_dge_effect_sizes()</a></code>

        </dt>
        <dd>Return the effect size results</dd>

        <dt>

          <code><a href="get_dge_limma_voom.html">get_dge_limma_voom()</a></code>

        </dt>
        <dd>Return the Limma Voom results</dd>

        <dt>

          <code><a href="get_dge_list.html">get_dge_list()</a></code>

        </dt>
        <dd>Return the DGEList</dd>

        <dt>

          <code><a href="get_dge_qc_plot.html">get_dge_qc_plot()</a></code>

        </dt>
        <dd>Return QC plots</dd>

        <dt>

          <code><a href="get_diffcor_graph.html">get_diffcor_graph()</a></code>

        </dt>
        <dd>Get differential correlation-based graph</dd>

        <dt>

          <code><a href="get_differential_abundance_res.html">get_differential_abundance_res()</a></code>

        </dt>
        <dd>Get the differential abundance results</dd>

        <dt>

          <code><a href="get_diffusion_perms.html">get_diffusion_perms()</a></code>

        </dt>
        <dd>Get the diffusion permutations</dd>

        <dt>

          <code><a href="get_diffusion_vector.html">get_diffusion_vector()</a></code>

        </dt>
        <dd>Get the diffusion vector</dd>

        <dt>

          <code><a href="get_embedding.html">get_embedding()</a></code>

        </dt>
        <dd>Get the embedding from the cache</dd>

        <dt>

          <code><a href="get_epsilon_res.html">get_epsilon_res()</a></code>

        </dt>
        <dd>Return the epsilon data</dd>

        <dt>

          <code><a href="get_fpkm_counts.html">get_fpkm_counts()</a></code>

        </dt>
        <dd>Return the FPKM-normalised counts</dd>

        <dt>

          <code><a href="get_gene_indices.html">get_gene_indices()</a></code>

        </dt>
        <dd>Get the index position for a gene</dd>

        <dt>

          <code><a href="get_gene_lengths.html">get_gene_lengths()</a></code>

        </dt>
        <dd>Get the gene lengths</dd>

        <dt>

          <code><a href="get_gene_lengths.bulk_coexp.html">get_gene_lengths.bulk_coexp</a></code>

        </dt>
        <dd>Get gene set lengths for a bulk_coexp class</dd>

        <dt>

          <code><a href="get_gene_lengths.matrix.html">get_gene_lengths(<i>&lt;matrix&gt;</i>)</a></code>

        </dt>
        <dd>Get gene set lengths for a matrix</dd>

        <dt>

          <code><a href="get_gene_names.html">get_gene_names()</a></code>

        </dt>
        <dd>Get the gene names</dd>

        <dt>

          <code><a href="get_gene_names_from_idx.html">get_gene_names_from_idx()</a></code>

        </dt>
        <dd>Get the gene names based on the gene idx</dd>

        <dt>

          <code><a href="get_go_data_human.html">get_go_data_human()</a></code>

        </dt>
        <dd>Wrapper function to load and process the gene ontology data.</dd>

        <dt>

          <code><a href="get_go_levels.html">get_go_levels()</a></code>

        </dt>
        <dd>Helper to get the Gene Ontology levels</dd>

        <dt>

          <code><a href="get_grid_search_res.html">get_grid_search_res()</a></code>

        </dt>
        <dd>Get the grid search results</dd>

        <dt>

          <code><a href="get_h5ad_dimensions.html">get_h5ad_dimensions()</a></code>

        </dt>
        <dd>Helper function to get the dimensions and compressed sparse format</dd>

        <dt>

          <code><a href="get_hotspot_membership.html">get_hotspot_membership()</a></code>

        </dt>
        <dd>Get the hotspot gene membership table</dd>

        <dt>

          <code><a href="get_hvg.html">get_hvg()</a></code>

        </dt>
        <dd>Get the HVG</dd>

        <dt>

          <code><a href="get_ica_stability_res.html">get_ica_stability_res()</a></code>

        </dt>
        <dd>Get the ICA component data (stability, convergence, nMI)</dd>

        <dt>

          <code><a href="get_index_cells.html">get_index_cells()</a></code>

        </dt>
        <dd>Get the index cells</dd>

        <dt>

          <code><a href="get_inflection_point.html">get_inflection_point()</a></code>

        </dt>
        <dd>Identify the inflection point for elbow-like data</dd>

        <dt>

          <code><a href="get_knn_dist.html">get_knn_dist()</a></code>

        </dt>
        <dd>Get the KNN distance measures</dd>

        <dt>

          <code><a href="get_knn_mat.html">get_knn_mat()</a></code>

        </dt>
        <dd>Get the KNN matrix</dd>

        <dt>

          <code><a href="get_meta_cell_matrices.html">get_meta_cell_matrices()</a></code>

        </dt>
        <dd>Create sparse dgRMatrix matrices for raw and norm counts</dd>

        <dt>

          <code><a href="get_meta_cells_sc.html">get_meta_cells_sc()</a></code>

        </dt>
        <dd>Generate meta cells based on hdWGCNA and return a <code>meta_cells</code> object</dd>

        <dt>

          <code><a href="get_metadata.html">get_metadata()</a></code>

        </dt>
        <dd>Return the metadata</dd>

        <dt>

          <code><a href="get_miloR_abundances_sc.html">get_miloR_abundances_sc()</a></code>

        </dt>
        <dd>Generate an miloR abundance object for differential abundance testing</dd>

        <dt>

          <code><a href="get_model_fit.html">get_model_fit()</a></code>

        </dt>
        <dd>Get the fitted model</dd>

        <dt>

          <code><a href="get_obs_data.html">get_obs_data()</a></code>

        </dt>
        <dd>Get the ready obs data from various sub method</dd>

        <dt>

          <code><a href="get_ontology_ancestry.html">get_ontology_ancestry()</a></code>

        </dt>
        <dd>Return ancestry terms from an ontology</dd>

        <dt>

          <code><a href="get_outputs.html">get_outputs()</a></code>

        </dt>
        <dd>Return the outputs</dd>

        <dt>

          <code><a href="get_params.html">get_params()</a></code> <code><a href="get_params.html">get_params.sc_hotspot()</a></code> <code><a href="get_params.html">get_params.sc_miloR()</a></code>

        </dt>
        <dd>Get the parameters that were used.</dd>

        <dt>

          <code><a href="get_pca_factors.html">get_pca_factors()</a></code>

        </dt>
        <dd>Get the PCA factors</dd>

        <dt>

          <code><a href="get_pca_loadings.html">get_pca_loadings()</a></code>

        </dt>
        <dd>Get the PCA loadings</dd>

        <dt>

          <code><a href="get_pca_singular_val.html">get_pca_singular_val()</a></code>

        </dt>
        <dd>Get the PCA singular values</dd>

        <dt>

          <code><a href="get_rbh_res.html">get_rbh_res()</a></code>

        </dt>
        <dd>Get the RBH results</dd>

        <dt>

          <code><a href="get_resolution_res.html">get_resolution_res()</a></code>

        </dt>
        <dd>Return the resolution results</dd>

        <dt>

          <code><a href="get_results.html">get_results()</a></code>

        </dt>
        <dd>Get the final results from the class</dd>

        <dt>

          <code><a href="get_rust_count_cell_f_path.html">get_rust_count_cell_f_path()</a></code>

        </dt>
        <dd>Get the file path to the counts in cell format</dd>

        <dt>

          <code><a href="get_rust_count_gene_f_path.html">get_rust_count_gene_f_path()</a></code>

        </dt>
        <dd>Get the file path to the counts in gene format</dd>

        <dt>

          <code><a href="get_sc_cache.html">get_sc_cache()</a></code>

        </dt>
        <dd>Getter the memory-stored data from the class</dd>

        <dt>

          <code><a href="get_sc_counts.html">get_sc_counts()</a></code>

        </dt>
        <dd>Getter the counts</dd>

        <dt>

          <code><a href="get_sc_duckdb.html">get_sc_duckdb()</a></code>

        </dt>
        <dd>Getter for the single cell DuckDB connection</dd>

        <dt>

          <code><a href="get_sc_map.html">get_sc_map()</a></code>

        </dt>
        <dd>Getter for the different maps in the object</dd>

        <dt>

          <code><a href="get_sc_obs.html">get_sc_obs()</a></code>

        </dt>
        <dd>Getter the obs table</dd>

        <dt>

          <code><a href="get_sc_rust_ptr.html">get_sc_rust_ptr()</a></code>

        </dt>
        <dd>Getter for the single cell Rust pointer</dd>

        <dt>

          <code><a href="get_sc_var.html">get_sc_var()</a></code>

        </dt>
        <dd>Getter the var table</dd>

        <dt>

          <code><a href="get_seacells_sc.html">get_seacells_sc()</a></code>

        </dt>
        <dd>Generate meta cells based on SEACells and return a <code>meta_cells</code> object</dd>

        <dt>

          <code><a href="get_seurat_counts_to_list.html">get_seurat_counts_to_list()</a></code>

        </dt>
        <dd>Transform Seurat raw counts into a List</dd>

        <dt>

          <code><a href="get_sim_matrix.html">get_sim_matrix()</a></code>

        </dt>
        <dd>Get the similarity matrix</dd>

        <dt>

          <code><a href="get_snf_adjcacency_mat.html">get_snf_adjcacency_mat()</a></code>

        </dt>
        <dd>Get an individual affinity matrix</dd>

        <dt>

          <code><a href="get_snf_final_mat.html">get_snf_final_mat()</a></code>

        </dt>
        <dd>Get the final SNF matrix</dd>

        <dt>

          <code><a href="get_snf_params.html">get_snf_params()</a></code>

        </dt>
        <dd>Get the SNF params</dd>

        <dt>

          <code><a href="get_snn_graph.html">get_snn_graph()</a></code>

        </dt>
        <dd>Get the sNN graph</dd>

        <dt>

          <code><a href="get_supercells_sc.html">get_supercells_sc()</a></code>

        </dt>
        <dd>Generate SuperCells and return a <code>meta_cells</code> object</dd>

        <dt>

          <code><a href="get_tpm_counts.html">get_tpm_counts()</a></code>

        </dt>
        <dd>Return the TPM-normalised counts</dd>

        <dt>

          <code><a href="gse_go_elim_method.html">gse_go_elim_method()</a></code>

        </dt>
        <dd>Run gene ontology enrichment with elimination method.</dd>

        <dt>

          <code><a href="gse_go_elim_method_list.html">gse_go_elim_method_list()</a></code>

        </dt>
        <dd>Run gene ontology enrichment with elimination method over a list.</dd>

        <dt>

          <code><a href="gse_hypergeometric.html">gse_hypergeometric()</a></code>

        </dt>
        <dd>Gene set enrichment (GSE) based on a hypergeometric test.</dd>

        <dt>

          <code><a href="gse_hypergeometric_list.html">gse_hypergeometric_list()</a></code>

        </dt>
        <dd>Gene set enrichment (GSE) based on a hypergeometric test over a list.</dd>

        <dt>

          <code><a href="hedges_g_dge.html">hedges_g_dge()</a></code>

        </dt>
        <dd>Calculate the effect size</dd>

        <dt>

          <code><a href="hotspot_autocor_sc.html">hotspot_autocor_sc()</a></code>

        </dt>
        <dd>Calculate the local auto-correlation of a gene</dd>

        <dt>

          <code><a href="hotspot_gene_cor_sc.html">hotspot_gene_cor_sc()</a></code>

        </dt>
        <dd>Calculate the local pairwise gene-gene correlation</dd>

        <dt>

          <code><a href="ica_evaluate_comp.html">ica_evaluate_comp()</a></code>

        </dt>
        <dd>Iterate over different ncomp parameters for ICA</dd>

        <dt>

          <code><a href="ica_optimal_ncomp.html">ica_optimal_ncomp()</a></code>

        </dt>
        <dd>Identify stability inflection point</dd>

        <dt>

          <code><a href="ica_processing.html">ica_processing()</a></code>

        </dt>
        <dd>Prepare class for ICA</dd>

        <dt>

          <code><a href="ica_stabilised_results.html">ica_stabilised_results()</a></code>

        </dt>
        <dd>Run stabilised ICA with a given number of components</dd>

        <dt>

          <code><a href="knn_graph_label_propagation.html">knn_graph_label_propagation()</a></code>

        </dt>
        <dd>kNN-based graph label propagation</dd>

        <dt>

          <code><a href="load_existing.html">load_existing()</a></code>

        </dt>
        <dd>Load an existing single_cell_exp from disk</dd>

        <dt>

          <code><a href="load_go_human_data.html">load_go_human_data()</a></code>

        </dt>
        <dd>Get the Gene Ontology data human</dd>

        <dt>

          <code><a href="load_h5ad.html">load_h5ad()</a></code>

        </dt>
        <dd>Load in h5ad to <code>single_cell_exp</code></dd>

        <dt>

          <code><a href="load_mtx.html">load_mtx()</a></code>

        </dt>
        <dd>Load in mtx/plain text files to <code>single_cell_exp</code> (nightly!)</dd>

        <dt>

          <code><a href="load_r_data.html">load_r_data()</a></code>

        </dt>
        <dd>Load in data directly from R objects.</dd>

        <dt>

          <code><a href="load_seurat.html">load_seurat()</a></code>

        </dt>
        <dd>Load in Seurat to <code>single_cell_exp</code></dd>

        <dt>

          <code><a href="meta_cells.html">meta_cells()</a></code>

        </dt>
        <dd>bixverse meta cell class</dd>

        <dt>

          <code><a href="module_scores_sc.html">module_scores_sc()</a></code>

        </dt>
        <dd>Calculate module activity scores</dd>

        <dt>

          <code><a href="multilevel_error.html">multilevel_error()</a></code>

        </dt>
        <dd>Helper function to calculate the multi-level error</dd>

        <dt>

          <code><a href="network_diffusions.html">network_diffusions()</a></code>

        </dt>
        <dd>Network diffusion class</dd>

        <dt>

          <code><a href="new_sc_cache.html">new_sc_cache()</a></code>

        </dt>
        <dd>Helper function to hold relevant cached data</dd>

        <dt>

          <code><a href="new_sc_hotspot_res.html">new_sc_hotspot_res()</a></code>

        </dt>
        <dd>Helper function to generate HotSpot data</dd>

        <dt>

          <code><a href="new_sc_knn.html">new_sc_knn()</a></code>

        </dt>
        <dd>Helper function to generate kNN data with distances</dd>

        <dt>

          <code><a href="new_sc_mapper.html">new_sc_mapper()</a></code>

        </dt>
        <dd>Helper function to construct relevant maps</dd>

        <dt>

          <code><a href="new_sc_miloR_res.html">new_sc_miloR_res()</a></code>

        </dt>
        <dd>Generate a new miloR result</dd>

        <dt>

          <code><a href="nightly_feature.html">nightly_feature()</a></code>

        </dt>
        <dd>Helper functions to mark something unstable/nightly</dd>

        <dt>

          <code><a href="normalise_bulk_dge.html">normalise_bulk_dge()</a></code>

        </dt>
        <dd>Normalise the count data for DGE.</dd>

        <dt>

          <code><a href="one_hot_encode.html">one_hot_encode()</a></code>

        </dt>
        <dd>Helper to generate one-hot encodings</dd>

        <dt>

          <code><a href="ontology.html">ontology()</a></code>

        </dt>
        <dd>Ontology class</dd>

        <dt>

          <code><a href="ot_harmonic_score.html">ot_harmonic_score()</a></code>

        </dt>
        <dd>Calculates a harmonic sum normalised between 0 to 1.</dd>

        <dt>

          <code><a href="params_boost.html">params_boost()</a></code>

        </dt>
        <dd>Wrapper function for Boost parameters</dd>

        <dt>

          <code><a href="params_cistarget.html">params_cistarget()</a></code>

        </dt>
        <dd>Wrapper function to CisTarget parameters</dd>

        <dt>

          <code><a href="params_community_detection.html">params_community_detection()</a></code>

        </dt>
        <dd>Wrapper function to generate community detection parameters</dd>

        <dt>

          <code><a href="params_cor_graph.html">params_cor_graph()</a></code>

        </dt>
        <dd>Wrapper function for graph generation</dd>

        <dt>

          <code><a href="params_coremo.html">params_coremo()</a></code>

        </dt>
        <dd>Wrapper function to generate CoReMo parameters</dd>

        <dt>

          <code><a href="params_dgrdl.html">params_dgrdl()</a></code>

        </dt>
        <dd>Wrapper function to generate DGRDL parameters</dd>

        <dt>

          <code><a href="params_graph_resolution.html">params_graph_resolution()</a></code>

        </dt>
        <dd>Wrapper function to generate resolution parameters for Leiden or Louvain clustering.</dd>

        <dt>

          <code><a href="params_gsea.html">params_gsea()</a></code>

        </dt>
        <dd>Wrapper function to generate GSEA parameters</dd>

        <dt>

          <code><a href="params_gsva.html">params_gsva()</a></code>

        </dt>
        <dd>Wrapper function to generate GSVA parameters</dd>

        <dt>

          <code><a href="params_hvg_defaults.html">params_hvg_defaults()</a></code>

        </dt>
        <dd>Helper function to generate HVG defaults</dd>

        <dt>

          <code><a href="params_ica_general.html">params_ica_general()</a></code>

        </dt>
        <dd>Wrapper function for standard ICA parameters</dd>

        <dt>

          <code><a href="params_ica_ncomp.html">params_ica_ncomp()</a></code>

        </dt>
        <dd>Wrapper function for ICA ncomp iterations</dd>

        <dt>

          <code><a href="params_ica_randomisation.html">params_ica_randomisation()</a></code>

        </dt>
        <dd>Wrapper function for ICA randomisation</dd>

        <dt>

          <code><a href="params_knn_defaults.html">params_knn_defaults()</a></code>

        </dt>
        <dd>Helper function to generate kNN defaults</dd>

        <dt>

          <code><a href="params_norm_doublet_detection_defaults.html">params_norm_doublet_detection_defaults()</a></code>

        </dt>
        <dd>Helper function to generate normalisation defaults for doublet detection.</dd>

        <dt>

          <code><a href="params_pca_defaults.html">params_pca_defaults()</a></code>

        </dt>
        <dd>Helper function to generate default parameters for PCA</dd>

        <dt>

          <code><a href="params_sc_bbknn.html">params_sc_bbknn()</a></code>

        </dt>
        <dd>Wrapper function for the BBKNN parameters</dd>

        <dt>

          <code><a href="params_sc_fastmnn.html">params_sc_fastmnn()</a></code>

        </dt>
        <dd>Wrapper function for the fastMNN parameters</dd>

        <dt>

          <code><a href="params_sc_hotspot.html">params_sc_hotspot()</a></code>

        </dt>
        <dd>Wrapper function for parameters for HotSpot</dd>

        <dt>

          <code><a href="params_sc_hvg.html">params_sc_hvg()</a></code>

        </dt>
        <dd>Wrapper function for HVG detection parameters.</dd>

        <dt>

          <code><a href="params_sc_metacells.html">params_sc_metacells()</a></code>

        </dt>
        <dd>Wrapper function for parameters for meta cell generation</dd>

        <dt>

          <code><a href="params_sc_miloR.html">params_sc_miloR()</a></code>

        </dt>
        <dd>Wrapper function for parameters for MiloR</dd>

        <dt>

          <code><a href="params_sc_min_quality.html">params_sc_min_quality()</a></code>

        </dt>
        <dd>Wrapper function to generate QC metric params for single cell</dd>

        <dt>

          <code><a href="params_sc_mtx_io.html">params_sc_mtx_io()</a></code>

        </dt>
        <dd>Wrapper function to provide data for mtx-based loading</dd>

        <dt>

          <code><a href="params_sc_neighbours.html">params_sc_neighbours()</a></code>

        </dt>
        <dd>Wrapper function for parameters for neighbour identification in single cell</dd>

        <dt>

          <code><a href="params_sc_seacells.html">params_sc_seacells()</a></code>

        </dt>
        <dd>Wrapper function for the SEACells parameters</dd>

        <dt>

          <code><a href="params_sc_supercell.html">params_sc_supercell()</a></code>

        </dt>
        <dd>Wrapper function for parameters for SuperCell generation</dd>

        <dt>

          <code><a href="params_sc_synthetic_data.html">params_sc_synthetic_data()</a></code>

        </dt>
        <dd>Default parameters for generation of synthetic data</dd>

        <dt>

          <code><a href="params_sc_vision.html">params_sc_vision()</a></code>

        </dt>
        <dd>Wrapper function for parameters for VISION with auto-correlation</dd>

        <dt>

          <code><a href="params_scrublet.html">params_scrublet()</a></code>

        </dt>
        <dd>Wrapper function for Scrublet doublet detection parameters</dd>

        <dt>

          <code><a href="params_snf.html">params_snf()</a></code>

        </dt>
        <dd>Wrapper function to generate SNF parameters</dd>

        <dt>

          <code><a href="params_ssgsea.html">params_ssgsea()</a></code>

        </dt>
        <dd>Wrapper function to generate ssGSEA parameters</dd>

        <dt>

          <code><a href="permute_seed_nodes.html">permute_seed_nodes()</a></code>

        </dt>
        <dd>Generate permuation scores for the diffusion</dd>

        <dt>

          <code><a href="plot.cpca_synthetic_data.html">plot(<i>&lt;cpca_synthetic_data&gt;</i>)</a></code>

        </dt>
        <dd>Plot the contrastive PCA example data</dd>

        <dt>

          <code><a href="plot.synthetic_matrix_simple.html">plot(<i>&lt;synthetic_matrix_simple&gt;</i>)</a></code>

        </dt>
        <dd>Plot the (simple) synthetic gene expression</dd>

        <dt>

          <code><a href="plot_boxplot_normalization.html">plot_boxplot_normalization()</a></code>

        </dt>
        <dd>Helper plot function for boxplot of normalized data</dd>

        <dt>

          <code><a href="plot_epsilon_res.html">plot_epsilon_res()</a></code>

        </dt>
        <dd>Plot the epsilon vs. power law goodness of fit result</dd>

        <dt>

          <code><a href="plot_hvgs.html">plot_hvgs()</a></code>

        </dt>
        <dd>Plot the highly variable genes</dd>

        <dt>

          <code><a href="plot_ica_ncomp_params.html">plot_ica_ncomp_params()</a></code>

        </dt>
        <dd>Plot various parameters with no comp</dd>

        <dt>

          <code><a href="plot_ica_stability_individual.html">plot_ica_stability_individual()</a></code>

        </dt>
        <dd>Plot the stability of the ICA components</dd>

        <dt>

          <code><a href="plot_optimal_cuts.html">plot_optimal_cuts()</a></code>

        </dt>
        <dd>Plot the k cuts vs median R2</dd>

        <dt>

          <code><a href="plot_pca.html">plot_pca()</a></code>

        </dt>
        <dd>Helper plot function for pca with contrasts</dd>

        <dt>

          <code><a href="plot_pca_res.html">plot_pca_res()</a></code>

        </dt>
        <dd>Plot the PCA data</dd>

        <dt>

          <code><a href="plot_preprocessing_genes.html">plot_preprocessing_genes()</a></code>

        </dt>
        <dd>Helper plot function of distribution of genes by samples</dd>

        <dt>

          <code><a href="plot_preprocessing_outliers.html">plot_preprocessing_outliers()</a></code>

        </dt>
        <dd>Helper plot function for identification of outliers</dd>

        <dt>

          <code><a href="plot_rbf_impact.html">plot_rbf_impact()</a></code>

        </dt>
        <dd>Helper function to plot distance to affinity relationship</dd>

        <dt>

          <code><a href="plot_resolution_res.html">plot_resolution_res()</a></code>

        </dt>
        <dd>Plot the resolution results.</dd>

        <dt>

          <code><a href="plot_voom_normalization.html">plot_voom_normalization()</a></code>

        </dt>
        <dd>Helper plot function for Voom normalisation</dd>

        <dt>

          <code><a href="pre_process_sim_onto.html">pre_process_sim_onto()</a></code>

        </dt>
        <dd>Pre-process data for subsequent ontology similarity</dd>

        <dt>

          <code><a href="prep_data_gower_hamming_dist.html">prep_data_gower_hamming_dist()</a></code>

        </dt>
        <dd>Transform data.tables into matrices for distance calculations</dd>

        <dt>

          <code><a href="prep_limma_contrasts.html">prep_limma_contrasts()</a></code>

        </dt>
        <dd>Create all limma contrasts from a provided string</dd>

        <dt>

          <code><a href="prep_stats_pathways.html">prep_stats_pathways()</a></code>

        </dt>
        <dd>Helper function to prepare data for GSEA</dd>

        <dt>

          <code><a href="preprocess_bulk_coexp.html">preprocess_bulk_coexp()</a></code>

        </dt>
        <dd>Process the raw data</dd>

        <dt>

          <code><a href="preprocess_bulk_dge.html">preprocess_bulk_dge()</a></code>

        </dt>
        <dd>QC on the bulk dge data (DEPRECATED!)</dd>

        <dt>

          <code><a href="print.bulk_coexp.html">print.bulk_coexp</a></code>

        </dt>
        <dd>print Method for bulk_coexp object</dd>

        <dt>

          <code><a href="print.gene_ontology_data.html">print.gene_ontology_data</a></code>

        </dt>
        <dd>print Method for gene_ontology_data object</dd>

        <dt>

          <code><a href="print.ontology.html">print.ontology</a></code>

        </dt>
        <dd>print Method for ontology object</dd>

        <dt>

          <code><a href="process_cistarget_res.html">process_cistarget_res()</a></code>

        </dt>
        <dd>Helper to process CisTarget results</dd>

        <dt>

          <code><a href="process_go_data.html">process_go_data()</a></code>

        </dt>
        <dd>Process Gene Ontology data into the right format</dd>

        <dt>

          <code><a href="qc_bulk_dge.html">qc_bulk_dge()</a></code>

        </dt>
        <dd>QC on the bulk dge data</dd>

        <dt>

          <code><a href="rbh_graph.html">rbh_graph()</a></code>

        </dt>
        <dd>Reciprocal best hit graph</dd>

        <dt>

          <code><a href="read_motif_annotation_file.html">read_motif_annotation_file()</a></code>

        </dt>
        <dd>Read in the motif annotation file</dd>

        <dt>

          <code><a href="read_motif_ranking.html">read_motif_ranking()</a></code>

        </dt>
        <dd>Read in the motif rankings and transform them into a matrix</dd>

        <dt>

          <code><a href="remove_knn.html">remove_knn()</a></code>

        </dt>
        <dd>Remove the KNN data</dd>

        <dt>

          <code><a href="remove_samples.html">remove_samples()</a></code>

        </dt>
        <dd>Remove samples from object</dd>

        <dt>

          <code><a href="remove_snn_graph.html">remove_snn_graph()</a></code>

        </dt>
        <dd>Remove the sNN graph</dd>

        <dt>

          <code><a href="robust_scale.html">robust_scale()</a></code>

        </dt>
        <dd>Robust scaler.</dd>

        <dt>

          <code><a href="rs_2d_loess.html">rs_2d_loess()</a></code>

        </dt>
        <dd>Rust implementation of a Loess function</dd>

        <dt>

          <code><a href="rs_aucell.html">rs_aucell()</a></code>

        </dt>
        <dd>Calculate AUCell in Rust</dd>

        <dt>

          <code><a href="rs_bbknn.html">rs_bbknn()</a></code>

        </dt>
        <dd>BBKNN implementation in Rust</dd>

        <dt>

          <code><a href="rs_bbknn_filtering.html">rs_bbknn_filtering()</a></code>

        </dt>
        <dd>Reduce BBKNN matrix to Top X neighbours</dd>

        <dt>

          <code><a href="rs_calc_es.html">rs_calc_es()</a></code>

        </dt>
        <dd>Calculates the traditional GSEA enrichment score</dd>

        <dt>

          <code><a href="rs_calc_gsea_stat_cumulative_batch.html">rs_calc_gsea_stat_cumulative_batch()</a></code>

        </dt>
        <dd>Helper function to generate fgsea simple-based permutations</dd>

        <dt>

          <code><a href="rs_calc_gsea_stat_traditional_batch.html">rs_calc_gsea_stat_traditional_batch()</a></code>

        </dt>
        <dd>Helper function to generate traditional GSEA-based permutations</dd>

        <dt>

          <code><a href="rs_calc_gsea_stats.html">rs_calc_gsea_stats()</a></code>

        </dt>
        <dd>Rust implementation of the fgsea::calcGseaStat() function</dd>

        <dt>

          <code><a href="rs_calc_multi_level.html">rs_calc_multi_level()</a></code>

        </dt>
        <dd>Calculates p-values for pre-processed data</dd>

        <dt>

          <code><a href="rs_calculate_dge_mann_whitney.html">rs_calculate_dge_mann_whitney()</a></code>

        </dt>
        <dd>Calculate DGEs between cells based on Mann Whitney stats</dd>

        <dt>

          <code><a href="rs_cistarget.html">rs_cistarget()</a></code>

        </dt>
        <dd>Run CisTarget motif enrichment analysis</dd>

        <dt>

          <code><a href="rs_cluster_stability.html">rs_cluster_stability()</a></code>

        </dt>
        <dd>Helper function to assess cluster stability</dd>

        <dt>

          <code><a href="rs_constrained_page_rank.html">rs_constrained_page_rank()</a></code>

        </dt>
        <dd>Calculate a constrained page-rank score</dd>

        <dt>

          <code><a href="rs_constrained_page_rank_list.html">rs_constrained_page_rank_list()</a></code>

        </dt>
        <dd>Calculate a constrained page-rank score over a list.</dd>

        <dt>

          <code><a href="rs_contrastive_pca.html">rs_contrastive_pca()</a></code>

        </dt>
        <dd>Calculate the contrastive PCA</dd>

        <dt>

          <code><a href="rs_cor.html">rs_cor()</a></code>

        </dt>
        <dd>Calculate the column wise correlations.</dd>

        <dt>

          <code><a href="rs_cor2.html">rs_cor2()</a></code>

        </dt>
        <dd>Calculate the column wise correlations.</dd>

        <dt>

          <code><a href="rs_cor_upper_triangle.html">rs_cor_upper_triangle()</a></code>

        </dt>
        <dd>Calculate the column wise correlations.</dd>

        <dt>

          <code><a href="rs_coremo_quality.html">rs_coremo_quality()</a></code>

        </dt>
        <dd>Helper function to assess CoReMo cluster quality</dd>

        <dt>

          <code><a href="rs_coremo_stability.html">rs_coremo_stability()</a></code>

        </dt>
        <dd>Helper function to assess CoReMo cluster stability</dd>

        <dt>

          <code><a href="rs_cos.html">rs_cos()</a></code>

        </dt>
        <dd>Calculate the column wise cosine similarities</dd>

        <dt>

          <code><a href="rs_count_zeroes.html">rs_count_zeroes()</a></code>

        </dt>
        <dd>Helper to get zero stats from a given matrix</dd>

        <dt>

          <code><a href="rs_cov2cor.html">rs_cov2cor()</a></code>

        </dt>
        <dd>Calculates the correlation matrix from the co-variance matrix</dd>

        <dt>

          <code><a href="rs_covariance.html">rs_covariance()</a></code>

        </dt>
        <dd>Calculate the column-wise co-variance.</dd>

        <dt>

          <code><a href="rs_create_random_aucs.html">rs_create_random_aucs()</a></code>

        </dt>
        <dd>Create random AUCs</dd>

        <dt>

          <code><a href="rs_critval.html">rs_critval()</a></code>

        </dt>
        <dd>Calculate the critical value</dd>

        <dt>

          <code><a href="rs_critval_mat.html">rs_critval_mat()</a></code>

        </dt>
        <dd>Calculate the critical value</dd>

        <dt>

          <code><a href="rs_dense_to_upper_triangle.html">rs_dense_to_upper_triangle()</a></code>

        </dt>
        <dd>Generate a vector-based representation of the upper triangle of a matrix</dd>

        <dt>

          <code><a href="rs_differential_cor.html">rs_differential_cor()</a></code>

        </dt>
        <dd>Calculate the column wise differential correlation between two sets of data.</dd>

        <dt>

          <code><a href="rs_dist.html">rs_dist()</a></code>

        </dt>
        <dd>Calculate the pairwise column distance in a matrix</dd>

        <dt>

          <code><a href="rs_fast_auc.html">rs_fast_auc()</a></code>

        </dt>
        <dd>Fast AUC calculation</dd>

        <dt>

          <code><a href="rs_fast_ica.html">rs_fast_ica()</a></code>

        </dt>
        <dd>Run the Rust implementation of fast ICA.</dd>

        <dt>

          <code><a href="rs_fdr_adjustment.html">rs_fdr_adjustment()</a></code>

        </dt>
        <dd>Calculate a BH-based FDR</dd>

        <dt>

          <code><a href="rs_filter_onto_sim.html">rs_filter_onto_sim()</a></code>

        </dt>
        <dd>Filter the term similarities for a specific critical value</dd>

        <dt>

          <code><a href="rs_generate_bulk_rnaseq.html">rs_generate_bulk_rnaseq()</a></code>

        </dt>
        <dd>Generation of bulkRNAseq-like data with optional correlation structure</dd>

        <dt>

          <code><a href="rs_geom_elim_fgsea_simple.html">rs_geom_elim_fgsea_simple()</a></code>

        </dt>
        <dd>Run fgsea simple method for gene ontology with elimination method</dd>

        <dt>

          <code><a href="rs_get_gs_indices.html">rs_get_gs_indices()</a></code>

        </dt>
        <dd>Helper function to rapidly retrieve the indices of the gene set members</dd>

        <dt>

          <code><a href="rs_get_metacells.html">rs_get_metacells()</a></code>

        </dt>
        <dd>Generate meta cells (hdWGCNA method)</dd>

        <dt>

          <code><a href="rs_get_seacells.html">rs_get_seacells()</a></code>

        </dt>
        <dd>Generate SEACells</dd>

        <dt>

          <code><a href="rs_gower_dist.html">rs_gower_dist()</a></code>

        </dt>
        <dd>Calculates the Gower distance for a given matrix</dd>

        <dt>

          <code><a href="rs_gse_geom_elim.html">rs_gse_geom_elim()</a></code>

        </dt>
        <dd>Run hypergeometric enrichment over the gene ontology</dd>

        <dt>

          <code><a href="rs_gse_geom_elim_list.html">rs_gse_geom_elim_list()</a></code>

        </dt>
        <dd>Run hypergeometric enrichment a list of target genes over the gene ontology</dd>

        <dt>

          <code><a href="rs_gsva.html">rs_gsva()</a></code>

        </dt>
        <dd>Rust version of the GSVA algorithm</dd>

        <dt>

          <code><a href="rs_h5ad_data.html">rs_h5ad_data()</a></code>

        </dt>
        <dd>Load in h5ad data via Rust</dd>

        <dt>

          <code><a href="rs_hamming_dist.html">rs_hamming_dist()</a></code>

        </dt>
        <dd>Calculates the Hamming distance between categorical columns</dd>

        <dt>

          <code><a href="rs_hedges_g.html">rs_hedges_g()</a></code>

        </dt>
        <dd>Calculate the Hedge's G effect</dd>

        <dt>

          <code><a href="rs_hotspot_autocor.html">rs_hotspot_autocor()</a></code>

        </dt>
        <dd>Calculate gene spatial auto-correlations</dd>

        <dt>

          <code><a href="rs_hotspot_cluster_genes.html">rs_hotspot_cluster_genes()</a></code>

        </dt>
        <dd>Cluster the genes by Z-score together</dd>

        <dt>

          <code><a href="rs_hotspot_gene_cor.html">rs_hotspot_gene_cor()</a></code>

        </dt>
        <dd>Calculate gene&lt;&gt;gene spatial correlations</dd>

        <dt>

          <code><a href="rs_hypergeom_test.html">rs_hypergeom_test()</a></code>

        </dt>
        <dd>Run a single hypergeometric test.</dd>

        <dt>

          <code><a href="rs_hypergeom_test_list.html">rs_hypergeom_test_list()</a></code>

        </dt>
        <dd>Run a hypergeometric test over a list of target genes</dd>

        <dt>

          <code><a href="rs_ica_iters.html">rs_ica_iters()</a></code>

        </dt>
        <dd>Run ICA over a given no_comp with random initilisations of w_init</dd>

        <dt>

          <code><a href="rs_ica_iters_cv.html">rs_ica_iters_cv()</a></code>

        </dt>
        <dd>Run ICA with cross-validation and random initialsiation</dd>

        <dt>

          <code><a href="rs_kbet.html">rs_kbet()</a></code>

        </dt>
        <dd>Calculate kBET type scores</dd>

        <dt>

          <code><a href="rs_knn_label_propagation.html">rs_knn_label_propagation()</a></code>

        </dt>
        <dd>kNN label propagation</dd>

        <dt>

          <code><a href="rs_knn_mat_to_edge_list.html">rs_knn_mat_to_edge_list()</a></code>

        </dt>
        <dd>Flatten kNN matrix to edge list</dd>

        <dt>

          <code><a href="rs_mad_outlier.html">rs_mad_outlier()</a></code>

        </dt>
        <dd>Calculate MAD outlier detection in Rust.</dd>

        <dt>

          <code><a href="rs_make_milor_nhoods.html">rs_make_milor_nhoods()</a></code>

        </dt>
        <dd>Generate the neighbourhoods akin to the miloR approach</dd>

        <dt>

          <code><a href="rs_mitch_calc.html">rs_mitch_calc()</a></code>

        </dt>
        <dd>Calculate mitch enrichment leveraging Rust under the hood</dd>

        <dt>

          <code><a href="rs_mnn.html">rs_mnn()</a></code>

        </dt>
        <dd>FastMNN batch correction in Rust</dd>

        <dt>

          <code><a href="rs_module_scoring.html">rs_module_scoring()</a></code>

        </dt>
        <dd>Calculate module activity scores in Rust</dd>

        <dt>

          <code><a href="rs_mutual_info.html">rs_mutual_info()</a></code>

        </dt>
        <dd>Calculates the mutual information matrix</dd>

        <dt>

          <code><a href="rs_onto_semantic_sim.html">rs_onto_semantic_sim()</a></code>

        </dt>
        <dd>Calculate the semantic similarity in an ontology</dd>

        <dt>

          <code><a href="rs_onto_semantic_sim_mat.html">rs_onto_semantic_sim_mat()</a></code>

        </dt>
        <dd>Calculate the semantic similarity in an ontology</dd>

        <dt>

          <code><a href="rs_onto_sim_wang.html">rs_onto_sim_wang()</a></code>

        </dt>
        <dd>Calculate the Wang similarity for specific terms</dd>

        <dt>

          <code><a href="rs_onto_sim_wang_mat.html">rs_onto_sim_wang_mat()</a></code>

        </dt>
        <dd>Calculate the Wang similarity matrix for an ontology</dd>

        <dt>

          <code><a href="rs_ot_harmonic_sum.html">rs_ot_harmonic_sum()</a></code>

        </dt>
        <dd>Calculate the OT harmonic sum</dd>

        <dt>

          <code><a href="rs_page_rank.html">rs_page_rank()</a></code>

        </dt>
        <dd>Rust version of calcaluting the personalised page rank</dd>

        <dt>

          <code><a href="rs_page_rank_parallel.html">rs_page_rank_parallel()</a></code>

        </dt>
        <dd>Calculate massively parallelised personalised page rank scores</dd>

        <dt>

          <code><a href="rs_phyper.html">rs_phyper()</a></code>

        </dt>
        <dd>Calculate the hypergeometric rest in Rust</dd>

        <dt>

          <code><a href="rs_pointwise_mutual_info.html">rs_pointwise_mutual_info()</a></code>

        </dt>
        <dd>Calculates the point wise mutual information</dd>

        <dt>

          <code><a href="rs_prcomp.html">rs_prcomp()</a></code>

        </dt>
        <dd>Rust implementation of prcomp</dd>

        <dt>

          <code><a href="rs_prepare_gsva_gs.html">rs_prepare_gsva_gs()</a></code>

        </dt>
        <dd>Prepare a pathway list for GSVA</dd>

        <dt>

          <code><a href="rs_prepare_whitening.html">rs_prepare_whitening()</a></code>

        </dt>
        <dd>Prepare the data for whitening</dd>

        <dt>

          <code><a href="rs_random_svd.html">rs_random_svd()</a></code>

        </dt>
        <dd>Run randomised SVD over a matrix</dd>

        <dt>

          <code><a href="rs_range_norm.html">rs_range_norm()</a></code>

        </dt>
        <dd>Apply a range normalisation on a vector.</dd>

        <dt>

          <code><a href="rs_rbf_function.html">rs_rbf_function()</a></code>

        </dt>
        <dd>Apply a Radial Basis Function</dd>

        <dt>

          <code><a href="rs_rbf_function_mat.html">rs_rbf_function_mat()</a></code>

        </dt>
        <dd>Apply a Radial Basis Function (to a matrix)</dd>

        <dt>

          <code><a href="rs_rbf_iterate_epsilons.html">rs_rbf_iterate_epsilons()</a></code>

        </dt>
        <dd>Helper to identify the right epsilon parameter</dd>

        <dt>

          <code><a href="rs_rbh_cor.html">rs_rbh_cor()</a></code>

        </dt>
        <dd>Generate reciprocal best hits based on correlations</dd>

        <dt>

          <code><a href="rs_rbh_sets.html">rs_rbh_sets()</a></code>

        </dt>
        <dd>Generate reciprocal best hits based on set similarities</dd>

        <dt>

          <code><a href="rs_sample_ids_for_cell_types.html">rs_sample_ids_for_cell_types()</a></code>

        </dt>
        <dd>Helper function to generate sample identifiers based on cells</dd>

        <dt>

          <code><a href="rs_sc_doublet_detection.html">rs_sc_doublet_detection()</a></code>

        </dt>
        <dd>Detect Doublets via BoostClassifier (in Rust)</dd>

        <dt>

          <code><a href="rs_sc_get_gene_set_perc.html">rs_sc_get_gene_set_perc()</a></code>

        </dt>
        <dd>Calculate the percentage of gene sets in the cells</dd>

        <dt>

          <code><a href="rs_sc_get_top_genes_perc.html">rs_sc_get_top_genes_perc()</a></code>

        </dt>
        <dd>Calculates the cumulative proportion of the top X genes</dd>

        <dt>

          <code><a href="rs_sc_hvg.html">rs_sc_hvg()</a></code>

        </dt>
        <dd>Calculate the percentage of gene sets in the cells</dd>

        <dt>

          <code><a href="rs_sc_hvg_batch_aware.html">rs_sc_hvg_batch_aware()</a></code>

        </dt>
        <dd>Calculate HVG per batch</dd>

        <dt>

          <code><a href="rs_sc_knn.html">rs_sc_knn()</a></code>

        </dt>
        <dd>Generates the kNN graph</dd>

        <dt>

          <code><a href="rs_sc_knn_w_dist.html">rs_sc_knn_w_dist()</a></code>

        </dt>
        <dd>Generates the kNN graph with additional distances</dd>

        <dt>

          <code><a href="rs_sc_pca.html">rs_sc_pca()</a></code>

        </dt>
        <dd>Calculates PCA for single cell</dd>

        <dt>

          <code><a href="rs_sc_scrublet.html">rs_sc_scrublet()</a></code>

        </dt>
        <dd>Scrublet Rust interface</dd>

        <dt>

          <code><a href="rs_sc_snn.html">rs_sc_snn()</a></code>

        </dt>
        <dd>Generates the sNN graph for igraph</dd>

        <dt>

          <code><a href="rs_set_similarity.html">rs_set_similarity()</a></code>

        </dt>
        <dd>Set similarities</dd>

        <dt>

          <code><a href="rs_set_similarity_list.html">rs_set_similarity_list()</a></code>

        </dt>
        <dd>Set similarities over one list</dd>

        <dt>

          <code><a href="rs_set_similarity_list2.html">rs_set_similarity_list2()</a></code>

        </dt>
        <dd>Set similarities over two list</dd>

        <dt>

          <code><a href="rs_simple_and_multi_err.html">rs_simple_and_multi_err()</a></code>

        </dt>
        <dd>Calculates the simple and multi error for fgsea multi level</dd>

        <dt>

          <code><a href="rs_simulate_dropouts.html">rs_simulate_dropouts()</a></code>

        </dt>
        <dd>Sparsify bulkRNAseq like data</dd>

        <dt>

          <code><a href="rs_snf.html">rs_snf()</a></code>

        </dt>
        <dd>Similarity network fusion</dd>

        <dt>

          <code><a href="rs_snf_affinity_cat.html">rs_snf_affinity_cat()</a></code>

        </dt>
        <dd>Calculate the SNF affinity matrix for categorical values</dd>

        <dt>

          <code><a href="rs_snf_affinity_continuous.html">rs_snf_affinity_continuous()</a></code>

        </dt>
        <dd>Calculate the SNF affinity matrix for continuous values</dd>

        <dt>

          <code><a href="rs_snf_affinity_mixed.html">rs_snf_affinity_mixed()</a></code>

        </dt>
        <dd>Calculate the SNF affinity matrix for mixed values</dd>

        <dt>

          <code><a href="rs_sparse_dict_dgrdl.html">rs_sparse_dict_dgrdl()</a></code>

        </dt>
        <dd>Generate a sparse dictionary with DGRDL</dd>

        <dt>

          <code><a href="rs_sparse_dict_dgrdl_grid_search.html">rs_sparse_dict_dgrdl_grid_search()</a></code>

        </dt>
        <dd>Generate a sparse dictionary with DGRDL</dd>

        <dt>

          <code><a href="rs_spectral_clustering.html">rs_spectral_clustering()</a></code>

        </dt>
        <dd>Rust implementation of spectral clustering</dd>

        <dt>

          <code><a href="rs_spectral_clustering_sim.html">rs_spectral_clustering_sim()</a></code>

        </dt>
        <dd>Rust implementation of spectral clustering</dd>

        <dt>

          <code><a href="rs_split_cor_signs.html">rs_split_cor_signs()</a></code>

        </dt>
        <dd>Helper function to split correlation matrices by sign</dd>

        <dt>

          <code><a href="rs_ssgsea.html">rs_ssgsea()</a></code>

        </dt>
        <dd>Rust version of the ssGSEA algorithm</dd>

        <dt>

          <code><a href="rs_supercell.html">rs_supercell()</a></code>

        </dt>
        <dd>Generate SuperCells.</dd>

        <dt>

          <code><a href="rs_synthetic_sc_data_csc.html">rs_synthetic_sc_data_csc()</a></code>

        </dt>
        <dd>Generate synthetic single cell data (Seurat type)</dd>

        <dt>

          <code><a href="rs_synthetic_sc_data_csr.html">rs_synthetic_sc_data_csr()</a></code>

        </dt>
        <dd>Generate synthetic single cell data (h5ad type)</dd>

        <dt>

          <code><a href="rs_synthetic_sc_data_with_cell_types.html">rs_synthetic_sc_data_with_cell_types()</a></code>

        </dt>
        <dd>Generates synthetic data for single cell</dd>

        <dt>

          <code><a href="rs_tied_diffusion_parallel.html">rs_tied_diffusion_parallel()</a></code>

        </dt>
        <dd>Calculate massively parallelised tied diffusion scores</dd>

        <dt>

          <code><a href="rs_tom.html">rs_tom()</a></code>

        </dt>
        <dd>Calculates the TOM over an affinity matrix</dd>

        <dt>

          <code><a href="rs_upper_triangle_to_dense.html">rs_upper_triangle_to_dense()</a></code>

        </dt>
        <dd>Reconstruct a matrix from a flattened upper triangle vector</dd>

        <dt>

          <code><a href="rs_upper_triangle_to_sparse.html">rs_upper_triangle_to_sparse()</a></code>

        </dt>
        <dd>Generate sparse data from an upper triangle</dd>

        <dt>

          <code><a href="rs_vision.html">rs_vision()</a></code>

        </dt>
        <dd>Calculate VISION pathway scores in Rust</dd>

        <dt>

          <code><a href="rs_vision_with_autocorrelation.html">rs_vision_with_autocorrelation()</a></code>

        </dt>
        <dd>Calculate VISION pathway scores in Rust with auto-correlation</dd>

        <dt>

          <code><a href="run_cistarget.html">run_cistarget()</a></code>

        </dt>
        <dd>Main function to run CisTarget</dd>

        <dt>

          <code><a href="run_limma_voom.html">run_limma_voom()</a></code>

        </dt>
        <dd>Wrapper for a Limma Voom analysis</dd>

        <dt>

          <code><a href="run_snf.html">run_snf()</a></code>

        </dt>
        <dd>Run the SNF algorithm</dd>

        <dt>

          <code><a href="save_sc_exp_to_disk.html">save_sc_exp_to_disk()</a></code>

        </dt>
        <dd>Save memory-bound data to disk</dd>

        <dt>

          <code><a href="scale_free_fit.html">scale_free_fit()</a></code>

        </dt>
        <dd>Calculate the goodness of fit for a power law distribution.</dd>

        <dt>

          <code><a href="scrublet_sc.html">scrublet_sc()</a></code>

        </dt>
        <dd>Doublet detection with Scrublet</dd>

        <dt>

          <code><a href="select_user_option.html">select_user_option()</a></code>

        </dt>
        <dd>Helper function for user option selection</dd>

        <dt>

          <code><a href="set_cell_mapping.html">set_cell_mapping()</a></code>

        </dt>
        <dd>Set cell mapping for sc_mapper object</dd>

        <dt>

          <code><a href="set_cells_to_keep.html">set_cells_to_keep()</a></code>

        </dt>
        <dd>Set cells to keep for sc_mapper object</dd>

        <dt>

          <code><a href="set_embedding.html">set_embedding()</a></code>

        </dt>
        <dd>Add additional embeddings to the class</dd>

        <dt>

          <code><a href="set_gene_mapping.html">set_gene_mapping()</a></code>

        </dt>
        <dd>Set gene mapping for sc_mapper object</dd>

        <dt>

          <code><a href="set_hotspot_membership.html">set_hotspot_membership()</a></code>

        </dt>
        <dd>Identify hotspot gene clusters</dd>

        <dt>

          <code><a href="set_hvg.html">set_hvg()</a></code>

        </dt>
        <dd>Set the HVG genes</dd>

        <dt>

          <code><a href="set_knn.html">set_knn()</a></code>

        </dt>
        <dd>Set/add KNN</dd>

        <dt>

          <code><a href="set_pca_factors.html">set_pca_factors()</a></code>

        </dt>
        <dd>Set/add PCA factors</dd>

        <dt>

          <code><a href="set_pca_loadings.html">set_pca_loadings()</a></code>

        </dt>
        <dd>Set/add PCA loadings</dd>

        <dt>

          <code><a href="set_pca_singular_vals.html">set_pca_singular_vals()</a></code>

        </dt>
        <dd>Set/add PCA singular values</dd>

        <dt>

          <code><a href="set_sc_new_obs_col.html">set_sc_new_obs_col()</a></code>

        </dt>
        <dd>Add a new column to the obs table</dd>

        <dt>

          <code><a href="set_sc_new_obs_col_multiple.html">set_sc_new_obs_col_multiple()</a></code>

        </dt>
        <dd>Add multiple new columns to the obs table</dd>

        <dt>

          <code><a href="set_sc_new_var_cols.html">set_sc_new_var_cols()</a></code>

        </dt>
        <dd>Add a new column to the var table</dd>

        <dt>

          <code><a href="set_snn_graph.html">set_snn_graph()</a></code>

        </dt>
        <dd>Set/add KNN</dd>

        <dt>

          <code><a href="simplify_hypergeom_res.html">simplify_hypergeom_res()</a></code>

        </dt>
        <dd>Simplify gene set results via ontologies</dd>

        <dt>

          <code><a href="simulate_dropouts.html">simulate_dropouts()</a></code>

        </dt>
        <dd>Simulate dropouts via different functions on synthetic bulk data</dd>

        <dt>

          <code><a href="single_cell_duckdb_base.html">single_cell_duckdb_base</a></code>

        </dt>
        <dd>Base class for the single cell DuckDB connection</dd>

        <dt>

          <code><a href="single_cell_duckdb_con.html">single_cell_duckdb_con</a></code>

        </dt>
        <dd>Class for storing single cell experimental data in DuckDB (nightly!)</dd>

        <dt>

          <code><a href="single_cell_exp.html">single_cell_exp()</a></code>

        </dt>
        <dd>bixverse single cell class</dd>

        <dt>

          <code><a href="snf.html">snf()</a></code>

        </dt>
        <dd>Similarity network fusion</dd>

        <dt>

          <code><a href="snf_process_aff_cat_mixed.html">snf_process_aff_cat_mixed()</a></code>

        </dt>
        <dd>Helper function to process categorical or mixed data for SNF</dd>

        <dt>

          <code><a href="snf_process_aff_continuous.html">snf_process_aff_continuous()</a></code>

        </dt>
        <dd>Helper function to process continuous data for SNF</dd>

        <dt>

          <code><a href="sparse_list_to_mat.html">sparse_list_to_mat()</a></code>

        </dt>
        <dd>Helper function to transform the Rust-exported sparse matrices into R ones</dd>

        <dt>

          <code><a href="sparse_mat_to_list.html">sparse_mat_to_list()</a></code>

        </dt>
        <dd>Helper function to transform R sparse matrices to list</dd>

        <dt>

          <code><a href="stream_h5ad.html">stream_h5ad()</a></code>

        </dt>
        <dd>Stream in h5ad to <code>single_cell_exp</code></dd>

        <dt>

          <code><a href="summarise_scores.html">summarise_scores()</a></code>

        </dt>
        <dd>Summarise gene scores if they are duplicates.</dd>

        <dt>

          <code><a href="synthetic_bulk_cor_matrix.html">synthetic_bulk_cor_matrix()</a></code>

        </dt>
        <dd>Generates synthetic bulk RNAseq data</dd>

        <dt>

          <code><a href="synthetic_c_pca_data.html">synthetic_c_pca_data()</a></code>

        </dt>
        <dd>Generates synthetic data for contrastive PCA exploration.</dd>

        <dt>

          <code><a href="synthetic_signal_matrix.html">synthetic_signal_matrix()</a></code>

        </dt>
        <dd>Generates a simple synthetic, pseudo gene expression matrix</dd>

        <dt>

          <code><a href="test_nhoods.html">test_nhoods()</a></code>

        </dt>
        <dd>Test neighbourhoods for differential abundance</dd>

        <dt>

          <code><a href="tied_diffusion.html">tied_diffusion()</a></code>

        </dt>
        <dd>Diffuse seed genes in a tied manner over a network</dd>

        <dt>

          <code><a href="to_snake_case.html">to_snake_case()</a></code>

        </dt>
        <dd>Helper function to transform strings to snake_case</dd>

        <dt>

          <code><a href="top_genes_perc_sc.html">top_genes_perc_sc()</a></code>

        </dt>
        <dd>Calculate the proportions of reads for the Top N genes</dd>

        <dt>

          <code><a href="tree_cut_iter.html">tree_cut_iter()</a></code>

        </dt>
        <dd>Coremo: Iterate over k for gene module detection.</dd>

        <dt>

          <code><a href="update_metadata_values.html">update_metadata_values()</a></code>

        </dt>
        <dd>Replace values in a metadata column</dd>

        <dt>

          <code><a href="upper_triangle_diffcor_mat.html">upper_triangle_diffcor_mat</a></code>

        </dt>
        <dd>Class for symmetric differential correlation matrices</dd>

        <dt>

          <code><a href="upper_triangle_to_sparse.html">upper_triangle_to_sparse()</a></code>

        </dt>
        <dd>Transform an upper triangle-stored matrix to a sparse one</dd>

        <dt>

          <code><a href="upper_triangular_sym_mat.html">upper_triangular_sym_mat</a></code>

        </dt>
        <dd>Class for symmetric correlation matrices</dd>

        <dt>

          <code><a href="vision_sc.html">vision_sc()</a></code>

        </dt>
        <dd>Calculate VISION scores</dd>

        <dt>

          <code><a href="vision_w_autocor_sc.html">vision_w_autocor_sc()</a></code>

        </dt>
        <dd>Calculate VISION scores (with auto-correlation scores)</dd>

        <dt>

          <code><a href="write_cellranger_output.html">write_cellranger_output()</a></code>

        </dt>
        <dd>Helper function to write data to a cell ranger like output</dd>

        <dt>

          <code><a href="write_h5ad_sc.html">write_h5ad_sc()</a></code>

        </dt>
        <dd>Helper function to write data to h5ad format</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gregor Lueg.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

