<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Gene Set Enrichment Methods • bixverse</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="gse_methods_files/libs/quarto-html/popper.min.js"></script><script src="gse_methods_files/libs/quarto-html/tippy.umd.min.js"></script><link href="gse_methods_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="gse_methods_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Gene Set Enrichment Methods">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">bixverse</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Gene Set Enrichment Methods</h1>





      <div class="d-none name"><code></code></div>
    </div>






    <section class="level2"><h2 id="gene-set-enrichment-methods-in-bixverse">Gene Set Enrichment Methods in bixverse<a class="anchor" aria-label="anchor" href="#gene-set-enrichment-methods-in-bixverse"></a>
</h2>
<p>This vignette shows you how to use the different implement gene set enrichment methods in <code>bixverse</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"fgsea"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"fgsea"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"msigdbr"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"msigdbr"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bixverse</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<section class="level3"><h3 id="hypergeometric-tests">Hypergeometric tests<a class="anchor" aria-label="anchor" href="#hypergeometric-tests"></a>
</h3>
<p><code>bixverse</code> has implementations of the standard hypergeometric tests to identify gene set enrichment for a given set of target genes. In this example, we will be loading the Hallmark gene sets from <code>msigdbr</code>. Let’s first load these in and create a list of gene sets (the expected format of most functions in <code>bixverse</code> for these type of analysis.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h_gene_sets</span> <span class="op">&lt;-</span> <span class="fu">msigdbr</span><span class="fu">::</span><span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr.html" class="external-link">msigdbr</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"human"</span>, collection <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">h_gene_sets_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">h_gene_sets</span><span class="op">$</span><span class="va">ensembl_gene</span>, <span class="va">h_gene_sets</span><span class="op">$</span><span class="va">gs_name</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123L</span><span class="op">)</span></span>
<span></span>
<span><span class="va">target_genes_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">h_gene_sets_ls</span><span class="op">[[</span><span class="st">"HALLMARK_MYC_TARGETS_V1"</span><span class="op">]</span><span class="op">]</span>, <span class="fl">25</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">h_gene_sets_ls</span><span class="op">[[</span><span class="st">"HALLMARK_MYC_TARGETS_V2"</span><span class="op">]</span><span class="op">]</span>, <span class="fl">25</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<p>To run the hypergeometric test against one target set, you can just use <code>gse_hypergeometric</code>. The function has parameters to define a minimum FDR threshold and minimum overlap science. Additionally, the gene universe will be set (if not provided) to all of the genes represented in the <code>gene_set_list</code>. You can provide your own gene universe if you wish to.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gse_hypergeometric.html">gse_hypergeometric</a></span><span class="op">(</span></span>
<span>  target_genes <span class="op">=</span> <span class="va">target_genes_1</span>,</span>
<span>  gene_set_list <span class="op">=</span> <span class="va">h_gene_sets_ls</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span>
<span><span class="co">#&gt;              gene_set_name odds_ratios        pvals          fdr  hits</span></span>
<span><span class="co">#&gt;                     &lt;char&gt;       &lt;num&gt;        &lt;num&gt;        &lt;num&gt; &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: HALLMARK_MYC_TARGETS_V2  170.749267 5.847007e-41 2.923504e-39    27</span></span>
<span><span class="co">#&gt; 2: HALLMARK_MYC_TARGETS_V1   35.385088 2.083184e-27 5.207959e-26    29</span></span>
<span><span class="co">#&gt; 3:    HALLMARK_E2F_TARGETS    4.219512 1.444580e-03 2.407633e-02     8</span></span>
<span><span class="co">#&gt;    gene_set_lengths target_set_lengths</span></span>
<span><span class="co">#&gt;               &lt;num&gt;              &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:               58                 50</span></span>
<span><span class="co">#&gt; 2:              200                 50</span></span>
<span><span class="co">#&gt; 3:              200                 50</span></span></code></pre></div>
</div>
<p>To run the hypergeometric test over a list of target gene sets, <code>bixverse</code> provides <code>gse_hypergeometric_list</code> that leverages <a href="https://github.com/rayon-rs/rayon" class="external-link">rayon</a> in Rust for multi-threading. This makes this function very fast even on large gene set libraries with lots of target gene sets.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">target_genes_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">h_gene_sets_ls</span><span class="op">[[</span><span class="st">"HALLMARK_TNFA_SIGNALING_VIA_NFKB"</span><span class="op">]</span><span class="op">]</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">h_gene_sets_ls</span><span class="op">[[</span><span class="st">"HALLMARK_IL6_JAK_STAT3_SIGNALING"</span><span class="op">]</span><span class="op">]</span>, <span class="fl">25</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">target_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  set_1 <span class="op">=</span> <span class="va">target_genes_1</span>,</span>
<span>  set_2 <span class="op">=</span> <span class="va">target_genes_2</span>,</span>
<span>  set_3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"random gene 1"</span>, <span class="st">"random gene 2"</span>, <span class="st">"random gene 3"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">results_multiple</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gse_hypergeometric_list.html">gse_hypergeometric_list</a></span><span class="op">(</span></span>
<span>  target_genes_list <span class="op">=</span> <span class="va">target_list</span>,</span>
<span>  gene_set_list <span class="op">=</span> <span class="va">h_gene_sets_ls</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results_multiple</span><span class="op">)</span></span>
<span><span class="co">#&gt;    target_set_name odds_ratios        pvals          fdr  hits gene_set_lengths</span></span>
<span><span class="co">#&gt;             &lt;char&gt;       &lt;num&gt;        &lt;num&gt;        &lt;num&gt; &lt;num&gt;            &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:           set_2  162.370744 4.603263e-43 2.301631e-41    30               91</span></span>
<span><span class="co">#&gt; 2:           set_1  170.868035 5.740508e-41 2.870254e-39    27               58</span></span>
<span><span class="co">#&gt; 3:           set_1   35.410526 2.043327e-27 5.108317e-26    29              200</span></span>
<span><span class="co">#&gt; 4:           set_2   29.978469 5.916124e-22 1.479031e-20    24              200</span></span>
<span><span class="co">#&gt; 5:           set_2   12.001728 2.323959e-10 3.873265e-09    15              201</span></span>
<span><span class="co">#&gt; 6:           set_2    8.575841 2.503092e-07 3.128865e-06    12              200</span></span>
<span><span class="co">#&gt;                         gene_set_name target_set_lengths</span></span>
<span><span class="co">#&gt;                                &lt;char&gt;              &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:   HALLMARK_IL6_JAK_STAT3_SIGNALING                 45</span></span>
<span><span class="co">#&gt; 2:            HALLMARK_MYC_TARGETS_V2                 50</span></span>
<span><span class="co">#&gt; 3:            HALLMARK_MYC_TARGETS_V1                 50</span></span>
<span><span class="co">#&gt; 4:   HALLMARK_TNFA_SIGNALING_VIA_NFKB                 45</span></span>
<span><span class="co">#&gt; 5:     HALLMARK_INFLAMMATORY_RESPONSE                 45</span></span>
<span><span class="co">#&gt; 6: HALLMARK_INTERFERON_GAMMA_RESPONSE                 45</span></span></code></pre></div>
</div>
<p>To exemplify the speed in Rust, in this case we run for 250 target gene sets over-enrichment tests against 5000 gene sets sampled from 20k genes. This totals 1.25m hypergeometric tests. Pending on your system, you can expect this to be done in a few seconds.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seed</span> <span class="op">=</span> <span class="fl">10101L</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">seed</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define the parameters</span></span>
<span><span class="va">universe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"gene_%i"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20000</span><span class="op">)</span></span>
<span><span class="va">gene_sets_no</span> <span class="op">&lt;-</span> <span class="fl">5000</span></span>
<span><span class="va">target_gene_sets_no</span> <span class="op">&lt;-</span> <span class="fl">250</span></span>
<span></span>
<span><span class="co"># Generate random gene sets</span></span>
<span><span class="va">gene_sets</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span></span>
<span>  <span class="fl">1</span><span class="op">:</span><span class="va">gene_sets_no</span>,</span>
<span>  <span class="op">~</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">seed</span> <span class="op">+</span> <span class="va">.x</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="va">size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">20</span><span class="op">:</span><span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">universe</span>, <span class="va">size</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_chr</a></span><span class="op">(</span></span>
<span>  <span class="fl">1</span><span class="op">:</span><span class="va">gene_sets_no</span>,</span>
<span>  <span class="op">~</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">seed</span> <span class="op">+</span> <span class="va">.x</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">3</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate random target sets</span></span>
<span><span class="va">target_gene_sets</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span></span>
<span>  <span class="fl">1</span><span class="op">:</span><span class="va">target_gene_sets_no</span>,</span>
<span>  <span class="op">~</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">.x</span> <span class="op">*</span> <span class="va">seed</span><span class="op">)</span></span>
<span>    <span class="va">size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">50</span><span class="op">:</span><span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">universe</span>, <span class="va">size</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">target_gene_sets</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_chr</a></span><span class="op">(</span></span>
<span>  <span class="fl">1</span><span class="op">:</span><span class="va">target_gene_sets_no</span>,</span>
<span>  <span class="op">~</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">seed</span> <span class="op">+</span> <span class="va">.x</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">letters</span>, <span class="fl">3</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run the Rust function</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">rs_results_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gse_hypergeometric_list.html">gse_hypergeometric_list</a></span><span class="op">(</span></span>
<span>  target_genes_list <span class="op">=</span> <span class="va">target_gene_sets</span>,</span>
<span>  gene_set_list <span class="op">=</span> <span class="va">gene_sets</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; 0.557 sec elapsed</span></span></code></pre></div>
</div>
</section><section class="level3"><h3 id="gene-ontology-aware-enrichment-tests-for-sets">Gene ontology aware enrichment tests (for sets)<a class="anchor" aria-label="anchor" href="#gene-ontology-aware-enrichment-tests-for-sets"></a>
</h3>
<p>Additionally, <code>bixverse</code> provides Rust-accelerated functions to run an ontology- aware enrichment test for the gene ontology in particular - the concept is based on <a href="https://academic.oup.com/bioinformatics/article/22/13/1600/193669" class="external-link">Adrian et al.</a>. Briefly, the function will start in the leafs of the ontology (lowest level) and calculate the enrichment statistic for that level. If an enrichment test reaches a threshold, the genes from that term are removed from all of its ancestor terms, ensuring that the more specific terms have stronger test statistics. To make it easy for the users, we provide the human gene ontology as data in the package. This is a data.table with 5 columns:</p>
<ul>
<li>go_id: the gene ontology term identifier.</li>
<li>go_name: the name gene ontology term name.</li>
<li>ensembl_id: the ensembl identifiers belonging to this term.</li>
<li>ancestors: in the ontology DAG all identified ancestors of this specific term</li>
<li>depth: the depth of that term (identified via BFS) for this term. Higher values indicate further distance from the top terms.</li>
</ul>
<p>These columns are expected by the <code>gene_ontology_data</code> class that stores the data and exposes it in Rust.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load in the package provided data and process it into the right format</span></span>
<span><span class="va">go_data_dt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_go_data_human.html">get_go_data_human</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading the data from the package.</span></span>
<span><span class="co">#&gt; Processing data for the gene_ontology class.</span></span>
<span><span class="co">#&gt; Warning: The `father` argument of `dfs()` is deprecated as of igraph 2.2.0.</span></span>
<span><span class="co">#&gt; ℹ Please use the `parent` argument instead.</span></span>
<span><span class="co">#&gt; ℹ The deprecated feature was likely used in the bixverse package.</span></span>
<span><span class="co">#&gt;   Please report the issue to the authors.</span></span>
<span></span>
<span><span class="co"># Generate the S7 object for subsequent usage</span></span>
<span><span class="va">go_data_s7</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_ontology_data.html">gene_ontology_data</a></span><span class="op">(</span><span class="va">go_data_dt</span>, min_genes <span class="op">=</span> <span class="fl">3L</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Similar to the more generic hypergeometric tests, you can provide either a single set of a target genes or a list of target genes. Leveraging Rusts borrowing, the overhead for multiple target gene sets is reduced and hence very fast.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">go_aware_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gse_go_elim_method.html">gse_go_elim_method</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">go_data_s7</span>,</span>
<span>  target_genes <span class="op">=</span> <span class="va">target_genes_1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">go_aware_res</span><span class="op">)</span></span>
<span><span class="co">#&gt;         go_id                                     go_name odds_ratios</span></span>
<span><span class="co">#&gt;        &lt;char&gt;                                      &lt;char&gt;       &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: GO:0003723                                 RNA binding   14.812443</span></span>
<span><span class="co">#&gt; 2: GO:0032040                    small-subunit processome   44.202960</span></span>
<span><span class="co">#&gt; 3: GO:0042274          ribosomal small subunit biogenesis   43.541132</span></span>
<span><span class="co">#&gt; 4: GO:0005730                                   nucleolus    6.009426</span></span>
<span><span class="co">#&gt; 5: GO:0000055 ribosomal large subunit export from nucleus  455.891304</span></span>
<span><span class="co">#&gt; 6: GO:0031428    box C/D methylation guide snoRNP complex  455.891304</span></span>
<span><span class="co">#&gt;           pvals          fdr  hits gene_set_lengths</span></span>
<span><span class="co">#&gt;           &lt;num&gt;        &lt;num&gt; &lt;num&gt;            &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: 3.122680e-16 3.441817e-12    23             1205</span></span>
<span><span class="co">#&gt; 2: 1.623943e-08 6.489265e-05     6               72</span></span>
<span><span class="co">#&gt; 3: 1.766267e-08 6.489265e-05     6               73</span></span>
<span><span class="co">#&gt; 4: 8.523856e-08 2.348748e-04    18             1866</span></span>
<span><span class="co">#&gt; 5: 2.368116e-07 4.350230e-04     3                6</span></span>
<span><span class="co">#&gt; 6: 2.368116e-07 4.350230e-04     3                6</span></span></code></pre></div>
</div>
<p>Here is the example on how to run this over a list.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">go_aware_res_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gse_go_elim_method_list.html">gse_go_elim_method_list</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">go_data_s7</span>,</span>
<span>  target_gene_list <span class="op">=</span> <span class="va">target_list</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">go_aware_res_2</span><span class="op">)</span></span>
<span><span class="co">#&gt;    target_set_name                                              go_name</span></span>
<span><span class="co">#&gt;             &lt;char&gt;                                               &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:           set_1                                          RNA binding</span></span>
<span><span class="co">#&gt; 2:           set_2 cell surface receptor signaling pathway via JAK-STAT</span></span>
<span><span class="co">#&gt; 3:           set_2                                inflammatory response</span></span>
<span><span class="co">#&gt; 4:           set_2                                    cytokine activity</span></span>
<span><span class="co">#&gt; 5:           set_2                     external side of plasma membrane</span></span>
<span><span class="co">#&gt; 6:           set_2                           cellular response to virus</span></span>
<span><span class="co">#&gt;         go_id odds_ratios        pvals          fdr  hits gene_set_lengths</span></span>
<span><span class="co">#&gt;        &lt;char&gt;       &lt;num&gt;        &lt;num&gt;        &lt;num&gt; &lt;num&gt;            &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: GO:0003723    14.81244 3.122680e-16 3.441817e-12    23             1205</span></span>
<span><span class="co">#&gt; 2: GO:0007259    97.16563 8.704402e-15 9.587028e-11     9               66</span></span>
<span><span class="co">#&gt; 3: GO:0006954    22.90818 1.408854e-13 7.758558e-10    14              447</span></span>
<span><span class="co">#&gt; 4: GO:0005125    31.85212 1.130962e-12 4.152138e-09    11              235</span></span>
<span><span class="co">#&gt; 5: GO:0009897    19.25374 1.941168e-10 5.133756e-07    11              379</span></span>
<span><span class="co">#&gt; 6: GO:0098586    56.46451 2.330559e-10 5.133756e-07     7               79</span></span></code></pre></div>
</div>
<section class="level4"><h4 id="speed-of-the-functions">Speed of the functions<a class="anchor" aria-label="anchor" href="#speed-of-the-functions"></a>
</h4>
<p>Also this functions were designed with speed in mind (leveraging Rust’s borrowing and parallelisations where possible). To run this over 100 potential target gene sets takes only a few seconds.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">go_gene_universe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">go_data_dt</span><span class="op">$</span><span class="va">ensembl_id</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate random target sets</span></span>
<span></span>
<span><span class="va">go_target_sets_no</span> <span class="op">&lt;-</span> <span class="fl">100L</span></span>
<span></span>
<span><span class="va">seed</span> <span class="op">&lt;-</span> <span class="fl">246L</span></span>
<span></span>
<span><span class="va">go_target_gene_sets</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span></span>
<span>  <span class="fl">1</span><span class="op">:</span><span class="va">go_target_sets_no</span>,</span>
<span>  <span class="op">~</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">.x</span> <span class="op">*</span> <span class="va">seed</span><span class="op">)</span></span>
<span>    <span class="va">size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">50</span><span class="op">:</span><span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">go_gene_universe</span>, <span class="va">size</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">go_target_gene_sets</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_chr</a></span><span class="op">(</span></span>
<span>  <span class="fl">1</span><span class="op">:</span><span class="va">go_target_sets_no</span>,</span>
<span>  <span class="op">~</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">seed</span> <span class="op">+</span> <span class="va">.x</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">letters</span>, <span class="fl">3</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Quick test with tictoc</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">rs_results_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gse_go_elim_method_list.html">gse_go_elim_method_list</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">go_data_s7</span>,</span>
<span>  target_gene_list <span class="op">=</span> <span class="va">go_target_gene_sets</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; 1.456 sec elapsed</span></span></code></pre></div>
</div>
</section></section><section class="level3"><h3 id="alternative-simplifying-results">Alternative: simplifying results<a class="anchor" aria-label="anchor" href="#alternative-simplifying-results"></a>
</h3>
<p>Another approach is to run the enrichment over the gene ontology as is and simplify subsequently. <code>bixverse</code> also has options for that. Let’s run an example with the gene ontology:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the go data that is part of the package</span></span>
<span><span class="va">go_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_go_human_data.html">load_go_human_data</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare the go &lt;&gt; gene information</span></span>
<span><span class="va">min_genes</span> <span class="op">&lt;-</span> <span class="fl">3L</span></span>
<span></span>
<span><span class="va">go_genes</span> <span class="op">&lt;-</span> <span class="va">go_data</span><span class="op">$</span><span class="va">go_to_genes</span></span>
<span><span class="va">go_genes_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">go_genes</span><span class="op">$</span><span class="va">ensembl_id</span>, <span class="va">go_genes</span><span class="op">$</span><span class="va">go_id</span><span class="op">)</span></span>
<span><span class="va">go_genes_ls</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/keep.html" class="external-link">keep</a></span><span class="op">(</span><span class="va">go_genes_ls</span>, \<span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">min_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run the hypergeometric test</span></span>
<span><span class="va">go_results_unfiltered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gse_hypergeometric.html">gse_hypergeometric</a></span><span class="op">(</span></span>
<span>  target_genes <span class="op">=</span> <span class="va">target_genes_1</span>,</span>
<span>  gene_set_list <span class="op">=</span> <span class="va">go_genes_ls</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">go_results_unfiltered</span><span class="op">)</span></span>
<span><span class="co">#&gt;    gene_set_name odds_ratios        pvals          fdr  hits gene_set_lengths</span></span>
<span><span class="co">#&gt;           &lt;char&gt;       &lt;num&gt;        &lt;num&gt;        &lt;num&gt; &lt;num&gt;            &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:    GO:0003723   22.120609 1.723343e-23 1.579099e-19    31             1546</span></span>
<span><span class="co">#&gt; 2:    GO:0006364   35.837205 2.936032e-12 1.345143e-08    10              159</span></span>
<span><span class="co">#&gt; 3:    GO:0005730    8.860092 6.601460e-12 1.581347e-08    23             1927</span></span>
<span><span class="co">#&gt; 4:    GO:0042254   43.069954 6.903186e-12 1.581347e-08     9              118</span></span>
<span><span class="co">#&gt; 5:    GO:0005634    8.314215 9.730968e-12 1.783297e-08    39             6736</span></span>
<span><span class="co">#&gt; 6:    GO:0005654    6.752334 9.279447e-11 1.417126e-07    30             4005</span></span>
<span><span class="co">#&gt;    target_set_lengths</span></span>
<span><span class="co">#&gt;                 &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:                 50</span></span>
<span><span class="co">#&gt; 2:                 50</span></span>
<span><span class="co">#&gt; 3:                 50</span></span>
<span><span class="co">#&gt; 4:                 50</span></span>
<span><span class="co">#&gt; 5:                 50</span></span>
<span><span class="co">#&gt; 6:                 50</span></span></code></pre></div>
</div>
<p>Now that we have the results, we can simplify them. The approach here is to identify similar terms via the Wang similarity and within a subset of similar terms take the term with the best test statistic, i.e., lowest FDR. Should the terms have the same FDR, the function will select for the most specific term (i.e., lower in the ontology).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">go_parent_child_dt</span> <span class="op">&lt;-</span> <span class="va">go_data</span><span class="op">$</span><span class="va">gene_ontology</span><span class="op">[</span></span>
<span>  <span class="va">relationship</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"is_a"</span>, <span class="st">"part_of"</span><span class="op">)</span></span>
<span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setattr.html" class="external-link">setnames</a></span><span class="op">(</span></span>
<span>    old <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"from"</span>, <span class="st">"to"</span>, <span class="st">"relationship"</span><span class="op">)</span>,</span>
<span>    new <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"parent"</span>, <span class="st">"child"</span>, <span class="st">"type"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">go_results_simplified</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simplify_hypergeom_res.html">simplify_hypergeom_res</a></span><span class="op">(</span></span>
<span>  res <span class="op">=</span> <span class="va">go_results_unfiltered</span>,</span>
<span>  parent_child_dt <span class="op">=</span> <span class="va">go_parent_child_dt</span>,</span>
<span>  weights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.8</span>, <span class="fl">0.6</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"is_a"</span>, <span class="st">"part_of"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">go_results_simplified</span><span class="op">)</span></span>
<span><span class="co">#&gt;    gene_set_name odds_ratios        pvals          fdr  hits gene_set_lengths</span></span>
<span><span class="co">#&gt;           &lt;char&gt;       &lt;num&gt;        &lt;num&gt;        &lt;num&gt; &lt;num&gt;            &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:    GO:0003723   22.120609 1.723343e-23 1.579099e-19    31             1546</span></span>
<span><span class="co">#&gt; 2:    GO:0006364   35.837205 2.936032e-12 1.345143e-08    10              159</span></span>
<span><span class="co">#&gt; 3:    GO:0005730    8.860092 6.601460e-12 1.581347e-08    23             1927</span></span>
<span><span class="co">#&gt; 4:    GO:0042254   43.069954 6.903186e-12 1.581347e-08     9              118</span></span>
<span><span class="co">#&gt; 5:    GO:0032040   44.202960 1.623943e-08 2.023038e-05     6               72</span></span>
<span><span class="co">#&gt; 6:    GO:0000055  455.891304 2.368116e-07 2.169905e-04     3                6</span></span>
<span><span class="co">#&gt;    target_set_lengths</span></span>
<span><span class="co">#&gt;                 &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:                 50</span></span>
<span><span class="co">#&gt; 2:                 50</span></span>
<span><span class="co">#&gt; 3:                 50</span></span>
<span><span class="co">#&gt; 4:                 50</span></span>
<span><span class="co">#&gt; 5:                 50</span></span>
<span><span class="co">#&gt; 6:                 50</span></span></code></pre></div>
</div>
</section><section class="level3"><h3 id="gsea">GSEA<a class="anchor" aria-label="anchor" href="#gsea"></a>
</h3>
<p><code>bixverse</code> also has implementations of the GSEA based on <a href="https://www.pnas.org/doi/10.1073/pnas.0506580102" class="external-link">Subramanian et al</a> to test against continuous scores. The package provides also the fgsea implementations of <a href="https://www.biorxiv.org/content/10.1101/060012v3" class="external-link">Korotkevich et al.</a>. These can be run like:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This is the example of fgsea</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/ctlab/fgsea/" class="external-link">"fgsea"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">examplePathways</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">exampleRanks</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42L</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fgsea_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fgsea/man/fgsea.html" class="external-link">fgsea</a></span><span class="op">(</span></span>
<span>  pathways <span class="op">=</span> <span class="va">examplePathways</span>, </span>
<span>  stats <span class="op">=</span> <span class="va">exampleRanks</span>,</span>
<span>  minSize <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  maxSize <span class="op">=</span> <span class="fl">500</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setorder.html" class="external-link">setorder</a></span><span class="op">(</span><span class="va">pathway</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fgsea_res</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                                                    pathway</span></span>
<span><span class="co">#&gt;                                                                                     &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:                                                                1221633_Meiotic_Synapsis</span></span>
<span><span class="co">#&gt; 2:                                   1445146_Translocation_of_Glut4_to_the_Plasma_Membrane</span></span>
<span><span class="co">#&gt; 3: 442533_Transcriptional_Regulation_of_Adipocyte_Differentiation_in_3T3-L1_Pre-adipocytes</span></span>
<span><span class="co">#&gt; 4:                                                                  508751_Circadian_Clock</span></span>
<span><span class="co">#&gt; 5:                                               5334727_Mus_musculus_biological_processes</span></span>
<span><span class="co">#&gt; 6:                                        573389_NoRC_negatively_regulates_rRNA_expression</span></span>
<span><span class="co">#&gt;         pval      padj    log2err         ES        NES  size</span></span>
<span><span class="co">#&gt;        &lt;num&gt;     &lt;num&gt;      &lt;num&gt;      &lt;num&gt;      &lt;num&gt; &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1: 0.5490534 0.7262873 0.06674261  0.2885754  0.9399884    27</span></span>
<span><span class="co">#&gt; 2: 0.6952862 0.8366277 0.05445560  0.2387284  0.8366856    39</span></span>
<span><span class="co">#&gt; 3: 0.1122449 0.2599823 0.21392786 -0.3640706 -1.3460572    31</span></span>
<span><span class="co">#&gt; 4: 0.7826888 0.8799951 0.05312981  0.2516324  0.7287088    17</span></span>
<span><span class="co">#&gt; 5: 0.3580060 0.5579562 0.08197788  0.2469065  1.0498921   106</span></span>
<span><span class="co">#&gt; 6: 0.4198895 0.6197865 0.08407456  0.3607407  1.0446784    17</span></span>
<span><span class="co">#&gt;                                 leadingEdge</span></span>
<span><span class="co">#&gt;                                      &lt;list&gt;</span></span>
<span><span class="co">#&gt; 1:                        15270,12189,71846</span></span>
<span><span class="co">#&gt; 2:  17918,19341,20336,22628,22627,20619,...</span></span>
<span><span class="co">#&gt; 3: 76199,19014,26896,229003,17977,17978,...</span></span>
<span><span class="co">#&gt; 4:                              20893,59027</span></span>
<span><span class="co">#&gt; 5:  60406,19361,15270,20893,12189,68240,...</span></span>
<span><span class="co">#&gt; 6:                              60406,20018</span></span></code></pre></div>
</div>
<p>And this is how you run it via <code>bixverse</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bixverse_fgsea</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_fgsea.html">calc_fgsea</a></span><span class="op">(</span></span>
<span>  stats <span class="op">=</span> <span class="va">exampleRanks</span>,</span>
<span>  pathways <span class="op">=</span> <span class="va">examplePathways</span>,</span>
<span>  gsea_params <span class="op">=</span> <span class="fu"><a href="../reference/params_gsea.html">params_gsea</a></span><span class="op">(</span>min_size <span class="op">=</span> <span class="fl">15L</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setorder.html" class="external-link">setorder</a></span><span class="op">(</span><span class="va">pathway_name</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bixverse_fgsea</span><span class="op">)</span></span>
<span><span class="co">#&gt;            es        nes     pvals n_more_extreme  size</span></span>
<span><span class="co">#&gt;         &lt;num&gt;      &lt;num&gt;     &lt;num&gt;          &lt;num&gt; &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:  0.2885754  0.9321129 0.5588723            336    27</span></span>
<span><span class="co">#&gt; 2:  0.2387284  0.8353752 0.7140600            451    39</span></span>
<span><span class="co">#&gt; 3: -0.3640706 -1.3241710 0.1282051             49    31</span></span>
<span><span class="co">#&gt; 4:  0.2516324  0.7211793 0.8109541            458    17</span></span>
<span><span class="co">#&gt; 5:  0.2469065  1.0551934 0.3751804            259   106</span></span>
<span><span class="co">#&gt; 6:  0.3607407  1.0338840 0.4204947            237    17</span></span>
<span><span class="co">#&gt;                                                                               pathway_name</span></span>
<span><span class="co">#&gt;                                                                                     &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:                                                                1221633_Meiotic_Synapsis</span></span>
<span><span class="co">#&gt; 2:                                   1445146_Translocation_of_Glut4_to_the_Plasma_Membrane</span></span>
<span><span class="co">#&gt; 3: 442533_Transcriptional_Regulation_of_Adipocyte_Differentiation_in_3T3-L1_Pre-adipocytes</span></span>
<span><span class="co">#&gt; 4:                                                                  508751_Circadian_Clock</span></span>
<span><span class="co">#&gt; 5:                                               5334727_Mus_musculus_biological_processes</span></span>
<span><span class="co">#&gt; 6:                                        573389_NoRC_negatively_regulates_rRNA_expression</span></span>
<span><span class="co">#&gt;                                leading_edge    log2err       fdr</span></span>
<span><span class="co">#&gt;                                      &lt;list&gt;      &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:                  15270,12189,71846,19357 0.06407038 0.7359532</span></span>
<span><span class="co">#&gt; 2:  17918,19341,20336,22628,22627,20619,... 0.05029481 0.8557113</span></span>
<span><span class="co">#&gt; 3: 76199,19014,26896,229003,17977,17978,... 0.19991523 0.2900703</span></span>
<span><span class="co">#&gt; 4:                        20893,59027,19883 0.04959020 0.9063587</span></span>
<span><span class="co">#&gt; 5:  60406,19361,15270,20893,12189,68240,... 0.07707367 0.5816288</span></span>
<span><span class="co">#&gt; 6:                 60406,20018,245688,20017 0.08175156 0.6207015</span></span></code></pre></div>
</div>
<p>We can appreciate the (more or less) same p-values estimated by both methods.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">fgsea_res</span><span class="op">$</span><span class="va">pval</span><span class="op">)</span>,</span>
<span>  y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">bixverse_fgsea</span><span class="op">$</span><span class="va">pvals</span><span class="op">)</span>,</span>
<span>  xlab <span class="op">=</span> <span class="st">"-log10(pval) fgsea"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"-log10(pval) bixverse"</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"fgsea and bixverse"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="gse_methods_files/figure-html/p-value%20comparison%20between%20fgsea%20and%20bixverse%20fgsea-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The speed is very comparable with the Rcpp implementation of fgsea, indicating no issues in terms of performance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">microbenchmark</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbenchmark/man/microbenchmark.html" class="external-link">microbenchmark</a></span><span class="op">(</span></span>
<span>  fgsea <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/fgsea/man/fgsea.html" class="external-link">fgsea</a></span><span class="op">(</span></span>
<span>    pathways <span class="op">=</span> <span class="va">examplePathways</span>, </span>
<span>    stats <span class="op">=</span> <span class="va">exampleRanks</span>,</span>
<span>    minSize <span class="op">=</span> <span class="fl">15</span>,</span>
<span>    maxSize <span class="op">=</span> <span class="fl">500</span></span>
<span>  <span class="op">)</span>,</span>
<span>  rust <span class="op">=</span> <span class="fu"><a href="../reference/calc_fgsea.html">calc_fgsea</a></span><span class="op">(</span></span>
<span>    stats <span class="op">=</span> <span class="va">exampleRanks</span>,</span>
<span>    pathways <span class="op">=</span> <span class="va">examplePathways</span>,</span>
<span>    gsea_params <span class="op">=</span> <span class="fu"><a href="../reference/params_gsea.html">params_gsea</a></span><span class="op">(</span>min_size <span class="op">=</span> <span class="fl">15L</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  times <span class="op">=</span> <span class="fl">10L</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Unit: seconds</span></span>
<span><span class="co">#&gt;   expr      min       lq     mean   median       uq      max neval</span></span>
<span><span class="co">#&gt;  fgsea 2.254475 2.317307 2.541921 2.371093 2.823325 3.018690    10</span></span>
<span><span class="co">#&gt;   rust 2.444489 2.446764 2.456520 2.456730 2.463731 2.475314    10</span></span></code></pre></div>
</div>
</section><section class="level3"><h3 id="gsea-gene-ontology-aware">GSEA gene ontology aware<a class="anchor" aria-label="anchor" href="#gsea-gene-ontology-aware"></a>
</h3>
<p><code>bixverse</code> also implements GSEA in a gene-ontology aware form. In this case, you use the same object as for the hypergeometric test, but instead provide the ranked list. Similar to the hyper geometric version, the ontology is traversed from the lowest level first and the permutation-based p-values are calculated for a given term. Should this p-value be below the elimination threshold, the genes from that given gene ontology term are removed from all of its ancestors, yielding more specific GO terms. Once this has been done with the full ontology, the multi-level method from fgsea is used to estimate lower p-values than just possible via permutation for terms that show significance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_universe_go</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">go_data_dt</span><span class="op">[</span>, <span class="st">"ensembl_id"</span><span class="op">]</span>, use.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42L</span><span class="op">)</span></span>
<span></span>
<span><span class="va">random_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gene_universe_go</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">random_stats</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gene_universe_go</span></span>
<span></span>
<span><span class="va">go_gsea_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fgsea_go_elim.html">fgsea_go_elim</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">go_data_s7</span>,</span>
<span>  stats <span class="op">=</span> <span class="va">random_stats</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">go_gsea_res</span><span class="op">)</span></span>
<span><span class="co">#&gt;         go_id         es       nes  size        pvals n_more_extreme</span></span>
<span><span class="co">#&gt;        &lt;char&gt;      &lt;num&gt;     &lt;num&gt; &lt;num&gt;        &lt;num&gt;          &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: GO:0031465 -0.9025468 -1.925965     6 6.165332e-05              0</span></span>
<span><span class="co">#&gt; 2: GO:0050877  0.5116421  1.950886    44 2.410338e-04              2</span></span>
<span><span class="co">#&gt; 3: GO:0072675 -0.8744037 -1.865910     6 2.625170e-04              0</span></span>
<span><span class="co">#&gt; 4: GO:0006066 -0.6718553 -2.026201    18 2.712234e-04              0</span></span>
<span><span class="co">#&gt; 5: GO:0071392 -0.5217830 -1.904151    36 3.788231e-04              0</span></span>
<span><span class="co">#&gt; 6: GO:0002090 -0.8302363 -1.868975     7 6.827804e-04              3</span></span>
<span><span class="co">#&gt;                                                                                           leading_edge</span></span>
<span><span class="co">#&gt;                                                                                                 &lt;list&gt;</span></span>
<span><span class="co">#&gt; 1:                                                     ENSG00000110844,ENSG00000180370,ENSG00000160685</span></span>
<span><span class="co">#&gt; 2: ENSG00000131263,ENSG00000229674,ENSG00000172987,ENSG00000108878,ENSG00000050165,ENSG00000143061,...</span></span>
<span><span class="co">#&gt; 3:                                                                                     ENSG00000170613</span></span>
<span><span class="co">#&gt; 4: ENSG00000287395,ENSG00000139354,ENSG00000211659,ENSG00000160345,ENSG00000007237,ENSG00000139679,...</span></span>
<span><span class="co">#&gt; 5: ENSG00000118193,ENSG00000134531,ENSG00000166897,ENSG00000162980,ENSG00000170348,ENSG00000103018,...</span></span>
<span><span class="co">#&gt; 6:                                     ENSG00000061938,ENSG00000169507,ENSG00000289721,ENSG00000083814</span></span>
<span><span class="co">#&gt;      log2err       fdr                                 go_name</span></span>
<span><span class="co">#&gt;        &lt;num&gt;     &lt;num&gt;                                  &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1: 0.5384341 0.4815124  Cul4B-RING E3 ubiquitin ligase complex</span></span>
<span><span class="co">#&gt; 2: 0.5188481 0.5295637                  nervous system process</span></span>
<span><span class="co">#&gt; 3: 0.4984931 0.5295637                       osteoclast fusion</span></span>
<span><span class="co">#&gt; 4: 0.4984931 0.5295637               alcohol metabolic process</span></span>
<span><span class="co">#&gt; 5: 0.4984931 0.5917217 cellular response to estradiol stimulus</span></span>
<span><span class="co">#&gt; 6: 0.4772708 0.8290480  regulation of receptor internalization</span></span></code></pre></div>
</div>
</section></section><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config);
    }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gregor Lueg.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
