<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Single Cell Analysis • bixverse</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="Single_cell_analysis_files/libs/quarto-html/popper.min.js"></script><script src="Single_cell_analysis_files/libs/quarto-html/tippy.umd.min.js"></script><link href="Single_cell_analysis_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Single_cell_analysis_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Single Cell Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">bixverse</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Single Cell Analysis</h1>





      <div class="d-none name"><code></code></div>
    </div>






    <section class="section level2"><h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette shows how to use the <code>bixverse</code> package to perform single cell analysis.</p>
<section class="level2"><h3 id="problem-statement">Problem statement<a class="anchor" aria-label="anchor" href="#problem-statement"></a>
</h3>
<p>The size of single cell datasets is growing rapidly, however the computational tools used to to analyse them are not keeping up. The current tools solution ot bigger datasets is just use a larger memory machine, however this is not always possible especially if analysis is required to be run locally on a laptop.</p>
<p>The aim of this package is to make it possible to analyse single-cell datasets containing millions of cells on a standard laptop. This is achieved through an out-of-memory design where large data objects—such as count matrices and cell or feature metadata—are stored on disk rather than held entirely in memory. Metadata tables are managed through DuckDB, allowing for fast and efficient access to subsets of the data when needed. While it may seem like overkill to use a duckdb to store the obs and var tables for 100,000 cells these tables can quickly become very large with millions of cells and become unwieldy to hold in memory on a standard laptop.</p>
<p>Computationally intensive tasks are handled by Rust back-end utilities that stream data from disk, perform the required operations, and return results to R. This approach combines the ease of use of R with the speed and efficiency of compiled Rust code, enabling high-performance single-cell analysis with minimal memory requirements.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bixverse</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="level2"><h3 id="create-test-data">Create test data<a class="anchor" aria-label="anchor" href="#create-test-data"></a>
</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## This function is used to create synthetic data</span></span>
<span><span class="va">single_cell_test_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_single_cell_test_data.html">generate_single_cell_test_data</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">f_path_csr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"csr_test.h5ad"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/write_h5ad_sc.html">write_h5ad_sc</a></span><span class="op">(</span></span>
<span>  f_path <span class="op">=</span> <span class="va">f_path_csr</span>,</span>
<span>  counts <span class="op">=</span> <span class="va">single_cell_test_data</span><span class="op">$</span><span class="va">counts</span>,</span>
<span>  obs <span class="op">=</span> <span class="va">single_cell_test_data</span><span class="op">$</span><span class="va">obs</span>,</span>
<span>  var <span class="op">=</span> <span class="va">single_cell_test_data</span><span class="op">$</span><span class="va">var</span>,</span>
<span>  .verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">single_cell_test_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_single_cell_test_data.html">generate_single_cell_test_data</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">f_path_v1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"cells_csv"</span><span class="op">)</span></span>
<span><span class="va">f_path_v2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"genes_tsv"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">f_path_v1</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">f_path_v2</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">counts_csc</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">single_cell_test_data</span><span class="op">$</span><span class="va">counts</span>, <span class="st">"CsparseMatrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># save a version with rows = cells and format type csv for the rest</span></span>
<span></span>
<span><span class="fu"><a href="../reference/write_cellranger_output.html">write_cellranger_output</a></span><span class="op">(</span></span>
<span>  f_path <span class="op">=</span> <span class="va">f_path_v1</span>,</span>
<span>  counts <span class="op">=</span> <span class="va">single_cell_test_data</span><span class="op">$</span><span class="va">counts</span>,</span>
<span>  obs <span class="op">=</span> <span class="va">single_cell_test_data</span><span class="op">$</span><span class="va">obs</span>,</span>
<span>  var <span class="op">=</span> <span class="va">single_cell_test_data</span><span class="op">$</span><span class="va">var</span>,</span>
<span>  rows <span class="op">=</span> <span class="st">"cells"</span>,</span>
<span>  format_type <span class="op">=</span> <span class="st">"csv"</span>,</span>
<span>  .verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># save a version with rows = genes and format type tsv for the rest</span></span>
<span></span>
<span><span class="fu"><a href="../reference/write_cellranger_output.html">write_cellranger_output</a></span><span class="op">(</span></span>
<span>  f_path <span class="op">=</span> <span class="va">f_path_v2</span>,</span>
<span>  counts <span class="op">=</span> <span class="va">single_cell_test_data</span><span class="op">$</span><span class="va">counts</span>,</span>
<span>  obs <span class="op">=</span> <span class="va">single_cell_test_data</span><span class="op">$</span><span class="va">obs</span>,</span>
<span>  var <span class="op">=</span> <span class="va">single_cell_test_data</span><span class="op">$</span><span class="va">var</span>,</span>
<span>  rows <span class="op">=</span> <span class="st">"genes"</span>,</span>
<span>  format_type <span class="op">=</span> <span class="st">"tsv"</span>,</span>
<span>  .verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</section></section><section class="section level2"><h2 id="single-cell-data-processing-using-bixverse">Single cell data processing using bixverse<a class="anchor" aria-label="anchor" href="#single-cell-data-processing-using-bixverse"></a>
</h2>
<section class="level2"><h3 id="load-in-data">Load in data<a class="anchor" aria-label="anchor" href="#load-in-data"></a>
</h3>
<p>The core of the single cell analysis is createing a <code>sc_object</code> this object contains the connections to the data on disk that can be accessed in rust and accessed and indexed by the duckdb.</p>
<p>Here we are setting the path to store the duckdb and files to <code>tempdir</code> but in reality you want to keep these accessible such that you can always access the data.</p>
<p>When generating the <code>sc_object</code> an <code>sc_qc_param</code> will need to be set, this defines the parameters used to filter out low-quality cells at the outset of the analysis. These cells are excluded during the creation of the on-disk data, thereby reducing computational load. Since poor-quality cells are not needed for downstream analysis, omitting them also minimizes both storage and memory requirements.</p>
<p>Currently these parameters are:</p>
<ul>
<li>min_unique_genes = minimum number of genes expressed in a cell for it to be considered</li>
<li>min_lib_size = minimum number of UMTs in a cell for it to be considered</li>
<li>min_cells = minimum number of cells for a gene to be expressed for the gene to be considered</li>
<li>target_size = library size to be normalised to during the lognorm step</li>
</ul>
<p>** Note your obs table will have a <code>to_keep</code> column appended to it when the <code>sc_object</code> is created this allows tracking of what cells are to be used in each analysis by default it will be set to all TRUE **</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## The QC parameters are set using the constructor function</span></span>
<span><span class="va">sc_qc_param</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/params_sc_min_quality.html">params_sc_min_quality</a></span><span class="op">(</span></span>
<span>  min_unique_genes <span class="op">=</span> <span class="fl">45L</span>,</span>
<span>  min_lib_size <span class="op">=</span> <span class="fl">300L</span>,</span>
<span>  min_cells <span class="op">=</span> <span class="fl">500L</span>,</span>
<span>  target_size <span class="op">=</span> <span class="fl">1000</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<p>There are different ways to process single cell data and users may have data stored in a number of ways we have tried to account for these when developing this package.</p>
<p>However when an <code>sc_object</code> is created from whatever the source a number of things happen</p>
<ol type="1">
<li>First pass: Identify genes that are expressed in a sufficient number of cells based on the specified threshold set in <code>sc_qc_param</code>
</li>
<li>Second pass: Using the selected genes from the first pass, determine which cells meet the inclusion threshold</li>
<li>CSR and CSC orientations of the data are saved in .bin files for rapid indexing of the data (more on this later)</li>
<li>Data is log normalised based on the <code>target_size</code> parameter</li>
</ol>
<p>Two columns are also added to the obs table <code>nnz</code> which is number of genes detected in the cell and <code>lib_size</code> which is number of UMTs</p>
<section class="level3"><h4 id="h5ad">H5AD<a class="anchor" aria-label="anchor" href="#h5ad"></a>
</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#  For demonstration</span></span>
<span><span class="va">sc_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/single_cell_exp.html">single_cell_exp</a></span><span class="op">(</span>dir_data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">sc_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_h5ad.html">load_h5ad</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">sc_object</span>,</span>
<span>  h5_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"csr_test.h5ad"</span>, package <span class="op">=</span> <span class="st">"bixverse"</span><span class="op">)</span>,</span>
<span>  sc_qc_param <span class="op">=</span> <span class="va">sc_qc_param</span>,</span>
<span>  .verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="level3"><h4 id="cell-ranger">Cell Ranger<a class="anchor" aria-label="anchor" href="#cell-ranger"></a>
</h4>
<p>We have accounted for all possible combinations of outputs when loading data from mtx files it can either be cells as rows or cells as columns, and the obs and var tables can either be loaded as <code>.tsv</code> or <code>.csv</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/single_cell_exp.html">single_cell_exp</a></span><span class="op">(</span>dir_data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">params_cells_rows_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/params_sc_mtx_io.html">params_sc_mtx_io</a></span><span class="op">(</span></span>
<span>  path_mtx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">f_path_v1</span>, <span class="st">"mat.mtx"</span><span class="op">)</span>,</span>
<span>  path_obs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">f_path_v1</span>, <span class="st">"barcodes.csv"</span><span class="op">)</span>,</span>
<span>  path_var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">f_path_v1</span>, <span class="st">"features.csv"</span><span class="op">)</span>,</span>
<span>  cells_as_rows <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  has_hdr <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Data can also be loaded in this way, not run here shown purely as an example</span></span>
<span></span>
<span><span class="co"># params_genes_rows_tsv &lt;- params_sc_mtx_io(</span></span>
<span><span class="co">#   path_mtx = file.path(f_path_v2, "mat.mtx"),</span></span>
<span><span class="co">#   path_obs = file.path(f_path_v2, "barcodes.tsv"),</span></span>
<span><span class="co">#   path_var = file.path(f_path_v2, "features.tsv"),</span></span>
<span><span class="co">#   cells_as_rows = FALSE,</span></span>
<span><span class="co">#   has_hdr = TRUE</span></span>
<span><span class="co"># )</span></span>
<span></span>
<span><span class="va">sc_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_mtx.html">load_mtx</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">sc_object</span>,</span>
<span>  sc_mtx_io_param <span class="op">=</span> <span class="va">params_cells_rows_csv</span>,</span>
<span>  sc_qc_param <span class="op">=</span> <span class="va">sc_qc_param</span>,</span>
<span>  streaming <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co">#  Shall the data be streamed during the conversion of CSR to CSC. Defaults to TRUE and should be used for larger data sets.</span></span>
<span>  .verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading observations data from flat file into the DuckDB.</span></span>
<span><span class="co">#&gt; Loading variable data from flat file into the DuckDB.</span></span></code></pre></div>
</div>
</section></section><section class="level2"><h3 id="accessing-data">Accessing data<a class="anchor" aria-label="anchor" href="#accessing-data"></a>
</h3>
<p>A duckdb with the single cell data has been created and it does not live in memory, however for some operations (plotting) you do need the data, A number of helpful getter functions have to been written to retrieve the data.</p>
<section class="level3"><h4 id="obs-table-cell-info">Obs table (cell info)<a class="anchor" aria-label="anchor" href="#obs-table-cell-info"></a>
</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this returns a data.table of the obs table which can then be interacted with</span></span>
<span><span class="va">sc_object</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt;      cell_idx   cell_id    cell_grp batch_index   nnz lib_size to_keep</span></span>
<span><span class="co">#&gt;         &lt;num&gt;    &lt;char&gt;      &lt;char&gt;       &lt;num&gt; &lt;num&gt;    &lt;num&gt;  &lt;lgcl&gt;</span></span>
<span><span class="co">#&gt;   1:        1 cell_0001 cell_type_1           1    52      371    TRUE</span></span>
<span><span class="co">#&gt;   2:        2 cell_0003 cell_type_3           1    52      495    TRUE</span></span>
<span><span class="co">#&gt;   3:        3 cell_0004 cell_type_1           1    48      551    TRUE</span></span>
<span><span class="co">#&gt;   4:        4 cell_0005 cell_type_2           1    53      422    TRUE</span></span>
<span><span class="co">#&gt;   5:        5 cell_0006 cell_type_3           1    57      394    TRUE</span></span>
<span><span class="co">#&gt;  ---                                                                  </span></span>
<span><span class="co">#&gt; 925:      925 cell_0995 cell_type_2           1    53      548    TRUE</span></span>
<span><span class="co">#&gt; 926:      926 cell_0996 cell_type_3           1    57      545    TRUE</span></span>
<span><span class="co">#&gt; 927:      927 cell_0997 cell_type_1           1    56      543    TRUE</span></span>
<span><span class="co">#&gt; 928:      928 cell_0998 cell_type_2           1    49      487    TRUE</span></span>
<span><span class="co">#&gt; 929:      929 cell_1000 cell_type_1           1    60      474    TRUE</span></span>
<span></span>
<span><span class="co"># The obs can also be retrieved using the obs getter which has mutliple arguments for fitlering</span></span>
<span><span class="fu"><a href="../reference/get_sc_obs.html">get_sc_obs</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span></span>
<span><span class="co">#&gt;      cell_idx   cell_id    cell_grp batch_index   nnz lib_size to_keep</span></span>
<span><span class="co">#&gt;         &lt;num&gt;    &lt;char&gt;      &lt;char&gt;       &lt;num&gt; &lt;num&gt;    &lt;num&gt;  &lt;lgcl&gt;</span></span>
<span><span class="co">#&gt;   1:        1 cell_0001 cell_type_1           1    52      371    TRUE</span></span>
<span><span class="co">#&gt;   2:        2 cell_0003 cell_type_3           1    52      495    TRUE</span></span>
<span><span class="co">#&gt;   3:        3 cell_0004 cell_type_1           1    48      551    TRUE</span></span>
<span><span class="co">#&gt;   4:        4 cell_0005 cell_type_2           1    53      422    TRUE</span></span>
<span><span class="co">#&gt;   5:        5 cell_0006 cell_type_3           1    57      394    TRUE</span></span>
<span><span class="co">#&gt;  ---                                                                  </span></span>
<span><span class="co">#&gt; 925:      925 cell_0995 cell_type_2           1    53      548    TRUE</span></span>
<span><span class="co">#&gt; 926:      926 cell_0996 cell_type_3           1    57      545    TRUE</span></span>
<span><span class="co">#&gt; 927:      927 cell_0997 cell_type_1           1    56      543    TRUE</span></span>
<span><span class="co">#&gt; 928:      928 cell_0998 cell_type_2           1    49      487    TRUE</span></span>
<span><span class="co">#&gt; 929:      929 cell_1000 cell_type_1           1    60      474    TRUE</span></span></code></pre></div>
</div>
<p><strong>Add data to obs</strong></p>
<p>Sometimes it is desirable to add data to the obs table for further analysis this can be acheived using <code><a href="../reference/set_sc_new_obs_col.html">set_sc_new_obs_col()</a></code> or <code><a href="../reference/set_sc_new_obs_col_multiple.html">set_sc_new_obs_col_multiple()</a></code></p>
<p>Note the new data must be of same length as <code><a href="../reference/get_cells_to_keep.html">bixverse::get_cells_to_keep()</a></code> and have the same order.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_cells_to_keep.html">get_cells_to_keep</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## new data to add single column</span></span>
<span><span class="va">new_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"random_new_data"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">n_cells</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Using function or using</span></span>
<span><span class="va">sc_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/set_sc_new_obs_col.html">set_sc_new_obs_col</a></span><span class="op">(</span></span>
<span>  <span class="va">sc_object</span>,</span>
<span>  col_name <span class="op">=</span> <span class="st">"random_new_data"</span>,</span>
<span>  new_data <span class="op">=</span> <span class="va">new_data</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or using [[]]</span></span>
<span><span class="va">sc_object</span><span class="op">[[</span><span class="st">"random_new_data2"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">new_data</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;    cell_idx   cell_id    cell_grp batch_index   nnz lib_size to_keep</span></span>
<span><span class="co">#&gt;       &lt;num&gt;    &lt;char&gt;      &lt;char&gt;       &lt;num&gt; &lt;num&gt;    &lt;num&gt;  &lt;lgcl&gt;</span></span>
<span><span class="co">#&gt; 1:        1 cell_0001 cell_type_1           1    52      371    TRUE</span></span>
<span><span class="co">#&gt; 2:        2 cell_0003 cell_type_3           1    52      495    TRUE</span></span>
<span><span class="co">#&gt; 3:        3 cell_0004 cell_type_1           1    48      551    TRUE</span></span>
<span><span class="co">#&gt; 4:        4 cell_0005 cell_type_2           1    53      422    TRUE</span></span>
<span><span class="co">#&gt; 5:        5 cell_0006 cell_type_3           1    57      394    TRUE</span></span>
<span><span class="co">#&gt; 6:        6 cell_0007 cell_type_1           1    49      593    TRUE</span></span>
<span><span class="co">#&gt;    random_new_data random_new_data2</span></span>
<span><span class="co">#&gt;             &lt;char&gt;           &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1: random_new_data  random_new_data</span></span>
<span><span class="co">#&gt; 2: random_new_data  random_new_data</span></span>
<span><span class="co">#&gt; 3: random_new_data  random_new_data</span></span>
<span><span class="co">#&gt; 4: random_new_data  random_new_data</span></span>
<span><span class="co">#&gt; 5: random_new_data  random_new_data</span></span>
<span><span class="co">#&gt; 6: random_new_data  random_new_data</span></span>
<span></span>
<span><span class="co">## Multiple columns can also be added must be a list of vector</span></span>
<span><span class="va">new_data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"other_random_data"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">n_cells</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="st">"even_different_random_data"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">n_cells</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## either like this</span></span>
<span><span class="va">sc_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/set_sc_new_obs_col_multiple.html">set_sc_new_obs_col_multiple</a></span><span class="op">(</span><span class="va">sc_object</span>, <span class="va">new_data_list</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or this</span></span>
<span><span class="co"># sc_object[[names(new_data_list)]] &lt;- new_data_list</span></span>
<span></span>
<span><span class="co"># or this</span></span>
<span><span class="va">sc_object</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"new_name_a"</span>, <span class="st">"new_name_b"</span><span class="op">)</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">new_data_list</span></span></code></pre></div>
</div>
</section><section class="level3"><h4 id="var-table-gene-info">Var table (gene info)<a class="anchor" aria-label="anchor" href="#var-table-gene-info"></a>
</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Returns gene info as a data.table</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_sc_var.html">get_sc_var</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;    gene_idx  gene_id ensembl_id no_cells_exp</span></span>
<span><span class="co">#&gt;       &lt;num&gt;   &lt;char&gt;     &lt;char&gt;        &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:        1 gene_001    ens_001          758</span></span>
<span><span class="co">#&gt; 2:        2 gene_002    ens_002          742</span></span>
<span><span class="co">#&gt; 3:        3 gene_003    ens_003          772</span></span>
<span><span class="co">#&gt; 4:        4 gene_004    ens_004          724</span></span>
<span><span class="co">#&gt; 5:        5 gene_005    ens_005          533</span></span>
<span><span class="co">#&gt; 6:        6 gene_006    ens_006          724</span></span></code></pre></div>
</div>
</section><section class="level3"><h4 id="accessing-count-data">Accessing Count data<a class="anchor" aria-label="anchor" href="#accessing-count-data"></a>
</h4>
<p>This is where things get interesting and were the power of saving the mtx as both a Compressed sparse Row (CSR) and compressed sparse column (CSC) format becomes apparent. For gene wise operations (highly variable genes,PCA) we want to be able to rapidly index through the columns (genes) so we would opt for selecting the CSC notation of the matrix. For cell wise operations (DGEs, AUCell) we want to be able to rapidly index through the rows (cells) so we would opt for selecting the CSR notation of the matrix.</p>
<p>Most of the algorithms will automatically select the correct orientation so you don’t need to worry about it but below is a demonstration of the time difference when retrieving a selection of cells from the db</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## We are selecting 20:30 cells in the matrix which is a cellwise opperation so the correct thing to do would be to return the cells</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_sc_counts.html">get_sc_counts</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">sc_object</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  cell_indices <span class="op">=</span> <span class="fl">20</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>  return_format <span class="op">=</span> <span class="st">"cell"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; 0.005 sec elapsed</span></span>
<span></span>
<span><span class="co">## Now we are to select the incorrect orientation</span></span>
<span><span class="co">## This dataset is too small to see a difference however try it on your own larger and it will shock you</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_sc_counts.html">get_sc_counts</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">sc_object</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  cell_indices <span class="op">=</span> <span class="fl">20</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>  return_format <span class="op">=</span> <span class="st">"gene"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; 0.019 sec elapsed</span></span>
<span></span>
<span></span>
<span><span class="co">## matrices can also be accessed with sinle brackets</span></span>
<span></span>
<span><span class="va">matrix</span> <span class="op">&lt;-</span> <span class="va">sc_object</span><span class="op">[</span>, , return_format <span class="op">=</span> <span class="st">"cell"</span>, assay <span class="op">=</span> <span class="st">"norm"</span><span class="op">]</span></span></code></pre></div>
</div>
</section></section><section class="level2"><h3 id="gene-proportions">Gene proportions<a class="anchor" aria-label="anchor" href="#gene-proportions"></a>
</h3>
<p>Usual QC procedure involves checking the proportions of mitochondrial and ribosomal genes in you sample bixverse can do this as follows</p>
<p>This operation also be chunked by using by setting <code>streaming = TRUE</code>, this is useful when working with very large datasets to avoid running into memory constraints.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gs_of_interest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  gs_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_001"</span>, <span class="st">"gene_002"</span>, <span class="st">"gene_003"</span>, <span class="st">"gene_004"</span><span class="op">)</span>,</span>
<span>  gs_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_096"</span>, <span class="st">"gene_097"</span>, <span class="st">"gene_100"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## This will add proportions to the obs table</span></span>
<span><span class="va">sc_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_set_proportions_sc.html">gene_set_proportions_sc</a></span><span class="op">(</span></span>
<span>  <span class="va">sc_object</span>,</span>
<span>  <span class="va">gs_of_interest</span>,</span>
<span>  streaming <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># can be chunked when working on large datasets to avoid running into memory constraints</span></span>
<span>  .verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;    cell_idx   cell_id    cell_grp batch_index   nnz lib_size to_keep</span></span>
<span><span class="co">#&gt;       &lt;num&gt;    &lt;char&gt;      &lt;char&gt;       &lt;num&gt; &lt;num&gt;    &lt;num&gt;  &lt;lgcl&gt;</span></span>
<span><span class="co">#&gt; 1:        1 cell_0001 cell_type_1           1    52      371    TRUE</span></span>
<span><span class="co">#&gt; 2:        2 cell_0003 cell_type_3           1    52      495    TRUE</span></span>
<span><span class="co">#&gt; 3:        3 cell_0004 cell_type_1           1    48      551    TRUE</span></span>
<span><span class="co">#&gt; 4:        4 cell_0005 cell_type_2           1    53      422    TRUE</span></span>
<span><span class="co">#&gt; 5:        5 cell_0006 cell_type_3           1    57      394    TRUE</span></span>
<span><span class="co">#&gt; 6:        6 cell_0007 cell_type_1           1    49      593    TRUE</span></span>
<span><span class="co">#&gt;    random_new_data random_new_data2 other_random_data</span></span>
<span><span class="co">#&gt;             &lt;char&gt;           &lt;char&gt;            &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1: random_new_data  random_new_data                 A</span></span>
<span><span class="co">#&gt; 2: random_new_data  random_new_data                 A</span></span>
<span><span class="co">#&gt; 3: random_new_data  random_new_data                 A</span></span>
<span><span class="co">#&gt; 4: random_new_data  random_new_data                 A</span></span>
<span><span class="co">#&gt; 5: random_new_data  random_new_data                 A</span></span>
<span><span class="co">#&gt; 6: random_new_data  random_new_data                 A</span></span>
<span><span class="co">#&gt;    even_different_random_data new_name_a new_name_b        gs_1        gs_2</span></span>
<span><span class="co">#&gt;                         &lt;int&gt;     &lt;char&gt;      &lt;int&gt;       &lt;num&gt;       &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:                          1          A          1 0.156334236 0.008086253</span></span>
<span><span class="co">#&gt; 2:                          2          A          2 0.022222223 0.006060606</span></span>
<span><span class="co">#&gt; 3:                          3          A          3 0.181488201 0.000000000</span></span>
<span><span class="co">#&gt; 4:                          4          A          4 0.007109005 0.033175357</span></span>
<span><span class="co">#&gt; 5:                          5          A          5 0.017766498 0.015228426</span></span>
<span><span class="co">#&gt; 6:                          6          A          6 0.242833048 0.079258010</span></span></code></pre></div>
</div>
<p>We can then set cells to keep based on proportions of reads mapping to genes, the filtered out cells will not actullay be removed from the duckdb rather just have their <code>to_keep</code> column in the obs table set to FALSE.</p>
<p><code><a href="../reference/set_cells_to_keep.html">set_cells_to_keep()</a></code> accepts both cell ids and cell indices as a vector of cells to keep.</p>
<p>By default using the [[]] notation to load in the obs will only retain the <code>to_keep</code> cells. To access all cells the <code><a href="../reference/get_sc_obs.html">get_sc_obs()</a></code> needs to be used.</p>
<p>Once this <code><a href="../reference/set_cells_to_keep.html">set_cells_to_keep()</a></code> function has been applied to the <code>sc_object</code> only cells marked TRUE in the <code>to_keep</code> column will be used in the downstream analysis</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">threshold</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span></span>
<span><span class="va">cells_to_keep</span> <span class="op">&lt;-</span> <span class="va">sc_object</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">gs_2</span> <span class="op">&lt;</span> <span class="va">threshold</span>, <span class="va">cell_id</span><span class="op">]</span></span>
<span></span>
<span><span class="va">sc_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/set_cells_to_keep.html">set_cells_to_keep</a></span><span class="op">(</span><span class="va">sc_object</span>, <span class="va">cells_to_keep</span><span class="op">)</span></span>
<span></span>
<span><span class="va">obs_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_sc_obs.html">get_sc_obs</a></span><span class="op">(</span><span class="va">sc_object</span>, filtered <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">obs_data</span><span class="op">$</span><span class="va">to_keep</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; FALSE  TRUE </span></span>
<span><span class="co">#&gt;   247   682</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">[[</span><span class="st">"to_keep"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; to_keep</span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt;  682</span></span></code></pre></div>
</div>
</section><section class="level2"><h3 id="highly-variable-genes-hvg">Highly Variable Genes (HVG)<a class="anchor" aria-label="anchor" href="#highly-variable-genes-hvg"></a>
</h3>
<p>It is most common to perform downstream analysis (PCA, Knn graph, clustering) on HVG as these usually are the most informative genes.</p>
<p>Currently the only implemented method is the VST-based version (known as Seurat v3). The other methods (meanvarbin, dispersion) will be implemented in the future.</p>
<p>See <code><a href="../reference/params_sc_hvg.html">params_sc_hvg()</a></code> for details.</p>
<p>This also has the option for the data to be chunked in which is useful for larger datasets where you may run into memory constraints.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hvg_params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/params_sc_hvg.html">params_sc_hvg</a></span><span class="op">(</span></span>
<span>  method <span class="op">=</span> <span class="st">"vst"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">sc_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_hvg_sc.html">find_hvg_sc</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">sc_object</span>,</span>
<span>  hvg_params <span class="op">=</span> <span class="va">hvg_params</span>,</span>
<span>  hvg_no <span class="op">=</span> <span class="fl">30L</span>, <span class="co"># Number of HVGs</span></span>
<span>  .verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_sc_var.html">get_sc_var</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;    gene_idx  gene_id ensembl_id no_cells_exp      mean      var  var_exp</span></span>
<span><span class="co">#&gt;       &lt;num&gt;   &lt;char&gt;     &lt;char&gt;        &lt;int&gt;     &lt;num&gt;    &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:        1 gene_001    ens_001          758 11.747801 511.5848 2.652159</span></span>
<span><span class="co">#&gt; 2:        2 gene_002    ens_002          742 11.024927 388.6639 2.604914</span></span>
<span><span class="co">#&gt; 3:        3 gene_003    ens_003          772 13.149560 541.0187 2.724832</span></span>
<span><span class="co">#&gt; 4:        4 gene_004    ens_004          724  9.629032 312.5029 2.507178</span></span>
<span><span class="co">#&gt; 5:        5 gene_005    ens_005          533  4.703812 153.0825 1.926829</span></span>
<span><span class="co">#&gt; 6:        6 gene_006    ens_006          724 10.136364 351.4315 2.546862</span></span>
<span><span class="co">#&gt;      var_std</span></span>
<span><span class="co">#&gt;        &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: 1.1396177</span></span>
<span><span class="co">#&gt; 2: 0.9652928</span></span>
<span><span class="co">#&gt; 3: 1.0194833</span></span>
<span><span class="co">#&gt; 4: 0.9720214</span></span>
<span><span class="co">#&gt; 5: 1.8117409</span></span>
<span><span class="co">#&gt; 6: 0.9976512</span></span>
<span></span>
<span></span>
<span><span class="co">## You can then which genes are highly variable as follows</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_sc_var.html">get_sc_var</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="../reference/get_hvg.html">get_hvg</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">]</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;    gene_idx  gene_id ensembl_id no_cells_exp      mean       var  var_exp</span></span>
<span><span class="co">#&gt;       &lt;num&gt;   &lt;char&gt;     &lt;char&gt;        &lt;int&gt;     &lt;num&gt;     &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:        4 gene_004    ens_004          724  9.629032 312.50287 2.507178</span></span>
<span><span class="co">#&gt; 2:       16 gene_016    ens_016          655  7.756598 226.22502 2.403411</span></span>
<span><span class="co">#&gt; 3:       39 gene_043    ens_043          549  3.527859  33.90901 1.522625</span></span>
<span><span class="co">#&gt; 4:       18 gene_018    ens_018          750 10.665689 364.52734 2.581662</span></span>
<span><span class="co">#&gt; 5:        9 gene_009    ens_009          590  7.294722 246.94691 2.360737</span></span>
<span><span class="co">#&gt;      var_std</span></span>
<span><span class="co">#&gt;        &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: 0.9720214</span></span>
<span><span class="co">#&gt; 2: 0.8935734</span></span>
<span><span class="co">#&gt; 3: 1.0178647</span></span>
<span><span class="co">#&gt; 4: 0.9551431</span></span>
<span><span class="co">#&gt; 5: 1.0761360</span></span></code></pre></div>
</div>
</section><section class="level2"><h3 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h3>
<p>PCA is one of the most common form of dimensionally reduction algorithms.</p>
<p><code>bixverse</code> offers currently offers two implementations of the PCA. Traditional PCA or the randomised PCA which is a faster approximation, for details see <a href="https://research.facebook.com/publications/an-implementation-of-a-randomized-algorithm-for-principal-component-analysis/" class="external-link">here</a>. In reality the first principal components with most of the signal are basically identical between the two methods and it is in the latter principal components were the uncertainty of the randomised version increases. Longer term, there is also the plan to implement SVD version that work purely on sparse data without scaling (densifying the data) for VERY large data sets. PCs will only be calculated using counts from the HVGs as such the HVG calculation will need to be run prior to PCA.</p>
<p>Note the PCS are not currently saved to to the duckdb and after calculations are currently stored in memory and will need to be recalculated on the object if session is closed. Alternatively if you wish to save PCA results this can be done using <code><a href="../reference/save_sc_exp_to_disk.html">save_sc_exp_to_disk()</a></code> these will then be loaded back in when the <code><a href="../reference/load_existing.html">load_existing()</a></code> function is called. More info in <a href="#sec-loading-existing" class="quarto-xref">Section 3</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_pca_sc.html">calculate_pca_sc</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">sc_object</span>,</span>
<span>  no_pcs <span class="op">=</span> <span class="fl">10L</span>, <span class="co"># Number PC to calculate</span></span>
<span>  randomised_svd <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  .verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the PCA factors</span></span>
<span><span class="fu"><a href="../reference/get_pca_factors.html">get_pca_factors</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                 PC_1       PC_2       PC_3       PC_4        PC_5</span></span>
<span><span class="co">#&gt; cell_0001  0.7916713 -2.1312935 -1.3696928  0.4485300  2.23568630</span></span>
<span><span class="co">#&gt; cell_0003 -2.4218905 -1.2342623 -0.3744556 -1.7982795 -1.55481982</span></span>
<span><span class="co">#&gt; cell_0004  2.5719275 -1.4721915  1.2367083  0.8602270 -2.01101637</span></span>
<span><span class="co">#&gt; cell_0005  1.8453631  3.3306789 -0.3060451  0.8307743  2.15308475</span></span>
<span><span class="co">#&gt; cell_0006 -1.9956303 -0.4817278 -0.7805332 -1.3205884  0.06048276</span></span>
<span></span>
<span><span class="co"># Get PCA loadings for each gene</span></span>
<span><span class="fu"><a href="../reference/get_pca_loadings.html">get_pca_loadings</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                 PC_1       PC_2        PC_3         PC_4        PC_5</span></span>
<span><span class="co">#&gt; gene_005  0.20317698 -0.3045997  0.05578877 -0.010422311  0.01976835</span></span>
<span><span class="co">#&gt; gene_017  0.15731421  0.3486673 -0.03532208 -0.046772514  0.08763239</span></span>
<span><span class="co">#&gt; gene_045 -0.02124721  0.0529221 -0.20211859  0.028571185 -0.01772384</span></span>
<span><span class="co">#&gt; gene_019  0.13870865  0.3136349  0.11827516  0.008420937 -0.00150319</span></span>
<span><span class="co">#&gt; gene_010  0.17827085 -0.3270922 -0.01653624  0.012350544  0.01573452</span></span></code></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_plot_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="../reference/get_pca_factors.html">get_pca_factors</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pca_plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC_1</span>, y <span class="op">=</span> <span class="va">PC_2</span>, color <span class="op">=</span> <span class="va">cell_grp</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="Single_cell_analysis_files/figure-html/plot-pca-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h3 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h3>
<section class="level3"><h4 id="nearest-neighbors-identification">Nearest Neighbors Identification<a class="anchor" aria-label="anchor" href="#nearest-neighbors-identification"></a>
</h4>
<p>Nearest neighbor identification is a foundational step in single-cell analysis that finds the k most similar cells for each cell based on their expression profiles. This creates a k-nearest neighbor (kNN) graph, which is then converted into a shared nearest neighbor (sNN) graph—a more robust representation where edges represent cells that share neighbors. Here to ease computation we use the PCA factors to generate the graph.</p>
<p>To identify nearest neighbors in bixverse, use two complementary functions:</p>
<ul>
<li><p><code><a href="../reference/params_sc_neighbours.html">params_sc_neighbours()</a></code>configures the parameters for neighbor identification. Specify <code>k</code> (how many neighbors to find) and choose your search algorithm: <code>annoy</code> for speed or <code>hnsw</code> for accuracy and memory efficiency. Set the distance metric with ann_dist (typically <code>cosine</code> or <code>euclidean</code>). You can also control the sNN graph generation with <code>full_snn</code> (whether to create edges between all cells or just neighbors), <code>pruning</code> (to remove weak connections), and <code>snn_similarity</code> (choose <code>rank</code> or <code>jaccard</code> to determine how neighbor weights are calculated). This function returns a parameter list ready for neighbor detection.</p></li>
<li><p><code>find_neighbours()</code> applies these parameters to your single-cell object to generate the actual kNN and sNN graphs. Specify which embedding to use (<code>embd_to_use</code>, currently <code>pca</code>) and optionally limit the number of embedding dimensions with <code>no_embd_to_use</code>. Set a seed for reproducibility and use .verbose to track runtime information. The function adds the computed neighbor matrices to your object for use in downstream analyses.</p></li>
</ul>
<p>Like PCA the results for these are not stored in the duckdb rather in memory so will need to be saved or rerun each time.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">neighbours_params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/params_sc_neighbours.html">params_sc_neighbours</a></span><span class="op">(</span>knn_algorithm <span class="op">=</span> <span class="st">"hnsw"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">sc_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_neighbours_sc.html">find_neighbours_sc</a></span><span class="op">(</span></span>
<span>  <span class="va">sc_object</span>,</span>
<span>  neighbours_params <span class="op">=</span> <span class="va">neighbours_params</span>,</span>
<span>  .verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Generating kNN data with hnsw method.</span></span>
<span><span class="co">#&gt; Generating sNN graph (full: FALSE).</span></span>
<span><span class="co">#&gt; Transforming sNN data to igraph.</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_knn_mat.html">get_knn_mat</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5]</span></span>
<span><span class="co">#&gt; [1,]  597  607  416   13  666</span></span>
<span><span class="co">#&gt; [2,]  419    4  308  645   24</span></span>
<span><span class="co">#&gt; [3,]    8  630  311  632  489</span></span>
<span><span class="co">#&gt; [4,]  430  178  449  678  413</span></span>
<span><span class="co">#&gt; [5,]   24    1  152  168  419</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_snn_graph.html">get_snn_graph</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;                               </span></span>
<span><span class="co">#&gt; [1,] . .         . . .        </span></span>
<span><span class="co">#&gt; [2,] . .         . . 0.9333333</span></span>
<span><span class="co">#&gt; [3,] . .         . . .        </span></span>
<span><span class="co">#&gt; [4,] . .         . . .        </span></span>
<span><span class="co">#&gt; [5,] . 0.9333333 . . .</span></span></code></pre></div>
</div>
</section><section class="level3"><h4 id="community-detection">community detection<a class="anchor" aria-label="anchor" href="#community-detection"></a>
</h4>
<p>Currently only Leiden clustering is implemented in the package these cluster are then added to the obs table of the <code>sc_object</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_clusters_sc.html">find_clusters_sc</a></span><span class="op">(</span><span class="va">sc_object</span>, name <span class="op">=</span> <span class="st">"leiden_clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="../reference/get_pca_factors.html">get_pca_factors</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>leiden_clusters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">leiden_clusters</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC_1</span>, y <span class="op">=</span> <span class="va">PC_2</span>, color <span class="op">=</span> <span class="va">leiden_clusters</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="Single_cell_analysis_files/figure-html/leiden-clustering-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section class="level2"><h3 id="differential-gene-expression">Differential gene expression<a class="anchor" aria-label="anchor" href="#differential-gene-expression"></a>
</h3>
<p>Once the clustering is complete we can perform differential gene expression analysis between the 2 groups of cell lines or group verses all style analysis.</p>
<p>Currently the wilcox method is used to assess DGE with a default <code>twosided</code> test run.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_names_1</span> <span class="op">&lt;-</span> <span class="va">sc_object</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">leiden_clusters</span> <span class="op">==</span> <span class="fl">1</span>, <span class="va">cell_id</span><span class="op">]</span></span>
<span><span class="va">cell_names_2</span> <span class="op">&lt;-</span> <span class="va">sc_object</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">leiden_clusters</span> <span class="op">==</span> <span class="fl">2</span>, <span class="va">cell_id</span><span class="op">]</span></span>
<span></span>
<span><span class="va">dge_leiden_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_markers_sc.html">find_markers_sc</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">sc_object</span>,</span>
<span>  cells_1 <span class="op">=</span> <span class="va">cell_names_1</span>,</span>
<span>  cells_2 <span class="op">=</span> <span class="va">cell_names_2</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"wilcox"</span>,</span>
<span>  alternative <span class="op">=</span> <span class="st">"twosided"</span>,</span>
<span>  .verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dge_leiden_clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt;     gene_id      lfc     prop1     prop2 z_scores     p_values          fdr</span></span>
<span><span class="co">#&gt;      &lt;char&gt;    &lt;num&gt;     &lt;num&gt;     &lt;num&gt;    &lt;num&gt;        &lt;num&gt;        &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: gene_001 2.335875 0.9819005 0.7521739 16.09610 2.717026e-58 2.445323e-57</span></span>
<span><span class="co">#&gt; 2: gene_002 2.352085 0.9819005 0.7217391 16.34724 4.551388e-60 4.608280e-59</span></span>
<span><span class="co">#&gt; 3: gene_003 2.470452 1.0000000 0.7739130 17.17004 4.451217e-66 9.834427e-65</span></span>
<span><span class="co">#&gt; 4: gene_004 2.318489 0.9728507 0.6869565 15.64946 3.350099e-55 2.087369e-54</span></span>
<span><span class="co">#&gt; 5: gene_005 1.743340 0.8371041 0.4000000 12.08653 1.244202e-33 5.304231e-33</span></span>
<span><span class="co">#&gt; 6: gene_006 2.235862 0.9728507 0.7130435 15.89808 6.533862e-57 5.292428e-56</span></span></code></pre></div>
</div>
<p>Specify a column, and the function will calculate differential gene expression for each cluster against all other cells combined. It processes clusters sequentially: cluster one vs. all others, cluster two vs. all others, and so on. To manage computational efficiency, the “all others” group is automatically downsampled to a random 100,000 cells if it exceeds that threshold.</p>
<p>Currently this functions alternative hypothesis is set to “greater”, i.e., genes upregulated in the group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dge_test_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_all_markers_sc.html">find_all_markers_sc</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">sc_object</span>,</span>
<span>  column_of_interest <span class="op">=</span> <span class="st">"leiden_clusters"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"wilcox"</span>,</span>
<span>  .verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="level2"><h3 id="aucell">AUcell<a class="anchor" aria-label="anchor" href="#aucell"></a>
</h3>
<p>AUCell calculates how well a set of genes (or pathway) “ranks” in each cell compared to all other genes. It produces a score between 0 and 1 for each cell—higher scores mean the genes in your set are highly expressed in that cell relative to the background.</p>
<p>The function <code>aucell_sc</code> offers two scoring methods: use <code>auroc</code> for identifying cells expressing specific marker genes, or <code>wilcox</code> for measuring overall pathway activity. For large datasets, you can load all cells at once or stream them in chunks of 50,000 cells for memory efficiency.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">auc_gene_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  markers_cell_type_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"gene_%03d"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>  markers_cell_type_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"gene_%03d"</span>, <span class="fl">11</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span>,</span>
<span>  markers_cell_type_3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"gene_%03d"</span>, <span class="fl">21</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">auc_res_auroc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aucell_sc.html">aucell_sc</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">sc_object</span>,</span>
<span>  gs_list <span class="op">=</span> <span class="va">auc_gene_sets</span>,</span>
<span>  auc_type <span class="op">=</span> <span class="st">"auroc"</span>,</span>
<span>  .verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">auc_res_auroc</span><span class="op">)</span>, <span class="fu"><a href="../reference/get_pca_factors.html">get_pca_factors</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">PC_1</span>, x <span class="op">=</span> <span class="va">PC_2</span>, color <span class="op">=</span> <span class="va">markers_cell_type_1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="Single_cell_analysis_files/figure-html/auc-cell-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h3 id="meta-cell">meta cell<a class="anchor" aria-label="anchor" href="#meta-cell"></a>
</h3>
<p>Metacells are aggregated groups of similar cells that are merged together to create a higher-level representation of your data. This approach reduces noise, improves computational efficiency, and is particularly useful for correlation-based analyses like identifying co-regulated genes. Metacells are generated by grouping nearby cells in the neighborhood graph, creating more stable and interpretable results from single-cell data.</p>
<p>To generate metacells in bixverse package, use two complementary functions: <code><a href="../reference/params_sc_metacells.html">params_sc_metacells()</a></code> configures the parameters for metacell generation. Specify the maximum number of allowed shared neighbors (<code>max_shared</code>), your target number of metacells (<code>target_no_metacells</code>), and iteration limits (<code>max_iter</code>). You can also fine-tune the k-nearest neighbor search using k, knn_method (choose between <code>annoy</code> or <code>hnsw</code>), distance metric (<code>ann_dist</code>), and algorithm-specific parameters like n_trees and search_budget. This function returns a list of parameters ready for the metacell generation pipeline. <code>generate_metacells()</code> takes your single-cell object and the parameter list from <code><a href="../reference/params_sc_metacells.html">params_sc_metacells()</a></code> to generate the actual metacells. You can optionally regenerate the k-NN graph with regenerate_knn, specify which embedding to use (embd_to_use, currently “pca”), and control library size normalization with target_size. Set a seed for reproducibility.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_meta_cell_params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/params_sc_metacells.html">params_sc_metacells</a></span><span class="op">(</span>target_no_metacells <span class="op">=</span> <span class="fl">10L</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta_cell_data_v1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_meta_cells_sc.html">get_meta_cells_sc</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">sc_object</span>,</span>
<span>  sc_meta_cell_params <span class="op">=</span> <span class="va">sc_meta_cell_params</span>,</span>
<span>  .verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">meta_cell_data_v1</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;bixverse::meta_cells&gt;</span></span>
<span><span class="co">#&gt;  @ obs_table          :Classes 'data.table' and 'data.frame':    10 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;  .. $ meta_cell_idx       : int  1 2 3 4 5 6 7 8 9 10</span></span>
<span><span class="co">#&gt;  .. $ meta_cell_id        : chr  "meta_cell_01" "meta_cell_02" "meta_cell_03" "meta_cell_04" ...</span></span>
<span><span class="co">#&gt;  .. $ no_originating_cells: num  16 14 15 12 15 12 16 14 14 16</span></span>
<span><span class="co">#&gt;  .. $ original_cell_idx   :List of 10</span></span>
<span><span class="co">#&gt;  ..  ..$ : int  91 102 116 148 166 226 280 370 438 528 ...</span></span>
<span><span class="co">#&gt;  ..  ..$ : int  15 34 36 85 129 170 270 288 310 350 ...</span></span>
<span><span class="co">#&gt;  ..  ..$ : int  6 10 20 106 109 171 248 313 316 486 ...</span></span>
<span><span class="co">#&gt;  ..  ..$ : int  172 196 283 323 341 372 382 457 647 663 ...</span></span>
<span><span class="co">#&gt;  ..  ..$ : int  214 231 274 402 406 497 508 533 537 560 ...</span></span>
<span><span class="co">#&gt;  ..  ..$ : int  97 142 169 251 329 337 432 435 451 483 ...</span></span>
<span><span class="co">#&gt;  ..  ..$ : int  53 55 93 154 230 307 371 413 445 477 ...</span></span>
<span><span class="co">#&gt;  ..  ..$ : int  66 67 86 88 90 133 188 247 276 295 ...</span></span>
<span><span class="co">#&gt;  ..  ..$ : int  68 108 241 271 305 320 442 467 513 547 ...</span></span>
<span><span class="co">#&gt;  ..  ..$ : int  24 207 222 292 296 355 399 439 464 476 ...</span></span>
<span><span class="co">#&gt;  .. - attr(*, ".internal.selfref")=&lt;externalptr&gt; </span></span>
<span><span class="co">#&gt;  @ var_table          :Classes 'data.table' and 'data.frame':    81 obs. of  2 variables:</span></span>
<span><span class="co">#&gt;  .. $ gene_idx: num  1 2 3 4 5 6 7 8 9 10 ...</span></span>
<span><span class="co">#&gt;  .. $ gene_id : chr  "gene_001" "gene_002" "gene_003" "gene_004" ...</span></span>
<span><span class="co">#&gt;  .. - attr(*, ".internal.selfref")=&lt;externalptr&gt; </span></span>
<span><span class="co">#&gt;  @ data               :List of 2</span></span>
<span><span class="co">#&gt;  .. $ raw :Formal class 'dgRMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;  ..  .. ..@ p       : int [1:11] 0 81 162 243 324 405 486 567 648 729 ...</span></span>
<span><span class="co">#&gt;  ..  .. ..@ j       : int [1:810] 0 1 2 3 4 5 6 7 8 9 ...</span></span>
<span><span class="co">#&gt;  ..  .. ..@ Dim     : int [1:2] 10 81</span></span>
<span><span class="co">#&gt;  ..  .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">#&gt;  ..  .. .. ..$ : chr [1:10] "meta_cell_01" "meta_cell_02" "meta_cell_03" "meta_cell_04" ...</span></span>
<span><span class="co">#&gt;  ..  .. .. ..$ : chr [1:81] "gene_001" "gene_002" "gene_003" "gene_004" ...</span></span>
<span><span class="co">#&gt;  ..  .. ..@ x       : num [1:810] 251 167 263 102 54 159 190 319 162 80 ...</span></span>
<span><span class="co">#&gt;  ..  .. ..@ factors : list()</span></span>
<span><span class="co">#&gt;  .. $ norm:Formal class 'dgRMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;  ..  .. ..@ p       : int [1:11] 0 81 162 243 324 405 486 567 648 729 ...</span></span>
<span><span class="co">#&gt;  ..  .. ..@ j       : int [1:810] 0 1 2 3 4 5 6 7 8 9 ...</span></span>
<span><span class="co">#&gt;  ..  .. ..@ Dim     : int [1:2] 10 81</span></span>
<span><span class="co">#&gt;  ..  .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">#&gt;  ..  .. .. ..$ : chr [1:10] "meta_cell_01" "meta_cell_02" "meta_cell_03" "meta_cell_04" ...</span></span>
<span><span class="co">#&gt;  ..  .. .. ..$ : chr [1:81] "gene_001" "gene_002" "gene_003" "gene_004" ...</span></span>
<span><span class="co">#&gt;  ..  .. ..@ x       : num [1:810] 7.99 7.58 8.04 7.09 6.45 ...</span></span>
<span><span class="co">#&gt;  ..  .. ..@ factors : list()</span></span>
<span><span class="co">#&gt;  @ original_assignment:List of 4</span></span>
<span><span class="co">#&gt;  .. $ assignments : int [1:682] -1 -1 -1 -1 -1 3 -1 -1 -1 3 ...</span></span>
<span><span class="co">#&gt;  .. $ n_cells     : num 682</span></span>
<span><span class="co">#&gt;  .. $ n_metacells : num 10</span></span>
<span><span class="co">#&gt;  .. $ n_unassigned: num 538</span></span>
<span><span class="co">#&gt;  @ dims               : int [1:2] 10 81</span></span>
<span><span class="co">#&gt;  @ other_data         : list()</span></span>
<span><span class="co">#&gt;  @ meta_cell_method   : chr "meta_cells_hdwgcna"</span></span></code></pre></div>
</div>
</section><section class="level2"><h3 id="save-object">Save object<a class="anchor" aria-label="anchor" href="#save-object"></a>
</h3>
<p>As analysis progresses some parts are written to the duckdb and saved and others are not.Any updates to the <code>obs</code> table and to the <code>to_keep</code> column (number of cells) will be saved to the duckdb and will be present when the data is loaded back in using <code><a href="../reference/load_existing.html">load_existing()</a></code>.</p>
<p>However, other aspects which are held in memory e.g HVG, PCA and AUcell will not be saved. To ensure these are saved and then loaded back in when <code><a href="../reference/load_existing.html">load_existing()</a></code> is used run <code><a href="../reference/save_sc_exp_to_disk.html">save_sc_exp_to_disk()</a></code>.</p>
<p>This will create an <code>memory.qs2</code> object in the same directory as the <code>counts_cells.bin</code>, <code>counts_genes.bin</code> and <code>sc_duckdb.db</code></p>
<p>Objects can either be saved as a <code>rds</code> or <code>qs2</code> it is highly recommended the <code>qs2</code> format is used as the reading in is far more performant.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/save_sc_exp_to_disk.html">save_sc_exp_to_disk</a></span><span class="op">(</span><span class="va">sc_object</span>, type <span class="op">=</span> <span class="st">"qs2"</span><span class="op">)</span></span></code></pre></div>
</div>
</section></section><section class="section level2"><h2 id="sec-loading-existing">Loading in existing object<a class="anchor" aria-label="anchor" href="#sec-loading-existing"></a>
</h2>
<p>A series of functions have been written to load in a already pre-processed sc_object where the <code>to_keep</code> cells have already been defined.</p>
<p>This is done by pointing the directory where the <code>sc_duckdb.db</code>, <code>counts_cells.bin</code> and <code>counts_genes.bin</code> have been created.</p>
<p>All information that was stored in the obs table leiden_clusters, gene proportions etc and any information in var table will be retained as it has been stored in the duckdb. However analysis that was created and stored in memory e.g knn and PCA will no longer exist and have to be rerun (unless data has been saved with <code><a href="../reference/save_sc_exp_to_disk.html">save_sc_exp_to_disk()</a></code>, recommended).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_object_loaded</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/single_cell_exp.html">single_cell_exp</a></span><span class="op">(</span>dir_data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sc_object_loaded</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_existing.html">load_existing</a></span><span class="op">(</span><span class="va">sc_object_loaded</span><span class="op">)</span></span>
<span><span class="co">#&gt; Found stored data from save_sc_exp_to_disk(). Loading that one into the object.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sc_object_loaded</span><span class="op">[</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 682  81</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">[</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 682  81</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 682  16</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sc_object_loaded</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 682  16</span></span></code></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## PCA in original object exists</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_pca_factors.html">get_pca_factors</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                 PC_1       PC_2       PC_3       PC_4        PC_5</span></span>
<span><span class="co">#&gt; cell_0001  0.7916713 -2.1312935 -1.3696928  0.4485300  2.23568630</span></span>
<span><span class="co">#&gt; cell_0003 -2.4218905 -1.2342623 -0.3744556 -1.7982795 -1.55481982</span></span>
<span><span class="co">#&gt; cell_0004  2.5719275 -1.4721915  1.2367083  0.8602270 -2.01101637</span></span>
<span><span class="co">#&gt; cell_0005  1.8453631  3.3306789 -0.3060451  0.8307743  2.15308475</span></span>
<span><span class="co">#&gt; cell_0006 -1.9956303 -0.4817278 -0.7805332 -1.3205884  0.06048276</span></span>
<span></span>
<span><span class="co">## PCA in loaded in object does exist as has been loaded back in from the qs2 file</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_pca_factors.html">get_pca_factors</a></span><span class="op">(</span><span class="va">sc_object_loaded</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 PC_1       PC_2       PC_3       PC_4        PC_5        PC_6</span></span>
<span><span class="co">#&gt; cell_0001  0.7916713 -2.1312935 -1.3696928  0.4485300  2.23568630  0.10833390</span></span>
<span><span class="co">#&gt; cell_0003 -2.4218905 -1.2342623 -0.3744556 -1.7982795 -1.55481982 -1.73168576</span></span>
<span><span class="co">#&gt; cell_0004  2.5719275 -1.4721915  1.2367083  0.8602270 -2.01101637 -0.98504061</span></span>
<span><span class="co">#&gt; cell_0005  1.8453631  3.3306789 -0.3060451  0.8307743  2.15308475 -0.88234264</span></span>
<span><span class="co">#&gt; cell_0006 -1.9956303 -0.4817278 -0.7805332 -1.3205884  0.06048276 -1.61810422</span></span>
<span><span class="co">#&gt; cell_0008  0.8654820  0.2562445 -1.4927127  1.1618458  1.81161833  0.09957793</span></span>
<span><span class="co">#&gt;                 PC_7       PC_8        PC_9      PC_10</span></span>
<span><span class="co">#&gt; cell_0001 -0.8307593 -0.9369759  0.11120906  0.1123955</span></span>
<span><span class="co">#&gt; cell_0003  0.4090152  0.5485656 -0.03805275  0.6408936</span></span>
<span><span class="co">#&gt; cell_0004 -2.1615980  0.6148190  0.08886870  0.5149482</span></span>
<span><span class="co">#&gt; cell_0005  1.3377242  0.8726504  0.93318295  0.5300551</span></span>
<span><span class="co">#&gt; cell_0006 -0.5391674  0.1609078 -0.24828078  1.1093334</span></span>
<span><span class="co">#&gt; cell_0008 -0.1490478  1.2157899  1.11126518 -1.0982429</span></span>
<span></span>
<span></span>
<span><span class="co">## Same for knn</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_knn_mat.html">get_knn_mat</a></span><span class="op">(</span><span class="va">sc_object</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5]</span></span>
<span><span class="co">#&gt; [1,]  597  607  416   13  666</span></span>
<span><span class="co">#&gt; [2,]  419    4  308  645   24</span></span>
<span><span class="co">#&gt; [3,]    8  630  311  632  489</span></span>
<span><span class="co">#&gt; [4,]  430  178  449  678  413</span></span>
<span><span class="co">#&gt; [5,]   24    1  152  168  419</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_knn_mat.html">get_knn_mat</a></span><span class="op">(</span><span class="va">sc_object_loaded</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14]</span></span>
<span><span class="co">#&gt; [1,]  597  607  416   13  666  300  177  453  654    96   507   127    99   605</span></span>
<span><span class="co">#&gt; [2,]  419    4  308  645   24  665  168  558  390   104   536   217   452    64</span></span>
<span><span class="co">#&gt; [3,]    8  630  311  632  489   83  383  381  183    40   119    32   253   671</span></span>
<span><span class="co">#&gt; [4,]  430  178  449  678  413  644  315  626  189   483    14   620   382    61</span></span>
<span><span class="co">#&gt; [5,]   24    1  152  168  419   94  514  221  308   190   602   277    15   231</span></span>
<span><span class="co">#&gt; [6,]  189  318  633  236  453  666  461  290  550    25   564   266   191   661</span></span>
<span><span class="co">#&gt;      [,15]</span></span>
<span><span class="co">#&gt; [1,]   424</span></span>
<span><span class="co">#&gt; [2,]   465</span></span>
<span><span class="co">#&gt; [3,]   292</span></span>
<span><span class="co">#&gt; [4,]   238</span></span>
<span><span class="co">#&gt; [5,]    73</span></span>
<span><span class="co">#&gt; [6,]   663</span></span></code></pre></div>
</div>
</section><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config);
    }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gregor Lueg.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
